<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" >
    <head>
        <title>KGGSeq Online User Manual</title>
        <style type="text/css">       

            /* overall styles */
            body {
                font:                   11pt Arial;
                margin-top:             0.5em;
                margin-bottom:          0.5em;
                margin-left:            2px;
                margin-right:           0;
                color:                  black;
                line-height:            1.3em;
            }
            #menu {
                display: block;
                top: 5px;
                right: 5px;
                width: 80px;
                position: fixed;
                border: solid 1px #cccccc;
                font-family: Arial, serif;
                background-color: #C7EDCC
            }
            * html #menu {
                position:absolute;
                top: expression(offsetParent.scrollTop+document.documentElement.clientHeight-this.offsetHeight)
            }

            .links
            {
                margin: 0px;
                padding:        6px;
                border: 0px inset #000000;
                max-width: 90%;
                background:     white;
            }

            /* various link styles */
            a
            {
                margin-right:           0;
                text-decoration:        none;
                border-bottom:          1px solid gray;
                white-space:            nowrap;
            }

            a:link
            {
                color:                  black;
            }

            a:visited
            {
                color:                  black;
                border-bottom:          1px solid gray;
            }

            a:active
            {
                color:                  yellow;
            }

            a:hover
            {
                color:                   #5A867B;
                border-bottom:          1px solid black;
            }

            h1 {
                margin-left:            0;
                margin-right:           0;
                width: 100%;
                padding-left:  5px;
                padding-right:  20px;
                margin-left:            0px;
                margin-bottom:          -1px;
                text-align: justify;
                font-size: 150%;
                font-weight:bold;
            }
            /* this is the indenting procedure for major categories */
            h2
            {
                border-left: 2px solid #225522;
                border-top: 2px solid #225522;
                margin-left:            0;
                margin-right:           1px;
                width: 100%;
                padding-left:  5px;
                padding-right:  20px;
                background:     #f0f5f0;
                text-align: justify;
                font-size: 120%;
            }

            /* this is for category headings */
            h3
            {
                border:                 1px dotted #000000;
                border-top:             1px dotted #000000;
                border-right:           1px dotted #000000;
                padding-bottom:         2px;
                padding-top:            0px;
                padding-right:          20px;
                padding-left:           20px;
                margin-left:            0px;
                margin-top:             3em;
                margin-bottom:          -1px;
                background:             #335522;
                font-size:              100%;
                color:                  #FFFFFF;
                font-weight:            normal;
            }

            h4
            {
                border:#CCC 1px dotted;
                padding-bottom:         2px;
                padding-top:            0px;
                padding-right:          20px;
                padding-left:           20px;
                margin-left:            10px;
                margin-right:           10px;
                /*                margin-top:             3em; */
                margin-bottom:          -1px;
                background:#E8E8E8           ;
                font-size:		90%;
                text-align: 		justify;
                font-weight:        normal;
                font-style:italic;
            }

            strong
            {
                border:                 1px solid #000000;
                padding-bottom:         1px;
                padding-top:            1px;
                padding-right:          20px;
                padding-left:           20px;
                margin-left:            1px;
                margin-top:             5em;
                margin-bottom:          2px;
                background:#FCF            ;
                font-size:              100%;
                font-weight:            normal;
            }

            code {
                display:block;
                style: overflow: auto;
                /* white-space: pre; */
                white-space: pre-wrap;       /* css-3 */
                white-space: -moz-pre-wrap !important;  /* Mozilla, since 1999 */
                white-space: -pre-wrap;      /* Opera 4-6 */
                white-space: -o-pre-wrap;    /* Opera 7 */
                word-wrap: break-word;       /* Internet Explorer 5.5+ */

                color: #333;
                font-family: courier;
                font-size: 95%;
                margin-top: 20px;
                margin-left: 40px;
                margin-right: 40px;
                margin-bottom: 10px;
                padding-left: 10px;
                padding-right: 10px;
                background-color: #F4F4FF;
                border-left: 2px solid;
                border-color: #337;
            }

            table.list {
                width:80%;
                text-align:justify;
                background-color:#F5F5F5;
                margin-left:auto;
                margin-right:auto;
                margin-top:5px;
                margin-bottom:5px;
                border-collapse: collapse;
                border:1px dotted #66CC00;
            }
            table.list TD {
                border:1px dashed  #999999;

            }


            .textsurround{float: right;
                          margin-right: 5px;}          


        </style>
        <!--[if IE 6]>
            <style type="text/css">
            /*<![CDATA[*/
        html {overflow-x:auto; overflow-y:hidden;}
            /*]]>*/
            </style>
        <![endif]-->
    </head>

    <body>
        <div style="position:absolute;left: 3px ;top: 3px">
            <table width="95%" border="0" cellspacing="0" cellpadding="3" align="center">
                <tr>
                    <td  width="180px" align="left" valign="top" bgcolor="C7EDCC">
                        <p><a name="Navi:introduction" href="../">KGGSeq Home</a></p>
                        <p><a name="Navi:introduction" href="#introduction">Introduction & References</a></p>
                        <p>   
                            <a name="Navi:installandrun">Install  and run</a><br />
                            <img src="imgs/head1.gif" width="12" height="12" />    <a href="#SystemRequirement">System       requirement</a><br />
                            <img src="imgs/head1.gif" width="12" height="12" />    <a href="#InstallingKGGSeq">Installing       KGGSeq</a>   <br />
                        </p>
                        <p> 
                            <a name="Navi:shorttutorials"> Short Tutorials</a><br />
                            <img src="imgs/head1.gif" width="12" height="12" />    <a href="MendelianDisease.htm" target="_blank">Mendelian Disease</a><br />
                            <img src="imgs/head1.gif" width="12" height="12" />      <a href="SomaticDriver.htm" target="_blank">Cancer Somatic</a><br />
                            <img src="imgs/head1.gif" width="12" height="12" />     <a href="DoubleHitGene.htm" target="_blank">Double Hit Gene</a><br />
                            <img src="imgs/head1.gif" width="12" height="12" />     <a href="ComplexDisease.htm" target="_blank">Complex Disease</a>
                        </p>
                        <p>
                            <a name="Navi:quickexamples"> Quick examples</a> <br />
                            <img src="imgs/head1.gif" width="12" height="12" />  <a href="#MendelianDiseaseHg19">Mendelian disease</a>  
                        </p>
                        <p>
                            <a name="Navi:options"> Options</a> <br />
                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#EnvironmentalSettings">General settings</a><br />
                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#InputFormats">Inputs</a><br />
                            &nbsp;&nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#VariantsAndGenotypes">Variants&amp;genotypes</a><br />
                            &nbsp;&nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#PhenotypesAndPedigree">Phenotype&amp;pedigree</a><br />
                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#Outputs"> Outputs</a><br />

                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#OdinaryOutputs">Ordinary outputs</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#BinaryOutputs">Binary outputs</a><br />
                            <br />

                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#QualityControl" target="_self">Quality control</a><br />

                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#GenotypeQC" target="_self">Genotype QC</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#VariantQC" target="_self">Variant QC</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#SampleQC" target="_self">Sample QC</a><br />
                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#FilteringAndPrioritization" target="_self">Filtering </a><br />

                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#GeneticInheritanceSharing" target="_self">Genetic inheritance </a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#GeneFeatureFiltering" target="_self">Gene feature</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#AlleleFrequencyTrimming" target="_self">Allele frequency</a> <br />                                   
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#VariantTypeAndRegionFiltering" target="_self">Type&amp;region&amp;gene</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#SupperDuplicateRegionFiltering" target="_self">Supper duplicate</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#TooManyVarGeneFiltering" target="_self">Gene&#39;s variant number</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#PhenotypeRelevantFiltering" target="_self">Phenotype</a><br />

                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#Annotation" target="_self">Annotation</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#dbSNPRSID" target="_self">dbSNP RSID</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#splicingAnnote" target="_self">Alternative splicing</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#dgvcnvAnnote" target="_self">Structure variation</a><br />

                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#SharedProteinProteinInteractions" target="_self">Protein interaction</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#SharedGeneSets" target="_self">GeneSet</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#MousePhenotype" target="_self">Mouse phenotype</a><br />            
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#zebrafihserPhenotype" target="_self">Zebrafish Phenotype</a><br />  
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#ddditem" target="_self">DDD Study</a><br />  
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#PubMedTextMining" target="_self">PubMed literature</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#OMIMTextMining" target="_self">OMIM</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#COSMICMining" target="_self">COSMIC</a><br />


                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#PredictionVar" target="_self">Prediction at variant</a><br />

                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#PredictingMendelianDiseaseCausingVariants" target="_self">Mendelian disease</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#PredictingCancerCausingVariants" target="_self">Cancer somatic driver</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#PredictingComplexDiseaseCausingVariants" target="_self">Complex disease</a><br />                                 
                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#PredictionGene" target="_self">Prediction at gene</a><br />

                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#Pathogenicity" target="_self">Pathogenicity </a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#PhenotypeMining" target="_self">Phenotype mining</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#Expression" target="_self">Expression</a><br />                                    
                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" /> <a href="#AssociationEnrichment" target="_self">Statistical tests</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#AssociationVariant" target="_self">Association at variant </a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#AssociationGene" target="_self">Association at gene</a><br />
<<<<<<< HEAD
                            &nbsp; &nbsp; &nbsp;<img src="imgs/circle.gif" width="10" height="8" /><a href="#GeneSkat" target="_self">SKAT</a><br />
                            &nbsp; &nbsp; &nbsp;<img src="imgs/circle.gif" width="10" height="8" /><a href="#GeneRvtests" target="_self">Rvtests</a><br />
                            <!--
                            &nbsp; &nbsp; &nbsp;<img src="imgs/circle.gif" width="10" height="8" /><a href="#GeneDoubleHitTrio" target="_self">Double-hit</a><br />
                            -->
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#MutationRateGene" target="_self">Mutation rate at gene</a><br />      
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#GeneSetAssociation" target="_self">Association at geneset</a><br /> 
                            &nbsp; &nbsp; &nbsp;<img src="imgs/circle.gif" width="10" height="8" /><a href="#GenesetSkat" target="_self">SKAT</a><br />
                            &nbsp; &nbsp; &nbsp;<img src="imgs/circle.gif" width="10" height="8" /><a href="#GenesetRvtests" target="_self">Rvtests</a><br />

=======
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#MutationRateGene" target="_self">Mutation rate at gene</a><br />      
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#GeneSetAssociation" target="_self">Association at geneset</a><br />      
>>>>>>> origin/master
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#GeneSetEnrichment" target="_self">Enrichment at geneset</a><br />      
                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" />  <a href="#Plotting" target="_self">Plotting</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#Plotting" target="_self">Allele frequency</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#Plotting" target="_self">Q-Q</a><br />
                            <br />
                            <img src="imgs/head1.gif" width="12" height="12" />  <a href="#Miscellaneous" target="_self">Miscellaneous</a><br />
<<<<<<< HEAD
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#calLD" target="_self">Linkage Disequilibrium</a><br />
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#ldPrune" target="_self">LD Pruning</a><br />
=======
                            &nbsp; &nbsp;<img src="imgs/leaf.gif" width="10" height="8" /><a href="#Miscellaneous" target="_self">Linkage Disequilibrium</a><br />

>>>>>>> origin/master

                        </p>
                        <a name="Navi:faq" href="#faq"> FAQ</a>
                    </td>

                    <td>
                        <table width="97%" border="0" cellspacing="5" cellpadding="0" align="center">
                            <tr> 
                                <td >
                                    <p> <h1> KGGSeq: A biological Knowledge-based  mining platform for Genomic and Genetic studies using Sequence data </h1> </p>
                                    <p align="center" >(Version 1.0)<br />
                                        Miaoxin Li,Jiang Li, Zhicheng Pan, Jun Li, Jacob Shujui Hsu, Hongsheng Gui, Suying Bao and  JS Kwan</p>
                                    <hr />
                                    <h1><a id="introduction"> Introduction </a> </h1><br />   
                                    <b>KGGSeq</b> is designed to make use of biological knowledge and statistical models to identify genetic loci responsible for human diseases/traits by high-throughput sequencing data. Compared with other tools like PLINK/Seq (&ldquo;A  library for the analysis of genetic variation data&rdquo;, http://atgu.mgh.harvard.edu/plinkseq/),  KGGSeq focuses more on integrative downstream analysis. Since version 1.0, it has been equipped with a series of integrative methods for characterizing genetic mutations/polymorphsim of Mendelian disorders, complex disorders and cancers by using whole genome sequencing data and versatile  public resources  .

                                </td>
                            </tr>
                            <tr> 
                                <td >
                                    <!-- 
                                                <img src="imgs/pipeline.jpg" width="463" height="508" alt="WorkFlowFig"  class="textsurround" /> 	
                                         width="90%" height="90%" 
<<<<<<< HEAD
                                    -->
=======
                                    !-->
>>>>>>> origin/master
                                    <img src="imgs/pipeline.png"   width="632" height="655"  alt="WorkFlowFig" /> 
                                </td>
                            </tr>
                            <tr >
                                <td><h3>References</h3>
                                    1. &nbsp;&nbsp;Li MX, Gui HS, Kwan JS, Bao SY, Sham PC. A comprehensive framework for prioritizing variants in exome sequencing studies of Mendelian diseases.Nucleic Acids Res. 2012 Apr;40(7):e53.  <a href="http://www.ncbi.nlm.nih.gov/pubmed/22241780" target="_blank">PubMed</a>&nbsp; &nbsp;<a href="http://nar.oxfordjournals.org/content/40/7/e53.long" target="_blank">NAR</a>
                                    <b>&lsaquo;For the entire filtration, annotation and prioritization framework&rsaquo;</b><br />
                                    2. &nbsp;&nbsp;Li et al. Predicting Mendelian disease-causing non-synonymous single nucleotide variants in exome sequencing studies. PLoS Genet. 2013 Jan;9(1):e1003143. <a href="http://www.ncbi.nlm.nih.gov/pubmed/23341771" target="_blank">PubMed</a> &nbsp; &nbsp;<a href="http://www.plosgenetics.org/article/info%3Adoi/10.1371/journal.pgen.1003143" target="_blank">PLoS Genet</a>
                                    <b>&lsaquo;For Mendelian pathogenic prediction of non-synonymous variants by logistic regression&rsaquo;</b><br />
                                    3. &nbsp;&nbsp;Li et al.KGGSeq: a software platform for a comprehensive downstream analysis in large-scale whole genome sequencing studies.  (Submitted)
                                    <b>&lsaquo;For the new version of KGGSeq since version 1.0&rsaquo;</b><br />
                                </td>
                            </tr>
                            <tr >
                                <td>
                                    <hr />
                                    <h1><a name="Navi:installation" href="#installation">Install and run</a></h1>
                                    <h2> <a id="SystemRequirement" > System requirement </a>   </h2>
                                    <p>Java Runtime Environment (JRE) version 1.6 (or up) is required for KGGSeq. It can be downloaded from the Java web site.
                                        Installing the JRE is very easy in Windows OS and Mac OS X.<br />
                                        <br />
                                        In Linux, you have more work to do. Details of the installation can be found <a  href="http://www.java.com/en/download/help/linux_install.xml" target="_blank">http://www.java.com/en/download/help/linux_install.xml.</a>                             <br />
                                        <br />
                                        In Ubuntu, if you have an error message like: &quot;Exception in thread &quot;AWT-EventQueue-0&quot; java.awt.HeadlessException ... &quot; , please install the Sun Java Running Environment (JRE) first.

                                        <h4>To install the Sun JRE on Ubuntu(10.04), please use the following commands:<br />
                                            sudo add-apt-repository "deb http://archive.canonical.com/ lucid partner"<br />
                                            sudo apt-get update<br />
                                            sudo apt-get install sun-java6-jre sun-java6-plugin sun-java6-fonts<br />
                                            Detailed explanation of above commands can be found at http://www.ubuntugeek.com/how-install-sun-java-runtime-environment-jre-in-ubuntu-10-04-lucid-lynx.html.</h4>
                                        <br />

                                        <p>For Mac OS, the JRE 1.6 has been available at http://developer.apple.com/java/download/ since April 2008. Mac OS users may need update the Java application to run KGGSeq. A potential problem is that this update does not replace the existing installation of J2SE 5.0 or change the default version of Java. Similar to the Linux OS, the Java_Home environmental variable has to be configured to initiate KGGSeq. </p>
                                        <p>&nbsp;</p>
                                </td>
                            </tr>
                            <tr >
                                <td >
                                    <h2><a  id="InstallingKGGSeq" >Installing KGGSeq</a> </h2>
                                    <p>Simply decompress the archive and run the  following command</p>
                                    <p><code> &nbsp; java -Xmx10g -jar ./kggseq.jar  &lt;arguments&gt; </code></p>
                                    <p>The arguments <tt>-Xmx10g</tt> set the maximum Java  heap sizes for KGGSeq as 10 gigabytes. Specifying a larger maximum heap size can speed up the analysis when processing large datasets. A higher setting like <tt>-Xmx20g</tt> is required when there is over 50 million. The number, however, should be less than the size of physical memory of a machine. </p>
                                    <p><strong>Note</strong>&nbsp;  <font class="textMark">&lt;arguments &gt;</font><b>can be saved in a flat text file.</b></p>
                                    <p>
                                        <strong>Note</strong>&nbsp;  For some advanced statistical analysis in which some third-party tools are called, you have to run R and C++ make command to install them. Please watch our <a  href="http://grass.cgs.hku.hk/limx/kggseq/doc10/demov10.mp4" target="_blank">demo video</a>  for details.
                                    </p>

                                    <br /> <strong>Note</strong>&nbsp;  To use KGGSeq under Proxy  Network, you need specifically use the following java command to configure the  proxy settings (Thank Daniele Yumi for this suggestion): <br />
                                    <code>  java -Dhttp.proxyHost=xxx.xxx.xxx -Dhttp.proxyPort=xxxx -Dhttp.proxyUser=xxx -Dhttp.proxyPassword=xxx -jar "./kggseq.jar"  </code> 
                                </td>
                            </tr>
                            <tr >
                                <td >
                                    <hr />
                                    <h1>Quick   example</h1>

                                    <p ><a   id="MendelianDiseaseHg19" class="h2"> Example : </a> <b> Prioritize variants for a rare Mendelian dominant disease, Arthrogryposis </b></p>
                                    <p>Files needed:<br />
                                        1)a VCF file (rare.disease.hg19.vcf); and<br />
                                        2)a linkage pedigree file (rare.disease.ped.txt). </p>
                                    <p><strong>Note</strong>&nbsp; <b> All files  were included in the examples folder of KGGSeq. </b></p>
                                    <p>Run the command below:</p>
                                    <code>java -Xmx5g -jar kggseq.jar examples/param.rare.disease.hg19.txt</code>
                                    <p>We now walk through the parameter file &ldquo;param.rare.disease.hg19.txt&rdquo;  before going into the results. Lines starting with hash sign <b>#</b> are  comments. Detailed interpretation for each argument in the parameter file is  included in 'Options' part.</p>

                                    <h4>

                                        <p>#one argument per line<br />
                                            #I.Environmental setting<br />
                                            --buildver hg19 \                           \ #line 1<br />
                                            --nt 4 \                    \ #line 2<br />
                                            #--lib-update                              \ #line 3<br />
                                            #--resource-update \                       \ #line 4</p>
                                        <p>#II. Specify the input files<br />
                                            --vcf-file examples/rare.disease.hg19.vcf    \ #line 5<br />
                                            --ped-file examples/rare.disease.ped.txt     \ #line 6, or specify --indiv-pheno X:1,Y:1,Z:2</p>
                                        <p>#III. Output setting<br />
                                            --out ./test1                           \ #line 7<br />
                                            --excel                                      \ #line 8<br />                                    
                                            --o-vcf                                      \ #line 10<br />
                                            #--o-flanking-seq 50                         \ #line 11, need large RAM memory</p>
                                        <p>#IV. QC<br />
                                            --gty-qual 20                                \ #line 12<br />
                                            --gty-dp 4                                   \ #line 13<br />
                                            --gty-af-ref 0.05                            \ #line 14<br />
                                            --gty-af-het 0.25                            \ #line 15<br />
                                            --vcf-filter-in PASS,VQSRTrancheSNP90.00to93.00,VQSRTrancheSNP93.00to95.00,VQSRTrancheSNP95.00to97.00,VQSRTrancheSNP97.00to99.00                         \ #line 16<br />
                                            --seq-qual 30                                \ #line 17<br />
                                            --seq-mq 20                                  \ #line 18<br />
                                            --seq-fs 60                                   \ #line 19<br />
                                            --min-obsa 1                                 \ #line 20<br />
                                        </p> 
                                        <p>#V. Filtering<br />
                                            --genotype-filter 3,4,5,6                          \ #line 21 for dominant mode<br />
                                            #--ibs-case-filter 1000                                 \ #line 22, or specify 'ibdregions.txt' file<br />
                                            --regions-out chrX,chrY                   \ #line 23<br />
                                            --db-filter 1kg201204,dbsnp141,ESP6500AA,ESP6500EA \ #line 24<br />
                                            --rare-allele-freq 0.006                     \ #line 25<br />
                                            --db-filter-hard dbsnp138nf   \ #line 26<br />
                                            --db-gene refgene,gencode,knowngene                            \ #line 27<br />
                                            --gene-feature-in 0,1,2,3,4,5,6                \ #line 28<br />
                                            --superdup-filter                        \  #line 29<br />
                                            --gene-var-filter 4                  \ #line 30
                                        </p>
                                        <p>#VI. Annotation <br />
                                            --scsnv-annot                                \#line 31<br />
                                            --dgv-cnv-annot                 \#line 32<br />
                                            --omim-annot                   \#line 32<br />
                                            --candi-list ECEL1,MYBPC1,TNNI2,TNNT3,TPM2   \#line 33<br />
                                            --geneset-annot cura                         \ #line 34<br />
                                            --ppi-annot string                           \ #line 35<br />
                                            --ppi-depth 1                                \  #line 36<br />
                                            --phenotype-term Arthrogryposis,Arthrogryposis+multiplex+congenita  \  #line 37<br />
                                            --pubmed-mining \  #line 38  <br />
                                            --mouse-pheno  \  #line 39  <br />
                                            --zebrafish-pheno \  #line 40  <br />
                                            --ddd-annot  \  #line 41  
                                        </p>
                                        <p>#VII. Prediction at variants <br />
                                            --db-score dbnsfp \ #line 42 <br />
<<<<<<< HEAD
                                            --mendel-causing-predict best \ #line 43 <br />
=======
                                            --mendel-causing-predict all \ #line 43 <br />
>>>>>>> origin/master
                                            --filter-nondisease-variant \ #line 44
                                        </p>

                                        <p>#VIII. Prediction at genes <br />
                                            --patho-gene-predict  \  #line 45<br />
                                            --phenolyzer-prediction \#   line 46 the '--phenotype-term' option has to be specified prior to this option.
                                        </p>

                                    </h4>

                                    <p><b>Part (I): Specify general environmental setting</b></p>
                                    <p> Argument &lsquo;line 1~4&rsquo; is used to set general KGGSeq running enviroment, including human genome build, resource path and program update </p>
                                    <p><b>Part (II): Specify the input files </b></p>
                                    <p><b> </b>Arguments &lsquo;line 5~6&rsquo; are used to specify the  input files which support various data format, and they are compulsory for running KGGSeq. </p>
                                    <p><b>Part (III): Output setting</b></p>
                                    <p> Argument &lsquo;line 7~11&rsquo; is used to set the output file name and type, which can be produced simutaneously by KGGSeq.</p>
                                    <p><b>Part (IV): Specify Quality control conditions</b></p>
                                    <p> Arguments &lsquo;line 12~20&rsquo; are used to apply QC conditions (Genotype or Variant level) on input dataset, and then exclude low quality genotypes or variants in the following analysis. </p>
                                    <p><b>Part (V): Filtering</b></p>
                                    <p> Arguments &lsquo;line 21~30&rsquo; are used to apply a combination of filters (genetic inheritance, gene feature, allele frequency, functional importance, etc) to enable better prioritization of the key variants/genes. </p>
                                    <p><b>Part (VI): Annotation</b></p>
                                    <p> Arguments &lsquo;line 31~41&rsquo; are used to annotate remained variants/genes by searching various genomic annotation database and literature evidence.</p>
                                    <p><b>Part (VII) & Part (VIII): Pathogenic prediction at variants and genes</b></p>
                                    <p> Arguments &lsquo;line 42~46&rsquo; are used to predict pathogenicity at remained variants/genes.</p>

                                    <p><strong>Notes</strong> Most of the above arguments are optional  (except line 5 and 6), so user can mask some lines by &ldquo;#&rdquo; or delete the  lines. Under this circumstance, user can have a systematic view of the impact  for each level or even steps. And it will be easier to produce this parameter  file by KGGSeq command generator we provide. </em></p>
                                    <p></p>
                                    <!--    
                                        <p>After running KGGSeq, you will get three file named &lsquo;<b>test0808_1.fit.xls</b>&rsquo;, &lsquo;<b>test0808_1.fit.vcf</b>&rsquo;, and &lsquo;<b>test0808_1.fit.seattleseq</b>&rsquo; in the same directory of KGGSeq.jar. For this example case, 2 missense variants residing in CTBP2 gene are retained and all of them are predicted to be disease  causal.
                                            We also provide another 5 quick examples (corresponding to different input datasets) in the same folder. Users are recommended to practice each of them for a better understanding of each parameter and their combinations.</p>
                                    -->
                                </td>
                            </tr>
                            <tr>
                                <td><hr /><h1> Options  </h1>
                                    <h2><a name="EnvironmentalSettings" >Performance  &amp;environmental settings</a></h2>

                                    <img src="imgs/leaf.gif" width="10" height="8" />  Set the number of parallel tasks for an analysis job.
                                    <code>java -jar kggseq.jar --nt #</code>
                                    <p>The # denotes the number of parallel threads to perform the analysis.<br /> 
                                    </p>
                                    <img src="imgs/leaf.gif" width="10" height="8" />  Set the version of reference human genome build<code>java -jar kggseq.jar --buildver hg##</code> 
                                    <p>All inputs  will be analyzed and outputted according to the set genome version.<br />
                                        <br />

                                        <strong>NOTE </strong>
                                        &nbsp;Currently, KGGSeq only supports hg18, hg19, and hg38. It is <tt>--buildver hg19</tt> by default.
                                    </p>
                                    <img src="imgs/leaf.gif" width="10" height="8"/>  Specify the local path to store the resource datasets <code>java -jar kggseq.jar --resource path/to/resource/files</code>
                                    <p>Once initiated, KGGSeq will automatically download the resource data from its web server to the local folder. By default, it is --resource "./"<br />
                                    </p>
                                    <img src="imgs/leaf.gif" width="10" height="8" /> Switch off the auto-checking function for itself.
                                    <code>java -jar kggseq.jar --no-lib-check</code>
                                    <p>By default,  KGGSeq will automatically check the latest version of the kggseq.jar.<br />
                                    </p>
                                    <img src="imgs/leaf.gif" width="10" height="8" />   Switch on the auto-checking and-updating function for itself.
                                    <code>java -jar kggseq.jar --lib-update</code>
                                    <p>With this option, KGGSeq will automatically check the latest version of the kggseq.jar and update it if necessary.<br />
                                    </p>

                                    <img src="imgs/leaf.gif" width="10" height="8" /> Switch on auto-checking and-updating for NEEDED resource datasets.
                                    <code>java -jar kggseq.jar --resource-update</code>
                                    <p>With this option, KGGSeq will automatically check and download the available latest version of resource datasets, once initiated.<br />
                                    </p>

                                    <img src="imgs/leaf.gif" width="10" height="8" /> Switch off the auto-log function to write log in a text file.
                                    <code>java -jar kggseq.jar --no-log</code>
                                    <p>By default, KGGSeq will automatically write log in a text file.<br />
                                    </p>

                                </td>
                            </tr>
                            <tr>
                                <td >
                                    <h2><a name="InputFormats" >Input formats </a></h2><h3><a name="VariantsAndGenotypes" >
                                            Variants and genotypes
                                        </a> </h3><br />
<<<<<<< HEAD
                                    <img src="imgs/leaf.gif" width="10" height="8" />  <a href="http://www.1000genomes.org/node/101" target="_blank">VCF format with genotypes</a><br />
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2</code> 
                                    The format that KGGSeq supporting best is the <a href="http://www.1000genomes.org/node/101" target="_blank">VCF</a> format, which stores sequence variants and called genotype information. The VCF file can be flatted as text or compressed as .gz.
=======
                                    <a href="http://www.1000genomes.org/node/101" target="_blank">VCF format with genotypes</a><br />
                                    <img src="imgs/leaf.gif" width="10" height="8" />  The format that KGGSeq supporting best is the <a href="http://www.1000genomes.org/node/101" target="_blank">VCF</a> format, which stores sequence variants and called genotype information. The VCF file can be flatted as text or compressed as .gz.
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2</code> 
>>>>>>> origin/master
                                    <p>
                                        <strong>NOTE </strong>
                                        &nbsp;To achieve the best of parallel computing advantages of KGGSeq, it is strongly recommended to compress the input data with blocked GNU Zip format by <tt> bgzip &#60;path/to/your/file&#62;</tt>. See more  <a href="http://www.htslib.org/doc/tabix.html" target="_blank">here</a>. Data compressed with conventional GNU Zip format cannot be handled in parallel. 
                                        <br />
                                        <br />
                                        <strong>NOTE </strong>
                                        &nbsp;The parallel computing also supports VCF data compressed by Burrows&#8211;Wheeler algorithm ( <a href="http://www.bzip.org/" target="_blank">bzip2</a>), which may save over 20% space than the standard GNU Zip. However,KGGSeq may take 3 to 4 folds longer time to decompress the data. 
                                        <br />                                       
                                        <br />
                                        <strong>NOTE</strong>
                                        &nbsp;To set KGGseq handling input file in parallel, the parameter <tt>--nt N</tt> should be added. Not only the .gz and .bz2 formats are supported, but the original VCF files can also be read parallel.
                                        <br />
                                        <br />  
                                        <strong>NOTE </strong>
                                        &nbsp;If the data were stored in different files, chromosome by chromosome, you can use _CHROM_ to denote the chromosome names [1...Y] or directly specify [1,2,X]in the file name so that KGGSeq can read data in multiple files in a run.In this scenario, the separated VCF files must contain the SAME subjects in the SAME order.
                                        <br />
                                        <br />
                                        <strong>NOTE </strong>
                                        &nbsp;In the VCF file, each subject should have a unique SubjectID, which is used to link the <a href="#PhenotypesAndPedigree" target="_self">phenotypes and/or pedigree structure</a>.
                                    </p>
                                    <p>
<<<<<<< HEAD
                                        <img src="imgs/leaf.gif" width="10" height="8" /> <a href="http://www.1000genomes.org/node/101" target="_blank">VCF format WITHOUT genotypes</a><br />
                                        KGGSeq can also accept VCF data without genotype information. But there are less usable functions for this type data.<br />
                                        <code>java -jar kggseq.jar --no-gty-vcf-file path/to/file1 </code>
                                    </p>   
                                    <p>
                                        <img src="imgs/leaf.gif" width="10" height="8" />   VCF format with a mixture of uphased and phased genotypes
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --force-gty-unphased </code> 
                                        KGGSeq will automatically recognize whether the genotypes are phased or not in an input VCF file. However, if there are both uphased and phased genotypes in a single VCF file, KGGSeq may throw errors. Uses have to use the option to convert all phased genotypes in to unphased genoypes.
                                    </p> 

                                    <p>
                                        <img src="imgs/leaf.gif" width="10" height="8" /><a href="http://annovar.openbioinformatics.org/en/latest/user-guide/input/#annovar-input-file" target="_blank">ANNOVAR format</a>
=======
                                        <a href="http://www.1000genomes.org/node/101" target="_blank">VCF format WITHOUT genotypes</a><br />
                                        KGGSeq can also accept VCF data without genotype information. But there are less usable functions for this type data.<br />
                                        <code>java -jar kggseq.jar --no-gty-vcf-file path/to/file1 </code>
                                    </p>                                  

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> In addition, KGGSeq also support other popular variants and/or genotypes formats which are popular in next generation studies.</p>

                                    <p>
                                        <a href="http://www.openbioinformatics.org/annovar/annovar_input.html" target="_blank">ANNOVAR format</a>
>>>>>>> origin/master
                                        <code>java -jar kggseq.jar --annovar-file path/to/file</code>
                                        <strong>NOTE </strong>
                                        Since v0.4, KGGSeq recognize an extended ANNOVAR format in which a head row and multiple columns for comments are allowed. <br />
                                        <h4>
                                            Example: <br />
                                            chr	startpos	endpos	ref		alt	comment1 comment2 comment3<br />
                                            1	69428	69428	T	G	T	92	129<br />
                                            1	69476	69476	T	C	T	1	0<br />
                                        </h4>
                                    </p>
<<<<<<< HEAD
                                    <p> <img src="imgs/leaf.gif" width="10" height="8" /> <a href="#kedformat">KGGSeq binary genotype file set</a> <tt>  --ked-file</tt></p>
=======
                                    <p><a href="#kedformat">KGGSeq binary genotype file set</a> <tt>  --ked-file</tt></p>
>>>>>>> origin/master
                                    <code>java -jar kggseq.jar --ked-file path/to/fileset</code>

                                    The KGGSeq binary genotype produced by KGGSeq, which is more flexible to store the different types of genotypes than the <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/binary.shtml" target="_blank">PLINK BED</a> format.

                                    <h3><a name="PhenotypesAndPedigree" >Phenotypes and pedigree</a></h3>
                                    <br />

                                    When there are <b>genotypes</b> inputted in VCF format, KGGSeq should know to whom they belong and the phenotypes of the samples. So you need provide your sample information in this case.<br /><br />
                                    <img src="imgs/leaf.gif" width="10" height="8" /> Concise subject ID and disease status: <tt>--indiv-pheno </tt>
                                    <code>java -jar kggseq.jar --vcf-file path/to/file  --indiv-pheno SubjectID1:0,SubjectID2:2,...</code>
                                    Specify the individual IDs and affection status, which is coded as: 0 missing or unknow,1 unaffected,2 affected.
                                    <br />
                                    <br />

                                    <strong>Note</strong> &nbsp; The individual IDs <b>must</b> be unique and identical to the ones defined in the VCF file(s).                                    </p>
                                    <img src="imgs/leaf.gif" width="10" height="8" /> Conventional <a href="http://www.broadinstitute.org/science/programs/medical-and-population-genetics/haploview/input-file-formats-0" target="_blank">linkage pedigree file</a>:<tt> --ped-file </tt>
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 (or --composite-subject-id)</code>
                                    Specify the subject IDs,  relationship between subjects, and disease status of them; it is a conventional Linkage Format, detailed <a href="http://www.broadinstitute.org/science/programs/medical-and-population-genetics/haploview/input-file-formats-0" target="_blank">here</a>.<br />
                                    <h4>
                                        <ul>
                                            <li> <b>Pedigree Name</b><br/>
                                                A unique alphanumeric identifier for this individual's family. Unrelated individuals should not share a pedigree name.
                                            </li>
                                            <li> <b>Individual ID </b><br/>
                                                An alphanumeric identifier for this individual. Should be unique within his family (see above).
                                            </li>
                                            <li> <b>Father's ID </b><br/>
                                                Identifier corresponding to father's individual ID or "0" if unknown father. Note
                                                that if a father ID is specified, the father must also appear in the file.
                                            </li>
                                            <li> <b>Mother's ID</b><br/>
                                                Identifier corresponding to mother's individual ID or "0" if unknown mother Note that if a mother ID is specified, the mother must also appear in the file.
                                            </li>
                                            <li><b> Sex </b><br/>
                                                Individual's gender (1=MALE, 2=FEMALE).
                                            </li>
                                            <li> <b>Affection status</b> <br/>
                                                Affection status to be used for association tests (0=UNKNOWN, 1=UNAFFECTED,2=AFFECTED).
                                            </li>
                                        </ul>
                                        <strong>Note</strong> By default, the subject IDs in the VCF file must be identical to Individual IDs in the pedigree file, which are assumed unique in the whole pedigree file. However, one can also ask KGGSeq to use a composite subject ID in the CVF file as "PedigreeName$IndividualID" by the option <tt>--composite-subject-id</tt>.
                                    </h4>
                                    <br/>
                                    <br/>
                                    Besides, KGGSeq also accepts a pedigree format with header information (See the following example),in which the phenotype names are specified. Some of the phenotype names may be designated as covariates in statistical association analysis.

                                    <h4>
                                        <pre>
                                        fid iid fatid matid sex DISEASE QT  AGE  
                                        1  NA12344 NA12347 NA12348 1   1   94.17   NA   
                                        1  NA12347 0   0   1   NA   19.54  40.0  
                                        1  NA12348 0   0   2   2   NA  36.6 
                                        </pre>
                                        <strong>Note</strong> The head line must start with 'fid iid'. Otherwise, KGGSeq will regard the first row as a subject-line. <br/>
                                        <strong>Note</strong> The missing phenotype values are denoted by "NA".
                                    </h4>
                                </td>
                            </tr>


                            <tr>
                                <td>                                    
                                    <h2><a name="Outputs" >Outputs </a></h2>
                                    <p>KGGSeq can flexibly output different formats of the prioritization and annotation results for validation or further analysis by third-party tools.</p>
                                    <p>Output file path and file name prefix: <tt>--out</tt><code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname</code>
                                        Specify path and prefix name of outputs. It is &quot;.<b>/kggseq</b>&quot;  by default. </p>
                                    <strong>Note</strong> Note all outputs (except for the Excel format) of KGGSeq are to compressed by  GNU Zip format automatically by default. Please use <tt>--no-gz</tt> to disable the automatic compression function.

                                    <h3><a name="OdinaryOutputs" >  Ordinary outputs
                                        </a> </h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Text format:<tt> This is by default</tt><code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 </code>
                                        By default, KGGSeq output results in TEXT format in a file named kggseq.flt.txt, in which the fields (or columns) are delimited by the tabs.</p>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Excel format: <tt>--excel</tt>
                                        <code>java -jar kggseq.jar --vcf-file path//to/file1 --ped-file path/to/file2  --excel</code>
                                        The results will be stored in a Excel file with the name kggseq.flt.xls.</p>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> For <a href="http://www.openbioinformatics.org/annovar/annovar_input.html" target="_blank">ANNOVAR</a> input: <tt>--o-annovar</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2  --o-annovar</code>
                                        Generate an extra copy of output for the prioritized variants in ANNOVAR input format, which can be further annotated by ANNOVAR.</p>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> For <a href="http://www.1000genomes.org/node/101" target="_blank">VCF</a> input: <tt>--o-vcf</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2  --o-vcf</code>
                                        Generate an extra copy of output for the prioritized variants in VCF format, which can be further analyzed by other tools.</p>
                                    <strong>Note</strong>  Please use <tt>--missing-gty X</tt> to denote missing genotypes in the output VCF file. It is <tt>--missing-gty ./.</tt> by default.
<<<<<<< HEAD

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> For <a href="http://www.1000genomes.org/node/101" target="_blank">VCF</a> input, but in a simplified format: <tt>--o-svcf</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2  --o-svcf</code>
                                        Generate an extra copy of output for the prioritized variants in simplified VCF format, in which only genotypes are included in the FORMAT column.</p>
=======
                                    
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> For <a href="http://www.1000genomes.org/node/101" target="_blank">VCF</a> input, but in a simplified format: <tt>--o-svcf</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2  --o-svcf</code>
                                        Generate an extra copy of output for the prioritized variants in simplified VCF format, in which only genotypes are kept in the FORMAT column.</p>
>>>>>>> origin/master
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> The group-file format for <a href="http://genome.sph.umich.edu/wiki/EPACTS#Creating_marker_group_file" target="_blank"> EPACTS </a> input or other tools : <tt>--o-gene-grp</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2  --o-gene-grp</code>
                                        Output coordinates, reference and alternative alleles of left variants of each gene in a group-file format, which can be further analyzed by other tools for more sophisticated gene-based association analysis.</p>
                                    <strong>Note</strong> According to coordinates, a variant may be mapped onto multiple genes.

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> For <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/" target="_blank">PLINK</a> input: <tt>--o-plink-ped</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2  --o-plink-ped</code>
                                        Generate an extra copy of output for the prioritized variants in PLINK <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml#ped" target="_blank">PED</a> and <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml#map" target="_blank">MAP</a> file, which can be further annotated by PLINK.</p>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Extract the flanking sequence of each sequence variant: <tt>--o-flanking-seq</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2  --o-flanking-seq 50</code>
                                        The source sequence information is the UCSC Chromosome dataset <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg18/chromosomes/" target="_blank">hg18</a> or <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/" target="_blank">hg19</a>, which are human genome reference sequences on the forward strand. The unit of the specified flanking sequence is base-pair(bp). The retrieved sequences will be in the *.flt.txt or *.flt.xls with the column label &quot;FlankingSeqXXbp&quot;.</p>
                                    <strong>Note</strong> Users might need set large memory (by <tt>-Xmx20g</tt>) to carry out this function as KGGSeq will load the whole sequences of a chromosome.

                                    <h3><a name="BinaryOutputs" >
                                            Binary outputs
                                        </a> </h3>
                                    <p>PLINK <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/binary.shtml" target="_blank"> binary genotypes files set</a>: <tt>--o-plink-bed</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --composite-subject-id --o-plink-bed</code>
                                        Generate an extra copy of output for the prioritized variants  in PLink binary genotype format (SNP-major model), which can store the genotype information more efficiently.<br />
                                        <strong>Note</strong> This format does not support variants with over 2 alleles and phased genotypes, which will be ignored by KGGSeq.To record this information, you are suggested to use the KGGSeq binary genotype format (below).

                                        <p><a name="kedformat">KGGSeq binary genotypes files set</a>: <tt>--o-ked</tt>
                                            <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --composite-subject-id --o-ked</code>
                                            Generate an extra copy of output for the prioritized variants in KGGSeq binary genotype format, which use much less  space to store the genotype information.<br />

                                            <h4>
                                                Here is an example VCF file:
                                                <table  class="list">
                                                    <tr>
                                                        <td>##fileformat=VCFv4.0<br />
                                                            #CHROM POS ID REF ALT QUAL FILTER INFO FORMAT 1$A 1$B 1$C<br />
                                                            chr1 4793 rs6682385 A G 596.57 PASS AC=2 GT 1/1 0/0 ./.<br />
                                                            chr1 53560 . T C 573.51 PASS AC=1; GT ./. 0/0 0/1<br />
                                                            chr1 12887054 rs3738105 T C,G 5572.99 PASS AC=2,1 GT 1/2 0/0 0/1</td>
                                                    </tr>
                                                </table>
                                                And the corresponding pedigree file:
                                                <table  class="list">
                                                    <tr>
                                                        <td >1 A 0 0 1 1<br />
                                                            1 B 0 0 2 1<br />
                                                            1 C A B 1 2</td>
                                                    </tr>
                                                </table>
                                                After running the command above, KGGSeq will generat three file:					           <br />
                                                <b>kggseq.ked</b> (the KGGSeq binary genotype file, genotype information)<br />
                                                <b>kggseq.kam</b> (the KGGSeq family data file, first six columns of the conventional <a href="http://www.broadinstitute.org/science/programs/medical-and-population-genetics/haploview/input-file-formats-0" target="_blank">Linkage Pedigree file</a>)<br />
<<<<<<< HEAD
                                                <b>kggseq.bim</b> (the KGGSeq variants map data file, two extra columns for reference and alternative allele names, like: A G,C). The <b>kggseq.kim</b> is similar to the  *.bim file of <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/binary.shtml" target="_blank">PLINK BED</a> format. But it has one additional column (the last column) which denotes the compression status in the <b> kggseq.ked</b> file.The '-1' means that the genotypes are not compressed while another value suggested that the genotypes are compressed for a variant. The value divided by 3 denotes the number of &#34;1&#34; bits in the encoded genotypes of a variant for all subjects. The actual positions of &#34;1&#34; bits in the encoded genotypes are stored in the <b>kggseq.ked</b> file.
                                                <br />
                                                An example:
                                                <table  class="list">
                                                    <tr>
                                                        <td > 1	rs558604819	0	10642	G	A	63<br />
                                                            1	rs575272151	0	11008	C	G	-1<br />
                                                            1	rs544419019	0	11012	C	G	-1<br />
                                                            1	rs561109771	0	11063	T	G	45<br /></td>
                                                    </tr>
                                                </table>

                                                <br />

                                                Compared to the plink *.bed format, the <b>kggseq.ked</b> has a different way to present genotypes, which is more flexible.
=======
                                                <b>kggseq.bim</b> (the KGGSeq variants map data file, two extra columns for reference and alternative allele names, like: A G,C)<br />
                                                The<b> kggseq.kam</b> and <b>kggseq.kim</b> are similar to the *.fam and *.bim files of <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/binary.shtml" target="_blank">PLINK BED</a> format.<br />
                                                <br />
                                                But the <b>kggseq.ked</b> has a different way to present genotypes, which is more flexible.
>>>>>>> origin/master
                                                Here is the binary view of  <b>kggseq.ked</b> file:<br />
                                                <table  class="list">
                                                    <tr>
                                                        <td >   |-magic &nbsp; &nbsp;&nbsp;number--| &nbsp;&nbsp;|phase mode| &nbsp;|------------------genotype data-------------------|<br />
                                                            10011110 10000010 &nbsp; 00000000&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<tt><font color="green">101</font><font color="red">001</font>00</tt> <tt><font color="green">100</font><font color="red">101</font>00</tt> <tt><font color="green">100</font><font color="red">000</font><font color="green">00</font></tt> <tt><font color="green">1</font>0000000 </tt> </td>
                                                    </tr>
                                                </table>

                                                <br />
                                                The followings are the interpretation of the KGGSeq binary genotype file format
                                                <ul>
                                                    <li> First two bytes <tt>10011110 10000010 </tt> for the validation checking of a KGGSeq binary genotype file.</li>
                                                    <li> Third byte is <tt>00000000</tt> (Unphased genotypes) or <tt>00000001</tt> (Phased genotypes) </li>
                                                    <li> Genotype data, either for unphased or phased ones in Variant-major order ((i.e. list all individuals for the first Variant, all individuals for the second Variant, etc) ).
<<<<<<< HEAD
                                                        The 4th byte (<tt><font color="green">101</font><font color="red">001</font>00</tt>) denotes genotypes of chr1:4793 by two bit-blocks (denoted by green and red respectively). At this site, the genotypes of A, B and C are first encoded as 10, 00 and 11 respectively (See details in the encoding table). The binary codes at first and second places are put into two blocks <font color="green">101</font> and <font color="red">001</font>. The block size is equal to the number of total subjects. It is 3 here. The remaining spaces in a byte, which can contain 8 bits, are set as zero bits. <br/>
                                                        The 5th byte (<tt><font color="green">100</font><font color="red">101</font>00</tt> ) denotes genotypes of chr1:53560. The 6th and 7th bytes (<tt><font color="green">100</font><font color="red">000</font><font color="green">00</font></tt> and <tt><font color="green">1</font>0000000 </tt> ) present genotypes of chr1:12887054, which has three alleles.</li>
=======
                                                        The 4th byte (<tt><font color="green">101</font><font color="red">001</font>00</tt>) denotes genotypes of chr1:4793 by two bit-blocks (marked in color). The A, B and C have bit genotypes 10,00 and 11 respectively. The 5th byte (<tt><font color="green">100</font><font color="red">101</font>00</tt> ) denotes genotypes of chr1:53560. The 6th and 7th bytes (<tt><font color="green">100</font><font color="red">000</font><font color="green">00</font></tt> and <tt><font color="green">1</font>0000000 </tt> ) present genotypes of chr1:12887054, which has three alleles.</li>
>>>>>>> origin/master
                                                    <li> New "row" in the VCF file always starts a new byte (8 bits) </li>   
                                                </ul>
                                                For each sequence variant, the number of bytes required to present a genotype is determined by the number of alleles and the genotype phase status.<br />
                                                <b>I. Unphased -genotypes</b><br />
                                                I.I Bi-allelic sequence variant (2 bits)
                                                <table  class="list">
                                                    <tr >
                                                        <td   valign="top">&nbsp;</td>                                                      
                                                        <td   valign="top">Reference homozygous</td>
                                                        <td  valign="top">Heterozygous </td>
                                                        <td   valign="top">Alternative homozygous</td>
                                                        <td   valign="top">missing</td>
                                                    </tr>
                                                    <tr >
                                                        <td  valign="top">VCF genotype</td>
                                                        <td  valign="top">0/0</td>
                                                        <td   valign="top">0/1</td>
                                                        <td  valign="top">1/1</td>
                                                        <td  valign="top">./.</td>
                                                    </tr>
                                                    <tr  >
                                                        <td width="127" valign="top">Bits</td>
                                                        <td width="130" valign="top">00</td>
                                                        <td width="127" valign="top">01</td>
                                                        <td width="127" valign="top">10</td>
                                                        <td width="127" valign="top">11</td>
                                                    </tr>
                                                    <tr >
                                                        <td width="127" valign="top">Decimal</td>
                                                        <td width="130" valign="top">0</td>
                                                        <td width="127" valign="top">1</td>
                                                        <td width="127" valign="top">2</td>
                                                        <td width="127" valign="top">3</td> 
                                                    </tr>
                                                </table>
                                                I.II Tri-allelic sequence variant (3 bits)
                                                <table  class="list">
                                                    <tr>
                                                        <td width="128" valign="top">&nbsp;</td>                                                      
                                                        <td width="115" valign="top">Reference homozygous</td>
                                                        <td width="116" valign="top">Heterozygous </td>
                                                        <td width="119" valign="top">Heterozygous</td>
                                                        <td width="95" valign="top">Alternative homozygous</td>
                                                        <td width="115" valign="top">Heterozygous</td>
                                                        <td width="121" valign="top">Alternative homozygous</td>
                                                        <td width="82" valign="top">missing</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="128" valign="top">VCF genotype</td>                                                       
                                                        <td width="115" valign="top">0/0</td>
                                                        <td width="116" valign="top">0/1</td>
                                                        <td width="119" valign="top">0/2</td>
                                                        <td width="95" valign="top">1/1</td>
                                                        <td width="115" valign="top">1/2</td>
                                                        <td width="121" valign="top">2/2</td>
                                                        <td width="82" valign="top">./.</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="128" valign="top">Bits</td>
                                                        <td width="82" valign="top">000</td>
                                                        <td width="115" valign="top">001</td>
                                                        <td width="116" valign="top">010</td>
                                                        <td width="119" valign="top">011</td>
                                                        <td width="95" valign="top">100</td>
                                                        <td width="115" valign="top">101</td>
                                                        <td width="121" valign="top">110</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="128" valign="top">Decimal</td>
                                                        <td width="82" valign="top">0</td>
                                                        <td width="115" valign="top">1</td>
                                                        <td width="116" valign="top">2</td>
                                                        <td width="119" valign="top">3</td>
                                                        <td width="95" valign="top">4</td>
                                                        <td width="115" valign="top">5</td>
                                                        <td width="121" valign="top">6</td>
                                                    </tr>
                                                </table>
                                                I.III Quad-allelic sequence variant (4 bits)
                                                <table class="list">
                                                    <tr>
                                                        <td width="129" valign="top">&nbsp;</td>                                                        
                                                        <td width="111" valign="top">Reference homozygous</td>
                                                        <td width="106" valign="top">Heterozygous </td>
                                                        <td width="115" valign="top">Heterozygous</td>
                                                        <td width="106" valign="top">Heterozygous</td>
                                                        <td width="100" valign="top">Alternative homozygous</td>
                                                        <td width="128" valign="top">Heterozygous</td>     
                                                        <td width="88" valign="top">Heterozygous</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="129" valign="top">VCF genotype</td>                                                       
                                                        <td width="111" valign="top">0/0</td>
                                                        <td width="106" valign="top">0/1</td>
                                                        <td width="115" valign="top">0/2</td>
                                                        <td width="106" valign="top">0/3</td>
                                                        <td width="100" valign="top">1/1</td>
                                                        <td width="128" valign="top">1/2</td>
                                                        <td width="88" valign="top">1/3</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="129" valign="top">Bits</td>
                                                        <td width="96" valign="top">0000</td>
                                                        <td width="111" valign="top">0001</td>
                                                        <td width="106" valign="top">0010</td>
                                                        <td width="115" valign="top">0011</td>
                                                        <td width="106" valign="top">0100</td>
                                                        <td width="100" valign="top">0101</td>
                                                        <td width="128" valign="top">0110</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="129" valign="top">Decimal</td>
                                                        <td width="96" valign="top">0</td>
                                                        <td width="111" valign="top">1</td>
                                                        <td width="106" valign="top">2</td>
                                                        <td width="115" valign="top">3</td>
                                                        <td width="106" valign="top">4</td>
                                                        <td width="100" valign="top">5</td>
                                                        <td width="128" valign="top">6</td>
                                                    </tr>
                                                </table>
                                                <table class="list">
                                                    <tr>
                                                        <td width="63" valign="top">&nbsp;</td>

                                                        <td width="88" valign="top">Alternative homozygous</td>
                                                        <td width="88" valign="top">Heterozygous</td>
                                                        <td width="88" valign="top">Alternative homozygous</td>
                                                        <td width="105" valign="top">missing</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">VCF genotype</td>

                                                        <td width="88" valign="top">2/2</td>
                                                        <td width="88" valign="top">2/3</td>
                                                        <td width="88" valign="top">3/3</td>
                                                        <td width="96" valign="top">./.</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">Bits</td>
                                                        <td width="88" valign="top">0111</td>
                                                        <td width="88" valign="top">1000</td>
                                                        <td width="88" valign="top">1001</td>
                                                        <td width="88" valign="top">1010</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">Decimal</td>
                                                        <td width="88" valign="top">7</td>
                                                        <td width="88" valign="top">8</td>
                                                        <td width="88" valign="top">9</td>
                                                        <td width="88" valign="top">10</td>
                                                    </tr>
                                                </table><br />
                                                <b>II. Phased -genotypes</b><br />
                                                II.I Bi-allelic sequence variants (3 bits)
                                                <table class="list">
                                                    <tr>
                                                        <td width="106" valign="top">&nbsp;</td>                                                      
                                                        <td width="112" valign="top">Reference homozygous</td>
                                                        <td width="114" valign="top">Heterozygous </td>
                                                        <td width="89" valign="top">Heterozygous </td>
                                                        <td width="112" valign="top">Alternative homozygous</td>
                                                        <td width="105" valign="top">missing</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="106" valign="top">VCF genotype</td>                                                      
                                                        <td width="112" valign="top">0|0</td>
                                                        <td width="114" valign="top">0|1</td>
                                                        <td width="89" valign="top">1|0</td>
                                                        <td width="112" valign="top">1|1</td>
                                                        <td width="105" valign="top">.|.</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="106" valign="top">Bits</td>
                                                        <td width="105" valign="top">000</td>
                                                        <td width="112" valign="top">001</td>
                                                        <td width="114" valign="top">010</td>
                                                        <td width="89" valign="top">011</td>
                                                        <td width="112" valign="top">100</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="106" valign="top">Decimal</td>
                                                        <td width="105" valign="top">0</td>
                                                        <td width="112" valign="top">1</td>
                                                        <td width="114" valign="top">2</td>
                                                        <td width="89" valign="top">3</td>
                                                        <td width="112" valign="top">4</td>
                                                    </tr>
                                                </table>
                                                II.II          Tri-allelic sequence variant (4 bits)
                                                <table class="list">
                                                    <tr>
                                                        <td width="133" valign="top">&nbsp;</td>                                                       
                                                        <td width="107" valign="top">Reference homozygous</td>
                                                        <td width="121" valign="top">Heterozygous </td>
                                                        <td width="100" valign="top">Heterozygous</td>
                                                        <td width="116" valign="top">Heterozygous</td>
                                                        <td width="102" valign="top">Heterozygous</td>
                                                        <td width="119" valign="top">Alternative homozygous</td>
                                                        <td width="56" valign="top">Heterozygous</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="133" valign="top">VCF genotype</td>

                                                        <td width="107" valign="top">0|0</td>
                                                        <td width="121" valign="top">0|1</td>
                                                        <td width="100" valign="top">0|2</td>
                                                        <td width="116" valign="top">1|0</td>
                                                        <td width="102" valign="top">1|1</td>
                                                        <td width="119" valign="top">1|2</td>
                                                        <td width="56" valign="top">2|0</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="133" valign="top">Bits</td>
                                                        <td width="93" valign="top">0000</td>
                                                        <td width="107" valign="top">0001</td>
                                                        <td width="121" valign="top">0010</td>
                                                        <td width="100" valign="top">0011</td>
                                                        <td width="116" valign="top">0100</td>
                                                        <td width="102" valign="top">0101</td>
                                                        <td width="119" valign="top">0110</td>

                                                    </tr>
                                                    <tr>
                                                        <td width="133" valign="top">Decimal</td>
                                                        <td width="93" valign="top">0</td>
                                                        <td width="107" valign="top">1</td>
                                                        <td width="121" valign="top">2</td>
                                                        <td width="100" valign="top">3</td>
                                                        <td width="116" valign="top">4</td>
                                                        <td width="102" valign="top">5</td>
                                                        <td width="119" valign="top">6</td>
                                                    </tr>
                                                </table>
                                                <table class="list">
                                                    <tr>
                                                        <td width="63" valign="top">&nbsp;</td>

                                                        <td width="83" valign="top">Heterozygous</td>
                                                        <td width="88" valign="top">Alternative homozygous</td>
                                                        <td width="93" valign="top">missing</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">VCF genotype</td>

                                                        <td width="83" valign="top">2|1</td>
                                                        <td width="88" valign="top">2|2</td>
                                                        <td width="93" valign="top">.|.</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">Bits</td>
                                                        <td width="56" valign="top">0111</td>
                                                        <td width="83" valign="top">1000</td>
                                                        <td width="88" valign="top">1001</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">Decimal</td>
                                                        <td width="56" valign="top">7</td>
                                                        <td width="83" valign="top">8</td>
                                                        <td width="88" valign="top">9</td>
                                                    </tr>
                                                </table>
                                                <br />
                                                II.III Quad-allelic sequence variants (5 bits)
                                                <table class="list">
                                                    <tr>
                                                        <td width="130" valign="top">&nbsp;</td>                                                    
                                                        <td width="116" valign="top">Reference homozygous</td>
                                                        <td width="104" valign="top">Heterozygous </td>
                                                        <td width="100" valign="top">Heterozygous</td>
                                                        <td width="111" valign="top">Heterozygous</td>
                                                        <td width="102" valign="top">Heterozygous</td>
                                                        <td width="125" valign="top">Heterozygous</td>
                                                        <td width="119" valign="top">Heterozygous</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="130" valign="top">VCF genotype</td>

                                                        <td width="116" valign="top">0|0</td>
                                                        <td width="104" valign="top">0|1</td>
                                                        <td width="100" valign="top">0|2</td>
                                                        <td width="111" valign="top">0|3</td>
                                                        <td width="102" valign="top">1|0</td>
                                                        <td width="125" valign="top">1|1</td>
                                                        <td width="119" valign="top">1|2</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="130" valign="top">Bits</td>
                                                        <td width="103" valign="top">00000</td>
                                                        <td width="116" valign="top">00001</td>
                                                        <td width="104" valign="top">00010</td>
                                                        <td width="100" valign="top">00011</td>
                                                        <td width="111" valign="top">00100</td>
                                                        <td width="102" valign="top">00101</td>
                                                        <td width="125" valign="top">00110</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="130" valign="top">Decimal</td>
                                                        <td width="103" valign="top">0</td>
                                                        <td width="116" valign="top">1</td>
                                                        <td width="104" valign="top">2</td>
                                                        <td width="100" valign="top">3</td>
                                                        <td width="111" valign="top">4</td>
                                                        <td width="102" valign="top">5</td>
                                                        <td width="125" valign="top">6</td>
                                                    </tr>
                                                </table>
                                                <table class="list">
                                                    <tr>
                                                        <td width="108" valign="top">&nbsp;</td>                                                     
                                                        <td width="125" valign="top">Alternative homozygous</td>
                                                        <td width="135" valign="top">Heterozygous</td>
                                                        <td width="135" valign="top">Heterozygous</td>
                                                        <td width="135" valign="top">Heterozygous</td>
                                                        <td width="138" valign="top">Heterozygous</td>
                                                        <td width="88" valign="top">Alternative homozygous</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="108" valign="top">VCF genotype</td>

                                                        <td width="125" valign="top">1|3</td>
                                                        <td width="135" valign="top">2|0</td>
                                                        <td width="135" valign="top">2|1</td>
                                                        <td width="135" valign="top">2|2</td>
                                                        <td width="138" valign="top">2|3</td>
                                                        <td width="88" valign="top">3|0</td>

                                                    </tr>
                                                    <tr>
                                                        <td width="108" valign="top">Bits</td>
                                                        <td width="119" valign="top">00111</td>
                                                        <td width="125" valign="top">01000</td>
                                                        <td width="135" valign="top">01001</td>
                                                        <td width="135" valign="top">01010</td>
                                                        <td width="135" valign="top">01011</td>
                                                        <td width="138" valign="top">01100</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="108" valign="top">Decimal</td>
                                                        <td width="119" valign="top">7</td>
                                                        <td width="125" valign="top">8</td>
                                                        <td width="135" valign="top">9</td>
                                                        <td width="135" valign="top">10</td>
                                                        <td width="135" valign="top">11</td>
                                                        <td width="138" valign="top">12</td>
                                                    </tr>
                                                </table>
                                                <table class="list">
                                                    <tr>
                                                        <td width="63" valign="top">&nbsp;</td>

                                                        <td width="56" valign="top">Heterozygous</td>
                                                        <td width="83" valign="top">Heterozygous</td>
                                                        <td width="83" valign="top">Alternative homozygous</td>
                                                        <td width="103" valign="top">missing</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">VCF genotype</td>

                                                        <td width="56" valign="top">3|1</td>
                                                        <td width="83" valign="top">3|2</td>
                                                        <td width="83" valign="top">3|3</td>
                                                        <td width="103" valign="top">.|.</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">Bits</td>
                                                        <td width="88" valign="top">01101</td>
                                                        <td width="56" valign="top">01110</td>
                                                        <td width="83" valign="top">01111</td>
                                                        <td width="83" valign="top">10000</td>
                                                    </tr>
                                                    <tr>
                                                        <td width="63" valign="top">Decimal</td>
                                                        <td width="88" valign="top">13</td>
                                                        <td width="56" valign="top">14</td>
                                                        <td width="83" valign="top">15</td>
                                                        <td width="83" valign="top">16</td>
                                                    </tr>
                                                </table>
                                            </h4>
                                        </p>
                                </td>
                            </tr>
                            <tr>
                                <td>
                                    <h2><a name="QualityControl" >Quality Control </a></h2>
                                    <p>After genetic variants were inputed from a variety of data formats, KGGSeq provides a few optional quality control filters on individual genotype, genomic variant and sample phenotype.
                                        By adopting one or a combination of these QCs before performing downstream prioritization/annotation, it could help the user to reduce false positives.</p>
                                    <h3><a name="GenotypeQC" >  Genotype QC </a> </h3><br />
                                    This is for genotype level quality control. Typical information in each individual genotype field covers read depth(<tt>--gty-dp</tt>), genotype quality (<tt>--gty-qual</tt>) and allelic balance ratio (<tt>--gty-af-ref</tt>, <tt>--gty-af-het</tt> and <tt>--gty-af-alt</tt>):
                                    <p><code>java -jar kggseq.jar --vcf-file path/to/file1 --indiv-pheno Indiv:1,Indiv:2 --gty-qual 20 --gty-dp 4 --gty-sec-pl 20 --gty-af-ref 0.05 --gty-af-het 0.25 --gty-af-alt 0.75 --o-vcf path/to/file2</code></p>
                                    <p> They are in relationship of logical AND, and not exclusive of each other,user can specify one or any combination of these parameters at one time.</p>
                                    <table class="list">
                                        <tr>
                                            <td>--gty-qual 20</td>
                                            <td width="83%">Exclude genotypes with the minimal genotyping quality (<a href="http://en.wikipedia.org/wiki/Phred_quality_score" target="_blank">Phred Quality Score</a>) per genotype &lt; 10. The default setting is <tt>--gty-qual 20</tt>.</td>
                                        </tr>
                                        <tr>
                                            <td>--gty-dp 4</td>
                                            <td>Exclude genotypes with the minimal read depth per genotype &lt; 4. The default setting is <tt>--gty-dp 4</tt>.</td>
                                        </tr>
                                        <tr>
                                            <td>--gty-sec-pl 20</td>
                                            <td>Exclude genotypes with the second smallest normalized, Phred-scaled likelihoods for genotypes &lt; 20. (otherwise, there would be confusing genotypes).The default setting is <tt>--gty-sec-pl 20</tt>.</td>
                                        </tr>
                                        <tr>
                                            <td>--gty-af-ref 0.05</td>
                                            <td>
<<<<<<< HEAD
                                                Exclude genotypes with the fraction of the reads carrying alternative allele &gt;= 5% at a reference-allele homozygous genotype, usually indicating by AF in a VCF file. The default setting is <tt>--gty-af-ref 1 </tt>, which disables this tag.</td>
=======
                                                Exclude genotypes with the fraction of the reads carrying alternative allele &gt;= 5% at a reference-allele homozygous genotype, usually indicating by AF in a VCF file. The default setting is <tt>--gty-af-ref 0.05 </tt>. This tag can be disabled by <tt>--gty-af-ref 1 </tt></td>
>>>>>>> origin/master
                                        </tr>
                                        <tr>
                                            <td>--gty-af-het 0.25</td>
                                            <td>Exclude genotypes with the fraction of the reads carrying alternative allele &lt;= 25% at a heterozygous genotype.
<<<<<<< HEAD
                                                The default setting is <tt>--gty-af-het 0</tt>, which disables this tag.</td>
=======
                                                The default setting is <tt>--gty-af-het 0.25</tt>. This tag can be disabled by <tt>--gty-af-het 0 </tt></td>
>>>>>>> origin/master
                                        </tr>
                                        <tr>
                                            <td>--gty-af-alt 0.75</td>
                                            <td>Exclude variants with the fraction of the reads carrying alternative allele &lt;= 75% at a alternative-allele homozygous genotype.
<<<<<<< HEAD
                                                The default setting is <tt>--gty-af-alt 0</tt>, which disables this tag.</td>
=======
                                                The default setting is <tt>--gty-af-alt 0.75</tt>. This tag can be disabled by <tt>--gty-af-alt 0 </tt></td>
>>>>>>> origin/master
                                        </tr>
                                        <tr>
                                            <td>--gty-somat-p 0.05</td>
                                            <td> Exclude genotypes at which the distribution of reads carrying reference and alternative alleles are not significantly (p&gt;0.05) different between tumor tissues and non-tumor tissues. By default, the p value threshold is 0.05.</td>
                                        </tr>
                                    </table>

                                    <p><strong>Note</strong> The AF is prioritized in the quality control process if the both AF and Ad are provided in the VCF file as the AF is usually estimated using high quality reads</p>
                                    <p>After genotype QC, two output files will be produced. 1)The default KGGSeq output in txt format (or Excel by specifying "--excel") which includes the basic annotations for each clean variant;
                                        2)The vcf file: the genotype field which does not meet the QC conditions will be replaced by missing value ./., hence not considered in following analysis.</p>
                                    <h4>
                                        ##Key information provided in default basic annotation file:<br />
                                        "AffectedRefHomGtyNum", Number of affected individuals with reference homozygote at this variant;<br />
                                        "AffectedHetGtyNum", Number of affected individuals with heterozygote at this variant;<br />
                                        "AffectedAltHomGtyNum", Number of affected individuals with non-ref homozygote;<br />
                                        "UnaffectedRefHomGtyNum", Number of unaffected individuals with reference homozygote at this variant;<br /> Gene feature filtering
                                        "UnaffectedHetGtyNum", Number of unaffected individuals with heterozygote at this variant;<br />
                                        "UnaffectedAltHomGtyNum", Number of unaffected individuals with non-ref homozygote;
                                    </h4>
                                    <p><strong>Note</strong> All above QC functions are designed for VCF input format,which may not be suitable for other <a href="#InputFormats" target="_self">input formats</a>.</p>
                                    <p><strong>Note</strong> All above QC functions can be turned off by  <tt>--no-qc</tt>.</p>
                                    <p><strong>Note</strong> In the LOG file, the null genotype refers to either 1) a missing genotype in the input VCF file or 2) a genotype failing QC. Similarly, the non-null genotype refers to a genotype pass QC.</p>
                                    <h3><a name="VariantQC" >  Variant QC </a> </h3><br />
                                    This is for variant level quality control. In general VCF format, each line corresponds to the variants on the same position. Typical information for one variant includes overall sequencing quality (<tt>--seq-qual</tt>), RMS Mapping Quality (<tt>--seq-mq</tt>),
                                    and strand bias (<tt>--seq-sb</tt> or Phred-scaled p-value using Fisher's exact test <tt>--seq-fs</tt>). To make use of this information:
                                    <p><code>java -jar kggseq.jar --vcf-file path/to/file1 --indiv-pheno Indiv:1,Indiv:2 --vcf-filter-in PASS,VQSRTrancheSNP95.00to97.00,VQSRTrancheSNP97.00to99.00 --seq-qual 30 --seq-mq 20  --seq-fs 60 --o-vcf path/to/file2</code></p>

                                    Interpretation of key parameters and their values:<br />
                                    <table  class="list">

                                        <tr>
                                            <td valign="top"><b>Arguments/values</b></td>
                                            <td valign="top"><b>Annotation</b></td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--vcf-filter-in</td>
                                            <td valign="top">exlude the variants with "FILTER" not matching the specified QC labels in the VCF input file. e.g., --vcf-filter-in PASS,VQSRTrancheSNP90.00to93.00,VQSRTrancheSNP93.00to95.00,VQSRTrancheSNP95.00to97.00,VQSRTrancheSNP97.00to99.00 . By default, this tag is ignored.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--seq-qual 30</td>
                                            <td valign="top">set the minimum overall sequencing quality score (<a href="http://en.wikipedia.org/wiki/Phred_quality_score" target="_blank">Phred Quality Score</a>) for the variant at 30;(default: --seq-qual 30) </td>
                                        </tr>

                                        <tr>
                                            <td valign="top">--seq-mq 20</td>
                                            <td valign="top">set the minimum overall mapping quality score (<a href="http://genome.sph.umich.edu/wiki/Mapping_Quality_Scores" target="_blank">Mapping Quality Score</a>) for the variant at 20;(default: --seq-mq 20) </td>
                                        </tr>
                                        <!--
                                        <tr>
                                            <td valign="top">--seq-sb -10</td>
                                            <td valign="top">set the maximal overall strand bias score for the variant at -10;(default: this tag is ignored.) </td>
                                        </tr>
<<<<<<< HEAD
                                        -->
=======
                                        !-->
>>>>>>> origin/master
                                        <tr>
                                            <td valign="top">--seq-fs 60</td>
                                            <td valign="top">set the maximal overall strand bias Phred-scaled p-value (using Fisher's exact test) for the variant at 60;(default: this tag is ignored.) </td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--max-allele 4</td>
                                            <td valign="top">ignore variant with alleles over 4; (default: --max-allele 4) </td>
                                        </tr>
                                    </table>

                                    <p><strong>Note</strong> All above quality control functions can be turned off by  <tt>--no-qc</tt>.</p>

                                    <p>In addition, KGGSeq can use phenotype information (specified by "--indiv-pheno" or "--ped-file") to filter variants, especially when multiple samples are included in one input file such as trio or case/control study design.</p>
                                    <p><code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --min-heta 1(or --min-homa 1) --min-hetu 1(or --min-homu1) --min-obsa 2(or --min-obsu 2) --min-obs 2 --o-vcf path/to/file3</code></p>

                                    <p>Moreover, It's flexible to match the sample size and missingness of genotypes. Here are the options for different study design. (cases = affected subjects ; controls = unaffected subjects)
                                        <table  class="list">
                                            <tr>
                                                <td valign="top"><b>Arguments/values</b></td>
                                                <td valign="top"><b>Annotation</b></td>
                                            </tr>
                                            <tr>
                                                <td valign="top">--min-heta 1</td>
                                                <td valign="top">Set minimal observed number of heterozygotes genotypes in cases as 1. </td>
                                            </tr>
                                            <tr>
                                                <td valign="top">--min-homa 1</td>
                                                <td valign="top">Set minimal observed number of alternate homozygsote(s) genotypes in cases as 1; <tt>--min-heta</tt> and <tt>--min-homa</tt> are logical OR relation.</td>
                                            </tr>
                                            <tr>
                                                <td valign="top">--min-hetu 1</td>
                                                <td valign="top">Set minimal observed number of heterozygotes genotypes in controls as 1. </td>
                                            </tr>
                                            <tr>
                                                <td valign="top">--min-homu 1</td>
                                                <td valign="top">Set minimal observed number of alternate homozygsote(s) genotypes in controls as 1; <tt>--min-hetu</tt> and <tt>--min-homu</tt> are logical OR relation.</td>
                                            </tr>
                                            <tr>
                                                <td valign="top">--min-obsa 2</td>
                                                <td valign="top">Set minimal observed number of non-missing genotypes in cases as 2.</td>
                                            </tr>
                                            <tr>
                                                <td valign="top">--min-obsu 2</td>
                                                <td valign="top">Set minimal observed number of non-missing genotypes in controls as 2; <tt>--min-obsa</tt> and <tt>--min-obsu</tt> are logical AND relation. </td>
                                            </tr>
                                            <tr>
                                                <td valign="top">--min-obs 2</td>
                                                <td valign="top">Set minimal observed number of non-missing genotypes in all samples as 2. </td>
                                            </tr>
                                            <tr>
                                                <td>--hwe-control 0.01</td>
                                                <td>Exclude variants in controls with the Hardy-Weinberg test p value &lt;= 0.01. By default, this tag is disabled.</td>
                                            </tr>
                                            <tr>
                                                <td>--hwe-case 0.01</td>
                                                <td>Exclude variants in cases with the Hardy-Weinberg test p value &lt;= 0.01. By default, this tag is disabled.</td>
                                            </tr>
                                            <tr>
                                                <td>--hwe-all 0.01</td>
                                                <td>Exclude variants in all subjects with the Hardy-Weinberg test p value &lt;= 0.01. By default, this tag is disabled.</td>
                                            </tr>
                                        </table>


                                        <p>After running individual genotype QC, there will be two output files produced: one default KGGSeq output stored in txt (or Excel by specifying "--excel"), in which the basic annotations for each clean variant are provided (see above section);
                                            the other one is clean vcf file, in which the variant row which does not meet the conditions specified will be removed completely, hence not considered in following analysis.</p>
                                        <p><strong>Note</strong> The genomic variant QC is also available for non-vcf type of input, by using type of variants, or genomic position only.</p>
                                        <h3><a name="SampleQC" > Sample QC </a> </h3><br />
                                        KGGSeq provides two set of functions to facilitate quality assessment of sequencing data of each subject. <br />

                                        Firstly, you can use KGGSeq to convert genotypes from  <a href="http://www.1000genomes.org/node/101" target="_blank">VCF</a> format to  PLINK binary genotype format (see <a href="#BinaryOutputs" target="_self">Outputs</a> part) so that routine sample QC (<a href="http://pngu.mgh.harvard.edu/~purcell/plink/ibdibs.shtml#genome" target="_blank">pairwise relatedness</a>, <a href="http://pngu.mgh.harvard.edu/~purcell/plink/summary.shtml#mendel" target="_blank">mendel error rate</a> and <a href="http://pngu.mgh.harvard.edu/~purcell/plink/summary.shtml#sexcheck" target="_blank">sex check</a>) in genome-wide association study (GWAS) analysis can be performed quickly. The samples with unexpected relationship or error rate are suggested to be removed.
                                        <p><code>java -jar kggseq.jar --vcf-file path-to-file1 --ped-file path/to/file2 &lt;Variants and Genotype QC Settings&gt; --db-merge &lt;Reference genotype set&gt; --o-plink-bed </code>
                                            <code>plink --bfile kggseq.merged --genome</code>
                                            The following is the reference genotypes integrated by KGGSeq: (KGGSeq will automatically download it for you according to the set name specified by <tt>--db-merge</tt> )<br />

                                            <table class="list">
                                                <tr >
                                                    <th >Reference genotype set</th>
                                                    <th >Description</th>
                                                </tr>
                                                <tr >
                                                    <td >hapmap2.r22.ceu.hg19</td>
                                                    <td rowspan="3">
                                                        <span>
                                                            Haplotypes of Hapmap 2 release 22. Convert the coordinates to be hg19 from hg18 by UCSC lift over function. Complied from
                                                            <a href="http://www.sph.umich.edu/csg/abecasis/MACH/download/HapMap-r22.html" target="_blank">here</a>.
                                                        </span>
                                                    </td>
                                                </tr>
                                                <tr >
                                                    <td >hapmap2.r22.chbjpt.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td >hapmap2.r22.yri.hg19</td>
                                                    <td ></td>
                                                </tr>
                                                <tr >
                                                    <td>hapmap3.r2.ceu.hg19</td>
                                                    <td rowspan="5">
                                                        <span>
                                                            Haplotpyes of Hapmap 3 release 2. Convert the coordinates to be hg19 from hg18 by UCSC lift over function. Compiled from
                                                            <a href="http://www.sph.umich.edu/csg/abecasis/MACH/download/HapMap3.r2.b36.html" target="_blank">here</a>.
                                                        </span>
                                                    </td>
                                                </tr>
                                                <tr >
                                                    <td>hapmap3.r2.chbjpt.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td>hapmap3.r2.mex.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td>hapmap3.r2.tsi.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td>hapmap3.r2.yri.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td>1kg.phase1.v3.shapeit2.asn.hg19</td>
                                                    <td rowspan="4">
                                                        Haplotpyes of 1000 Genomes Project version 3. Donwload from
                                                        <a href="http://www.sph.umich.edu/csg/abecasis/MACH/download/1000G.2012-03-14.html" target="_blank">here</a>.
                                                    </td>
                                                </tr>
                                                <tr>
                                                    <td>1kg.phase1.shapeit2.v3.eur.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td >1kg.phase1.shapeit2.v3.afr.hg19</td>
                                                    <td ></td>
                                                </tr>
                                                <tr>
                                                    <td>1kg.phase1.v3.shapeit2.amr.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td>1kg.phase3.v5.shapeit2.eur.hg19</td>
                                                    <td rowspan="5">
                                                        Haplotpyes of 1000 Genomes Project Phase3 of version 5. Donwload from
                                                        <a href="ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20130502/" target="_blank">here</a>.

                                                    </td>
                                                </tr>
                                                <tr>
                                                    <td >1kg.phase3.v5.shapeit2.afr.hg19</td>
                                                    <td ></td>
                                                </tr>
                                                <tr>
                                                    <td>1kg.phase3.v5.shapeit2.amr.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td>1kg.phase3.v5.shapeit2.sas.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr>
                                                    <td>1kg.phase3.v5.shapeit2.eas.hg19</td>
                                                    <td></td>
                                                </tr>
                                                <tr><td colspan="2">We acknowledge the complied VCF data of 1000 Genomes projects by the author of MACH, Dr. LI Yun. To see detailed the description about the data,
                                                        please visit<br/><a href="http://www.sph.umich.edu/csg/abecasis/MACH/download/1000G.2012-03-14.html" target="_blank">http://www.sph.umich.edu/csg/abecasis/MACH/download/1000G.2012-03-14.html</a></td></tr>
                                            </table>  <br />
<<<<<<< HEAD
                                            An unexpectedly high pairwise relatedness (denoted by the PI_HAT of plink) suggests cross-individual contamination<br/>
=======
                                            An unexpectedly high pair-wise relatedness (denoted by the PI_HAT of plink) suggests cross-individual contamination<br/>
>>>>>>> origin/master
                                            <br />
                                            <strong>Note</strong> To estimate the relatedness (the PI_HAT), you normally do NOT need the genotypes datasets from 1000 Genome Projects (which will require rather large memory). The merging with ancestry-matched HapMap genotype data is often sufficient for an accurate estimation of relatedness. However, the estimation of PI_HAT without reference genotypes in a small local sample will result in an over-estimated value because of the inaccurate allele frequencies derived from the tiny sample.
                                            <code>plink --bfile kggseq --mendel</code>
                                            <code>plink --bfile kggseq --check-sex</code>
                                        </p>
                                        Secondly, we provide plotting function (see Plotting part) to visualization the distribution of minor alleles per individual. The outlier from the distribution of total rare variants (MAF < 0.01, or novel in the common database) across all individuals can be easily detected and then ignored in the following analysis.
                                        <p><code>java -jar kggseq.jar --vcf-file path-to-file1 --out path/to/prefixname --mafplot</code>
                                            <code>java -jar kggseq.jar --vcf-file path-to-file1 --ped-file path/to/file2  --db-filter hg18_dbsnp138,hg18_dbsnp141 --allele-freq 0,0.01 --o-vcf path/to/file3</code></p>
                                        <p><strong>Note</strong> The meanings for <tt>-db-filter</tt> and <tt>--allele-freq</tt> in this section are provied in  "<a href="#AlleleFrequencyTrimming" target="_self">Allele Frequency Trimming</a>" section.</p>
                                        <p>&nbsp;</p>
                                </td>
                            </tr>

                            <tr>
                                <td>
                                    <h2><a  id="FilteringAndPrioritization" >Filtering</a></h2><hr />
                                    KGGSeq offers  simple but potentially powerful strategies to variants filtering and prioritization, which help users isolate most promising disease causal candiate variants from sequence variants to be considered.Users could prioritize variants based on some or all of the following evidences and/or functions: Genetic inheritance sharing, gene feature, allele frequency,  biological function prediction, sequence variant type, and phenotype-relevant filtering.
                                    <h3><a  id="GeneticInheritanceSharing" > Genetic inheritance sharing</a></h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Inheritance pattern filter: <tt>--genotype-filter</tt></p>

                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2  --genotype-filter 1,2</code></p>
                                    <p>Filter out variants for which their genotypes are not consistent with the assumption of disease inheritance pattern. </p>
                                    Functions and applicable models for each code of <tt> '--genotype-filter'</tt>options are listed below:
                                    <table   class="list">
                                        <tr>
                                            <td valign="top"><b>Code</b></td>
                                            <td valign="top"><b>Function</b></td>
                                            <td valign="top"><b>Applicable model</b></td>
                                        </tr>
                                        <tr>
                                            <td valign="top">1</td>
                                            <td valign="top">Exclude variants at which affected subjects have heterozygous genotypes.</td>
                                            <td valign="top">Recessive</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">2</td>
                                            <td valign="top">Exclude variants at which both affected and unaffected subjects have the same homozygous genotypes.</td>
                                            <td valign="top">Recessive and full penetrance causal mutation(s) in the sample; <b>compound-heterozygosity</b></td>
                                        </tr>
                                        <tr>
                                            <td valign="top">3</td>
                                            <td valign="top">Exclude variants at which affected subjects have reference homozygous genotypes.</td>
                                            <td valign="top">Dominant without consanguineous mating </td>
                                        </tr>
                                        <tr>
                                            <td valign="top">4</td>
                                            <td valign="top">Exclude variants at which both affected and unaffected subjects have the same heterozygous genotypes.</td>
                                            <td valign="top">Dominant with full penetrance causal mutation(s) in the sample</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">5</td>
                                            <td valign="top">Exclude variants at which affected subjects have alternative homozygous genotypes.</td>
                                            <td valign="top">Dominant  without heterogeneity</b></td>
                                        </tr>
                                        <tr>
                                            <td valign="top">6</td>
                                            <td valign="top">Exclude variants at which affected family members have NO shared alleles.</td>
                                            <td valign="top">Full penetrance causal mutation(s) in the sample</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">7</td>
                                            <td valign="top">ONLY include variants at which an offspring has one or two non-inherited alleles.
                                                <br />
                                                <br />
                                                <strong>Note</strong>
                                                Used as <tt> --genotype-filter 4,7 </tt> to filter out the de novo mutations where both affected and unaffected subjects have the same heterozygous genotypes. <br />
                                                <strong>Note</strong>  Used as <tt> --ignore-homo  </tt> to ignore the  de novo mutations where affected and unaffected subjects have the same ALTERNATIVE homozygous genotypes.

                                            </td>
                                            <td valign="top">Detect <I>de novo</I> mutations which have nearly full penetrance.<br />
                                                In the main output file, there is a column named <B>DenovoMutationEvent</B> to record the genotypes of a child and his or her parents.<br />
                                                <br />
                                                Example: <I>N140_0:0/1:46,59&N140_1:0/0:57,0&N140_2:0/0:68,0</I><br />
                                                The child N140_0 has genotype 0/1 with 46 and 59 reads carrying reference alleles and alternative alleles respectively. The father N140_1 and mother N140_2 are homozygous 0/0.
                                            </td>
                                        </tr>
                                        <tr>
                                            <td valign="top">8</td>
                                            <td valign="top">ONLY include variants at which 1) the paired non-tumor (unaffected) and tumor (affected) samples have different genotypes.<br />
                                                <br />
                                                <strong>Note</strong> The tag <tt> --indiv-pair </tt> has to be used together to specify the matched non-tumor and tumor samples, e.g.<br />  <tt> --indiv-pair tumorIndivID1:normalIndivID1,tumorIndivID2:normalIndivID2</tt>
                                            </td>
                                            <td valign="top">Detect <I>somatic mutation(s) </I> among matched tumor and non-tumor samples. </td>
                                        </tr>
                                    </table>
                                    <strong>Note</strong> Multiple filtration codes have logical OR relationship.<br /><br />
                                    <strong>Note</strong> The inheritance mode based filtration is proposed under strong assumptions for rare Mendelian disease with clear inheritance mode. If the inheritance mode is elusive, such filtration is not suggested as it may lead to the miss of genuine causal mutation(s). This is particularly true when one has no sufficient information to distinguish the <a href="http://en.wikipedia.org/wiki/Compound_heterozygosity" target="_blank"><b>compound-heterozygosity</b></a> diseases from the <a href="http://en.wikipedia.org/wiki/Recessive" target="_blank"><b>recessive</b></a> diseases.

                                    <p><img src="imgs/leaf.gif" width="10" height="8" />  Extract mutations affecting both copies of a gene in a diploid: <tt>--double-hit-gene-trio-filter</tt> or <tt>--double-hit-gene-phased-filter</tt></p>
<<<<<<< HEAD
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --db-gene refgene --gene-feature-in 0,1,2,3,4,5,6 --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.033 --db-score dbnsfp --mendel-causing-predict best --double-hit-gene-trio-filter (OR --double-hit-gene-phased-filter)</code></p>
                                    <p>This function is designed for a disorder suspected to be under compound-heterozygous or recessive inheritance mode, in which both copies of a gene on the two homologous chromosomes of a patient are damaged by two different mutations or the same mutation. For recessive mode, it simply checks variants with homozygous genotypes in patients. For the compound-heterozygous mode, it can use two different input data, <b>phased genotypes of a patient</b> or  <b>unphased genotypes in a trio</b>. Here the trio refers to <b>the two parents and an offspring</b>.  When these alleles causing a disease at one locus, it follows the <a href="http://en.wikipedia.org/wiki/Recessive" target="_blank"><b>recessive</b></a> model; and when they are at two loci, it follows the <a href="http://en.wikipedia.org/wiki/Compound_heterozygosity" target="_blank"><b>compound-heterozygosity</b></a> model. In both cases, a gene is hit twice. This is the reason why it has the name 'double-hit gene'. </p>
=======
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --db-gene refgene --gene-feature-in 0,1,2,3,4,5,6 --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.033 --db-score dbnsfp --mendel-causing-predict all --double-hit-gene-trio-filter (OR --double-hit-gene-phased-filter)</code></p>
                                    <p>This function is designed for a disorder suspected to be under compound-heterozygous or recessive inheritance mode, in which both copies of a gene on the two ortholog chromosomes of a patient are damaged by two different mutations or the same mutation. For recessive mode, it simply checks variants with homozygous genotypes in patients. For the compound-heterozygous mode, it can use two different input data, <b>phased genotypes of a patient</b> or  <b>unphased genotypes in a trio</b>. Here the trio refers to <b>the two parents and an offspring</b>.  When these alleles causing a disease at one locus, it follows the <a href="http://en.wikipedia.org/wiki/Recessive" target="_blank"><b>recessive</b></a> model; and when they are at two loci, it follows the <a href="http://en.wikipedia.org/wiki/Compound_heterozygosity" target="_blank"><b>compound-heterozygosity</b></a> model. In both cases, a gene is hit twice. This is the reason why it has the name 'double-hit gene'. </p>
>>>>>>> origin/master
                                    <p>Besides the file storing prioritized sequence variants, this analysis with <tt>--double-hit-gene-trio-filter</tt> option will lead to a list of double-hit genes in at least one subject. The counts of double-hit genes of each subject are saved in *.doublehit.gene.trios.flt.count.xls (or *.doublehit.gene.trios.flt.count.txt) and the genotypes of involved sequence variants are saved in *.doublehit.gene.trios.flt.gty.xls (or *.doublehit.gene.trios.flt.gty.txt).  <br/></p>
                                    <p>
                                        The '*doublehit.gene.trios.flt.count*' and the '*doublehit.gene.trios.flt.gty*' files or sheets contain multiple columns:
                                        <h4>
                                            Gene: Gene symbols	<br />
                                            PubMed: The PumMed ID(s) of papers co-mentioning the gene and phenotypes(or diseases) specified by '--phenotype-term'. <br />
                                            ExonLen: The length of total exons in kb. The overlapped regions are ignored.<br />                                           
<<<<<<< HEAD
                                            CountHitCase: Total number of subjects having double-hit variants in the gene.<br />
=======
                                            CountHit: Total number of subjects having double-hit variants in the gene.<br />
                                            CountHitPseudoControl: Total number of subjects having double-hit variants on the untranslated chromosomes copies of the gene. THe higher value in this column, the less likely the gene cause the disease. <br />
>>>>>>> origin/master
                                            &#60;Subjects&#62;: Whether a subject carry the double-hit variants,1 for yes and 0 for no in the 'flt.count' file or sheet. In the 'gty.count' file or sheet, the variants positions and genotypes are shown. <br/>
                                            Example: <br/>
                                            17797224:0/0,0/1,0/1;17827193:0/1,0/0,0/1: The genotypes of <b>father</b>, <b>mother</b>, and <b>child</b> are <b>0/0</b>, <b>0/1</b> and <b>0/1</b> at the variant with coordinate 17797224. At the other variant, the genotypes of <b>father</b>, <b>mother</b>, and <b>child</b> are <b>0/1</b>, <b>0/0</b> and <b>0/1</b>. Apparently, both copies of this gene in the child are hit by the two mutations. 
                                        </h4>
                                    </p>

                                    The corresponding files produced by the analysis with <tt>--double-hit-gene-phased-filter</tt> option are  *.doublehit.gene.phased.flt.count.xls (or *.doublehit.gene.phased.flt.count.txt) and  *.doublehit.gene.phased.flt.gty.xls (or *.doublehit.gene.phased.flt.gty.txt).
                                    <p>
                                        The '*doublehit.gene.phased.flt.count*' and the '*doublehit.gene.phased.flt.gty*' files or sheets contain multiple columns:
                                        <h4>
                                            Gene: Gene symbols	<br />
                                            PubMed: The PumMed ID(s) of papers co-mentioning the gene and phenotypes(or diseases) specified by '--phenotype-term'. <br />
                                            P: The p-values of chi-square test to examine whether the double-hit mutations are associated with a disease.<br />
                                            CountCtl: Total number of controls having double-hit variants in the gene.<br />
                                            CountCase: Total number of patients having double-hit variants in the gene.<br />
                                            &#60;Subjects&#62;: Whether a subject carry the double-hit variants,1 for yes and 0 for no in the 'flt.count' file or sheet. In the 'gty.count' file or sheet, the variants positions and genotypes are shown. <br/>
                                            Example: <br/>
                                            105352884:0|1;105353305:1|0: The genotypes of a subject are <b>0|1</b> and <b>1|0</b> at the variants with coordinate 105352884 and 105353305. Apparently, both copies of this gene in the subject are hit by the two mutations. 
                                        </h4>
                                    </p>
<<<<<<< HEAD
                                    <p><strong>Note</strong> Both filters have a strong assumption of full penetrance. </p>
                                    <p><strong>Note</strong> The <tt>--double-hit-gene-trio-filter</tt> or <tt>--double-hit-gene-phased-filter</tt> only works for inputs specified by  <tt>--vcf-file</tt> and requires parents' genotypes to infer the phases of the alleles. And sequence variants with missed parents' genotypes are ignored.</p>
                                    <p><strong>Note</strong> To ignore homozygous mutation (i.e., implicating the autosomal recessive model), please add the tag <tt>--ignore-homo</tt>. </p>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" />  Extract a gene on which EVERY affected subject has at least one case-unique alternative allele. But these alleles may be from different variants: <tt>--unique-gene-filter</tt></p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.02 --db-score dbnsfp --mendel-causing-predict best --unique-gene-filter</code></p>
=======
                                    <p><strong>Note</strong> The <tt>--double-hit-gene-trio-filter</tt> or <tt>--double-hit-gene-phased-filter</tt> only works for inputs specified by  <tt>--vcf-file</tt> and requires parents' genotypes to infer the phases of the alleles. And sequence variants with missed parents' genotypes are ignored.</p>
                                    <p><strong>Note</strong> To ignore homozygous mutation (i.e., implicating the autosomal recessive model), please add the tag <tt>--ignore-homo</tt>. </p>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" />  Extract a gene on which EVERY affected subject has at least one case-unique alternative allele. But these alleles may be from different variants: <tt>--unique-gene-filter</tt></p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.02 --db-score dbnsfp --mendel-causing-predict all --unique-gene-filter</code></p>
>>>>>>> origin/master
                                    <p>Sequence variants beyond this type of genes will be filtered out.</p>
                                    <p><strong>Note</strong> <tt>--unique-gene-filter</tt> only works for inputs specified by  <tt>--vcf-file</tt>.</p>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Filter sequence variants by Identical by state (IBS): <tt>--ibs-case-filter X</tt></p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --ibs-case-filter X </code></p>
                                    <p>Search the longest region (in minimal length X kb) of each interesting variant, in which there is at least one allele identical (IBS&#62;0) among all cases at all consecutive variants. This function can be used to highlight potential causal mutations or exclude non-disease-relevant sequence variants as the causal mutation(s) is very likely to have long shared regions among all affected family members of Mendelian dominant and recessive diseases. However, IBS checking is just a quick but over-estimation of <a href="http://en.wikipedia.org/wiki/Identity_by_descent" target="_blank">identical by descent (IBD)</a>, which is a more accurate measure of shared regions among related individuals. KGGSeq currently has no function to estimate IBD regions but allow users to input the IBD information estimated by third-party tools into KGGSeq for annotation.
                                    </p>
                                    <p><strong>Note</strong> <tt>--ibs-case-filter</tt> only works for inputs specified by  <tt>--vcf-file</tt>.</p>
                                    In the main output file, two columns will be appended:
                                    <h4>
                                        LongestIBSRegion:The longest region in which all consecutive variants hae IBS over 0 in patients. <br/>
                                        Example: 'chr2:46259552-57303414#Var:26469StartAtTailEndAtTail' means in the IBS region chr2:46259552-57303414 there are 26469 consecutive variants. This region start with and end at the boundaries of the input regions or chromosomes  <br />
                                        LongestIBSRegionLength(bp): The length of the region. <br />
                                    </h4>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Identical by descent (IBD) annotation: <tt>--ibd-annot</tt></p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --ibd-annot path/to/file3</code></p>
                                    <p>User can also provide <b>IBD</b> (i.e.,  the two alleles arose from the same allele in an earlier generation) or  significant linkage regions, which are estimated by a third party statistical genetics tools (such as <a href="http://faculty.washington.edu/browning/beagle/beagle.html" target="_blank">BEAGLE</a>, <a href="http://www.sph.umich.edu/csg/abecasis/Merlin/index.html" target="_blank">Merlin</a> and <a href="http://www.genetics.ucla.edu/software/simwalk_doc/SW2_Overview.html" target="_blank">SimWalk2</a>). Variants within these regions will be highlighted. For an ibd or linkage annotation file, title line is needed, and the  file format is like: </p>
                                    <h4>
                                        CHR	START	END<br />
                                        1	6134174 	10082841<br />
                                        1	202855724	207862411<br />
                                        <hr />
                                        <strong>Note</strong> Title line is needed.
                                    </h4>
                                    In the main output file, two columns will be appended:
                                    <h4>
                                        IBDRegion: The region with IBD information. <br />
                                        IBDRegionAnnot: The annotation information at the region. <br />
                                    </h4>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" />  Filter sequence variants by homozygosity: <tt>--homozygosity-case-filter X</tt></p>
                                    <p><code>java -jar ./kggseq.jar  --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --homozygosity-case-filter X</code></p>
                                    <p>
                                        Search the longest region (in minimal length X kb) of consecutive <a href="http://en.wikipedia.org/wiki/Zygosity#Homozygous" target="_blank"><b>homozygous</b></a> genotype for each interesting variant with minimal length X kb in cases but not in controls. For cancer diseases, this function can be used to examine the <a href="http://en.wikipedia.org/wiki/Loss_of_heterozygosity" target="_blank">loss of heterozygosity</a> straightforwardly. For rare recessive diseases, it can be used to examine the  <a href="http://www.orcades.ed.ac.uk/popgenetics/homozygosity.html" target="_blank">runs of homozygosity</a>  straightforwardly. But KGGSeq currently does not have methods to test the statistical significance.
                                    </p>
                                    <p><strong>Note</strong> <tt>--homozygosity-case-filter</tt> only works for inputs specified by  <tt>--vcf-file</tt>.</p>

                                    In the main output file, two columns will be appended:
                                    <h4>
                                        LongestHomozygosityRegion longest region in which all consecutive variants hae IBS over 0 in patients. <br/>
                                        Example: 'chr2:46259552-57303414#Var:26469StartAtTailEndAtTail' means in the IBS region chr2:46259552-57303414 there are 26469 consecutive variants. This region start with and end at the boundaries of the input regions or chromosomes  <br />
                                        LongestHomozygosityRegionLength(bp): The length of the region. <br />
                                    </h4>

                                    <h3><a  id="GeneFeatureFiltering" > Gene feature filtering</a> </h3>
                                    <p>Filtration according to gene features:  <tt> --db-gene & --gene-feature-in</tt></p>
                                    <p><code>java -jar ./kggseq.jar  --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --db-gene refgene,gencode,knowngene --gene-feature-in 0,1,2,3,4,5,6</code></p>
                                    <p>The <tt>'--dbgene'</tt> sets database(s) to annotate and filter variants.</p>

                                    <table class="list">
                                        <tr>
                                            <td valign="top"><b>Gene database label</b></td>
                                            <td valign="top"><b>Supported reference genome</b></td>
                                            <td valign="top"><b>Description</b></td>
                                            </b>
                                        </tr>
                                        <tr>
                                            <td valign="top">refgene</td>
                                            <td valign="top">hg18, hg19 and hg38</td>
                                            <td valign="top">The RefGene database compiled by UCSC from <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/refGene.txt.gz">hg18 refGene</a> and <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/refGene.txt.gz">hg19 refGene</a>
                                                <br />  Note: RefSeq has NO mitochondria gene definition.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">gencode</td>
                                            <td valign="top">hg19 and hg38</td>
                                            <td valign="top">The GENCODE gene sets, gencode.v19.annotation.gtf for hg19 and gencode.v23.annotation.gtf for hg38.See <a href="http://www.gencodegenes.org/">home page </a> and the <a href="http://genome.cshlp.org/content/22/9/1760.abstract">paper</a> of GENCODE for details.
                                                <br /> Note: GENCODE contains similar number of coding genes but more transcripts than RefGene. It HAS the mitochondria gene definition.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">knowngene</td>
                                            <td valign="top">hg18, hg19 and hg38</td>
                                            <td valign="top">The UCSC knonwGene database compiled by UCSC from <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/knownGene.txt.gz">hg18 knownGene</a> and <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/knownGene.txt.gz">hg19 knownGene</a>
                                            </td>
<<<<<<< HEAD
                                        </tr> 
                                        <tr>
                                            <td valign="top">path/to/file</td>
                                            <td valign="top">-</td>
                                            <td valign="top">KGGSeq also accepts a gene dataset customized by users. The example can be seen <a href="http://grass.cgs.hku.hk/limx/kggseq/download/customGeneDB.txt", target="_blank">here</a>. The following are the column names:<br/>

                                                bin: The bin ID number <br/>
                                                name: The name of a transcript <br/>
                                                chrom: The chromosome name <br/>
                                                strand: The strand information of the transcript <br/>
                                                txStart: The start site of a transcript <br/>
                                                txEnd: The end site of a transcript <br/>
                                                cdsStart: The start site of coding DNA sequence <br/>
                                                cdsEnd: The end site of coding DNA sequence <br/>
                                                exonCount: The number of exons <br/>
                                                exonStarts: The start sites of all exons <br/>
                                                exonEnds: The end sites of all exons <br/>
                                                score: mapping score <br/>
                                                name2: Gene name <br/>
                                                cdsStartStat: Un-used <br/>
                                                cdsEndStat: Un-used <br/>
                                                exonFrames: Un-used <br/>
                                                sequences: DNA sequences of exons <br/>
                                                delSites: Deletion sites on the human reference genome, compared to the latest cDNA <br/>
                                                insSites: Insertion sites on the human reference genome, compared to the latest cDNA 


                                            </td>
                                        </tr>                                         
                                    </table>

=======
                                        </tr>                                         
                                    </table>
>>>>>>> origin/master
                                    <p>  With this filter, variants beyond <tt> '--gene-feature-in' </tt> region will be excluded. It is <tt> '--gene-feature-in 0,1,2,3,4,5...,15' </tt> by default.</p>
                                    Illustration of number codes for the gene features:

                                    <table class="list">
                                        <tr>
                                            <td valign="top"><b>Feature</b></td>
                                            <td valign="top"><b>Code</b></td>
                                            <td valign="top"><b>Explanation</b></td>
                                            </b>
                                        </tr>
                                        <tr>
                                            <td valign="top">Frameshift</td>
                                            <td valign="top">0</td>
                                            <td valign="top">Short insertion or deletion result in a completely different translation from the original.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Nonframeshift</td>
                                            <td valign="top">1</td>
                                            <td valign="top">Short insertion or deletion results in loss of amino acids in the translated proteins.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Startloss</td>
                                            <td valign="top">2</td>
                                            <td valign="top">Indels or nucleotide substitution result in the loss of start codon(ATG) (mutated into a non-start codon).</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Stoploss</td>
                                            <td valign="top">3</td>
                                            <td valign="top">Indels or nucleotide substitution result in the loss of stop codons (TAG, TAA, TGA)</td>
                                        </tr>

                                        <tr>
                                            <td valign="top">Stopgain</td>
                                            <td valign="top">4</td>
                                            <td valign="top">Indels or nucleotide substitution result in the new stop codons (TAG, TAA, TGA), which may truncate the protein.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Splicing</td>
                                            <td valign="top">5</td>
                                            <td valign="top">variant is within 2-bp of a splicing junction (use <tt>--splicing x</tt> to change this, the unit of x is base-pair)</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Missense</td>
                                            <td valign="top">6</td>
                                            <td valign="top">Variants result in a codon coding for a different amino acid (missense)</td>
                                        </tr>                                       
                                        <tr>
                                            <td height="22" valign="top">Synonymous</td>
                                            <td valign="top">7</td>
                                            <td valign="top">Nucleotide substitution does not change amino acid.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Exonic</td>
                                            <td valign="top">8</td>
                                            <td valign="top">Due to loss of sequences in reference database, this variant can only be mapped into exonic region without more precise annotation.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">UTR5</td>
                                            <td valign="top">9</td>
                                            <td valign="top">Within a 5' untranslated region</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">UTR3</td>
                                            <td valign="top">10</td>
                                            <td valign="top">Within a 3' untranslated region</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Intronic</td>
                                            <td valign="top">11</td>
                                            <td valign="top">Within an intron</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Upstream</td>
                                            <td valign="top">12</td>
                                            <td valign="top">Within 1-kb region upstream of transcription start site (use <tt>--neargene x</tt> to change this, the unit of x is base-pair)</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Downstream</td>
                                            <td valign="top">13</td>
                                            <td valign="top">Within 1-kb region downtream of transcription end site (use <tt>--neargene x</tt> to change this, the unit of x is base-pair)</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">ncRNA</td>
                                            <td valign="top">14</td>
                                            <td valign="top">Within a transcript without protein-coding annotation in the gene definition (see Notes below for more explanation)</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Intergenic</td>
                                            <td valign="top">15</td>

                                            <td valign="top">variant is in intergenic region</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Monomorphic</td>
                                            <td valign="top">16</td>
                                            <td valign="top">It is not a sequence variation actually, which may be resulted from bugs in reference genome in variant calling. </td>
                                        </tr>
                                        <tr>
                                            <td valign="top">Unknown</td>
                                            <td valign="top">17</td>
                                            <td valign="top">Variants has no annotation.</td>
                                        </tr>
                                    </table>

                                    The output of this function is a list of variables providing all involved gene features of a considered variant.
                                    <h4>
                                        <b> MostImportantFeatureGene: </b>The gene corresponds to the variant&rsquo;s smallest gene feature code.<br />
                                        <b> MostImportantGeneFeature: </b>The variant&rsquo;s smallest gene feature code.<br />
                                        <b> RefGeneFeatures: </b>All matched gene features in the reference gene database.<br />
                                        <b> GENCODEFeatures: </b>All matched gene features in the GENCODE gene database. 
                                    </h4>
                                    <br />
                                    <br />
                                    The gene feature annotations are described in accordance with the <a href="http://www.hgvs.org/mutnomen/examplesDNA.html" target="_blank"><b>Human Genome Variation Society (HGVS) recommendations</b></a>. The followings are some examples:
                                    <h4>
                                        <i>KLHL17:NM_198317:c.1918A>C:p.T640P:(12Exons):exon12:missense</i><br />	
                                        <b>Means: At nucleotide 76 of gene KLHL17's transcript NM_198317, an A is changed to a C, which changes the 640<sup>th</sup> amino-acid T to be P in this transcript's protein. This transcript has 12 exons in total and this variant is in the 12<sup>th</sup> exon.</b><br />								
<<<<<<< HEAD
                                        <i>SYNGR1:NM_004711:(4Exons):exon4:c.C605ins+CAA:p.P202ins??</i><br />										
                                        <b>Means: At nucleotide 605 of gene SYNGR1's transcript NM_004711, there is a 3bp insertion. This transcript has 4 exons in total and this variant is in the 4<sup>th</sup> exon. The '+' denotes an insertion. The affected peptide is unknown ('??')</b><br />                                        
                                        <i>PKD1L2:NM_001076780:c.A705delAA-:p.N236:(18Exons):exon4:frameshift</i><br />										
                                        <b>Means: At nucleotide 705 of gene PKD1L2's transcript NM_001076780, there is a 2bp deletion. This transcript has 18 exons in total and this variant is in the 4<sup>th</sup> exon. The '-' denotes an deletion. The variant results in a frameshift of protein-coding .</b><br />                                       
=======
>>>>>>> origin/master
                                        <i>DNAJC8:NM_014280:c.237+2T>G:(9Exons):exon3GTdonor</i><br />
                                        <b>Means:This is a T to G substitution of the 2<sup>nd</sup> nucleotide of the 2 <sup>nd</sup> intron  (from its 5' part) positioned between coding DNA nucleotides 237 and 238, which will affect the GT splicing donor.</b><br />
                                        <i>HNRNPCL2:NM_001136561:c.-169C>G:(2Exons):5UTR</i><br />
                                        <b>Means: This is a C to G substitution of nucleotide, 169 nucleotides upstream of the ATG translation initiation codon (coding DNA -169).</b><br />										 
                                        <i>PRAMEF4:NM_001009611:c.*62C>T:(4Exons):3UTR</i><br />
                                        <b>Means: This is a C to T substitution of nucleotide coding DNA 62, located in the 3' UTR.</b><br />
                                        <i>C1orf159:NM_017891:c.311-23T>C:(10Exons):intronic6</i><br />	
                                        <b>Means: This is a T to C substitution of the 23 nucleotide of the 6 <sup>th</sup> intron  (from its 3' part) positioned between coding DNA nucleotides 310 and 311.</b><br />                                        
                                        <i>HNRNPCL1:NM_001013631:c.*138G>A:(2Exons):downstream</i><br />
                                        <b>Means: This is a C to A substitution of nucleotide located 293 nucleotides downstream of the gene, i.e. the polyA-addition site.</b><br />
                                        <i>TNFRSF18:NM_148901:c.-61C>T:(4Exons):upstream</i><br />										
<<<<<<< HEAD
                                        <b>Means: This is a C to T substitution of nucleotide, 61 nucleotides upstream of the ATG translation initiation codon (coding DNA -61).</b><br />    
=======
                                        <b>Means: This is a C to T substitution of nucleotide, 61 nucleotides upstream of the ATG translation initiation codon (coding DNA -61).</b><br />                                        
>>>>>>> origin/master
                                    </h4>
                                    <br/>
                                    <p >
                                        <b> Table: Compare non-synonymous gene feature annotation of three popular tools and two gene models for variants discovered in 1000 Genomes Project AFR panel </b> <br/>
                                        <img   src="imgs/annotationcmp.png"   width="90%" height="90%"/> <br/>
                                        Note: The RefGene model was provided by UCSC on Oct 10, 2015 for KGGSeq; and for ANNOVAR and the latest RefGene model (by Oct 10, 2015) from their websites was used. a: Because ANNOVAR has no startloss category but annotates startloss as missense, no comparison was made for this category. b:  Because dbSNP has no startloss category, no consistency was calculated for this category. c: rate of consistency with annotations in the dbSNP. The default parameters settings of each tool were used for the annotation. 
                                    </p>


                                    <h3><a  id="AlleleFrequencyTrimming" > Allele frequency trimming by 4 ways</a> </h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> <b>1. </b> Allele frequency trimming according to  database: <tt> --db-filter </tt> </p>
                                    <code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2  --db-filter 1kg201204,dbsnp141,ESP6500AA,ESP6500EA  --rare-allele-freq 0.005 --db-filter-hard dbsnp138nf</code>
                                    This mode enables  variants filtering according to allele frequencies recorded in public data sources. <br />
                                    The <tt>'--db-filter' </tt>is use to set databases for filtration. In the above case, variants with alternative allele frequency EQUAL to or over 0.005 will be excluded. <b>No spaces </b> are allowed between and within database names.
                                    <br />
                                    Moreover, <tt>'--db-filter-hard' </tt>is use to filter <b>ALL</b>  existing  variants in a database. But it should be always careful to use this hard filtering.
                                    <br />
                                    <p>                                        
                                        <strong>Note</strong>1. To keep variants in a frequency range,[a,b], please use  <tt>'--allele-freq a,b'</tt> option. The boundaries a and b are included.  <br />
                                        <strong>Note</strong>2. The options, <tt>'--allele-freq' </tt>and <tt>'--rare-allele-freq'</tt>, are mutually exclusive and the latter has higher priority.  <br />
                                        <strong>Note</strong>3. By default,  <tt>'--rare-allele-freq 0.01'</tt>, will be used if allele frequency cutoff is set.<br />
                                        <strong>Note</strong>4. Difference between  <tt>'--rare-allele-freq'</tt> and <tt>'--allele-freq'</tt>: The <tt>'--rare-allele-freq 0,0.01'</tt> does  not consider the variants with missing frequency in public databases, but <tt>'--rare-allele-freq 0.01'</tt> does. So <tt>'--rare-allele-freq 0.01'</tt> returns much more variants than <tt>'--rare-allele-freq 0,0.01'</tt>.
                                    </p>
                                    <br />
                                    List of public databases for allele frequency trimming

                                    <table  class="list">
                                        <tr>                                            
                                            <td valign="top"><b>DB Label</b></td>
                                            <td valign="top"><b>Explanation</b></td>
                                        </tr>
                                        <tr>                                            
                                            <td valign="top">1kgeur201305</td>
                                            <td valign="top">495 subjects in the EUR panel of 1000 Genomes Project release in 2013 May (around 24.0 million sequence variants)</td>
                                        </tr>
                                        <tr>                                            
                                            <td valign="top">1kgeas201305</td>
                                            <td valign="top">496 subjects in the EAS panel of 1000 Genomes Project release in 2013 May (around 23.5 million sequence variants)</td>
                                        </tr>
                                        <tr>                                            
                                            <td valign="top">1kgafr201305</td>
                                            <td valign="top">645 subjects in the AFR panel of 1000 Genomes Project release in 2013 May (around 41.7 million sequence variants)</td>
                                        </tr>
                                        <tr>                                            
                                            <td valign="top">1kgsas201305</td>
                                            <td valign="top">485 subjects in the SAS panel of 1000 Genomes Project release in 2013 May (around 26.7 million sequence variants)</td>
                                        </tr>
                                        <tr>                                            
                                            <td valign="top">1kgamr201305</td>
                                            <td valign="top">346 subjects in the AMR panel of 1000 Genomes Project release in 2013 May (around 28.2 million sequence variants)</td>
                                        </tr>                                       
                                        <tr>

                                            <td valign="top">1kg201305</td>
                                            <td valign="top">All sample of 1000 Genomes Project release 2013 May (over 81 million sequence variants from 2504 subjects)</td>
                                        </tr>
                                        <tr>

                                            <td valign="top">1kg201204</td>
                                            <td valign="top">All sample of 1000 Genomes Project release 2012 April </td>
                                        </tr>
                                        <tr>

                                            <td valign="top">1kgafr201204</td>
                                            <td valign="top">African descendent samples of 1000 Genomes Project release in 2012 April </td>
                                        </tr>
                                        <tr>

                                            <td valign="top">1kgeur201204</td>
                                            <td valign="top">European descendent samples of 1000 Genomes Project release in 2012 April </td>
                                        </tr>
                                        <tr>

                                            <td valign="top">1kgamr201204</td>
                                            <td valign="top">American descendent samples of 1000 Genomes Project release in 2012 April </td>
                                        </tr>
                                        <tr>

                                            <td valign="top">1kgasn201204</td>
                                            <td valign="top">Asian descendent samples of 1000 Genomes Project release in 2012 April </td>
                                        </tr>

                                        <tr>

                                            <td valign="top">dbsnp135</td>
                                            <td valign="top">dbSNP version 135, compiled from UCSC.</td>
                                        </tr>
                                        <tr>

                                            <td valign="top">dbsnp137</td>
                                            <td valign="top">dbSNP version 137, compiled from UCSC.</td>
                                        </tr>
                                        <tr>

                                            <td valign="top">dbsnp138</td>
                                            <td valign="top">dbSNP version 138, compiled from UCSC.</td>
                                        </tr>
                                        <tr>

                                            <td valign="top">dbsnp138nf</td>
                                            <td valign="top">dbSNP version 138 without the flagged SNPs by UCSC (so nonflagged: nf). In UCSC the flagged SNPs include SNPs clinically associated by dbSNP, mapped to a single location in the reference genome assembly, and not known to have a minor allele frequency of at least 1%. </td>
                                        </tr>
                                        <tr>

                                            <td valign="top">dbsnp141</td>
                                            <td valign="top">dbSNP version 141, compiled from NCBI dbSNP V141. Note Indels are ingored.</td>
                                        </tr>
                                        <tr>

                                            <td valign="top">ESP5400</td>
                                            <td valign="top">a public variants dataset from NHLBI GO Exome Sequencing Project (<a href="http://evs.gs.washington.edu/EVS/" target="_blank">ESP</a>)</td>
                                        </tr>
                                        <tr>

                                            <td valign="top">ESP6500AA</td>
                                            <td valign="top">a public variants dataset of African American from NHLBI GO Exome Sequencing Project (<a href="http://evs.gs.washington.edu/EVS/" target="_blank">ESP</a>)</td>
                                        </tr>
                                        <tr>

                                            <td valign="top">ESP6500EA</td>
                                            <td valign="top">a public variants dataset of European American from NHLBI GO Exome Sequencing Project (<a href="http://evs.gs.washington.edu/EVS/" target="_blank">ESP</a>)</td>
                                        </tr>
                                        <tr>

                                            <td valign="top">exac</td>
<<<<<<< HEAD
                                            <td valign="top"> <a href="http://exac.broadinstitute.org/" target="_blank">Exome Aggregation Consortium (ExAC r0.3.1)</a>: Variants from 60,706 unrelated individuals sequenced. 
                                                The samples are partitioned in to 7 ancestrally different panels, East Asian(eas, n=4327), South Asian (sas, n=8256), African/African American(afr, n= 5203), American(amr, n=5789 ), 
                                                Finnish(fin, n=3307), Non-Finnish European(nfe, n=33370), Other (oth, n=454). By default (<tt>--db-filter exac</tt>), KGGSeq will use frequencies of all of the 7 panels for annotation and filtering. 
                                                You can also specify the panels of your interests by the abbreviations. For example, you can specify the frequencies by of East Asian and Non-Finnish European  by <tt> --db-filter exac.eas.nfe</tt>. </td>
=======
                                            <td valign="top"> <a href="http://exac.broadinstitute.org/" target="_blank">Exome Aggregation Consortium (ExAC)</a>: Variants from 61,486 unrelated individuals sequenced as part of various disease-specific and population genetic studies (<a href="http://evs.gs.washington.edu/EVS/" target="_blank">ESP</a>)</td>
>>>>>>> origin/master
                                        </tr>
                                    </table>

                                    The output of this function is a list of variables providing all involved gene features of a considered variant.
                                    <h4>
                                        <b> MaxDBAltAF: </b>The maximum alternative allele frequency across datasets specified.<br />
                                        <b> altFreq@hgXX_XXXXX: </b>The alternative allele frequency at a dataset hgXX_XXXXX.<br />
                                        <strong>Note</strong>Variants without allele frequencies in reference databases are marked by '.' in the results. Moreover, Variants not EXISTING in reference database are marked by 'N'.
                                    </h4>

                                    <br />
                                    <img src="imgs/leaf.gif" width="10" height="8" /> <b>2. </b> Users can also provide local filters to filter variants. <tt> --local-filter</tt><br />
                                    <p><code>java -jar ./kggseq.jar  --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2   --local-filter path/to/file3  --rare-allele-freq 0.005 --local-filter-hard path/to/file4</code></p>

                                    Moreover, <tt>'--local-filter-hard' </tt>is use to filter <b>ALL</b>  existing variants in a database.

                                    <p>The <tt>'--local-filter'</tt> option sets local datasets for filterating. KGGSeq accepts local dataset files in the following format: each file has 5 columns [Chromosome, Physical Position, Reference Allele, Alternative Allele(s), and Frequency of alternative allele(s)], separated by tab character. KGGSeq provides a function to convert data in VCF format to that in the local dataset format:</p>
                                    <code>java -jar ./kggseq.jar --make-filter --vcf-file path/to/file --out test.var</code>

                                    An example file is like:
                                    <h4>

                                        Chrom	Position	Reference	Alternative	Freq<br />
                                        1	469	C	G	0.150<br />
                                        1	492	C	T	0.175<br />
                                        1	519	G	C	0.067<br />
                                        1	874290	-	0ACAGAG	0.809<br />
                                        1	875913	CAG	3	0.8431<br />
                                        <hr />
                                        <strong>Note</strong> The title line is needed and Freq is optional.
                                    </h4>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> <b>3. </b> An in-house VCF file can be used for allele frequency trimming: <tt>--local-filter-vcf </tt></p>
                                    <p><code>java -jar ./kggseq.jar  --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2   --local-filter-vcf path/to/file3  --rare-allele-freq 0.005 --local-filter-vcf-hard path/to/file4</code></p>
                                    <p> If the data were stored in different files chromosome by chromosome, you can use <tt>' _CHROM_'</tt> to denote the chromosome names [1...Y] or directly specify [1,2,X] in the file name so that KGGSeq can read data in multiple files in a run.</p>

                                    Similarly, <tt>'--local-filter-vcf-hard' </tt>is use to filter <b>ALL</b>  existing variants in a database.

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> <b>4. </b>An in-house VCF file without genotypes can also be used for allele frequency trimming: <tt>--local-filter-no-gty-vcf </tt></p>
                                    <p><code>java -jar ./kggseq.jar  --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2   --local-filter-no-gty-vcf path/to/file3  --rare-allele-freq 0.005 --local-filter-no-gty-vcf-hard path/to/file4</code></p>

                                    Similarly, <tt>'--local-filter-no-gty-vcf-hard' </tt>is use to filter <b>ALL</b>  existing variants in a database.


                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> <b>5. </b>Filter variants by minor allele frequency in the input sample: <tt>--filter-sample-maf-le</tt> or <tt>--filter-sample-maf-oe</tt></p>
                                    <p><code>java -jar ./kggseq.jar  --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2  --filter-sample-maf-le 0.01</code></p>

                                    <tt>--filter-sample-maf-le</tt>: Filter out variants with minor allele frequency less than OR equal to a cutoff (x) in the input samples.<br/>
                                    <tt>--filter-sample-maf-oe</tt>: Filter out variants with minor allele frequency over OR equal to a cutoff (x) in the input samples.<br/>



                                    <h3><a  id="VariantTypeAndRegionFiltering" >  Variant type, region and gene filtering</a> </h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Ingore insertion and deletion sequence variants (indels): <tt> --ignore-indel</tt></p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --ignore-indel</code></p>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" />  Accordingly, users can also  ask KGGSeq to  ingore Single nucleotide variants (SNVs) : <tt> --ignore-snv</tt></p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --ignore-snv</code></p>
                                    <p>By default, KGGSeq will process both SNV and Indel.
                                    </p>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Ignore some genomic regions: <tt>--regions-out</tt> </p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --regions-out chrX,chrY:1-10000</code>
                                        <img src="imgs/leaf.gif" width="10" height="8" />  Alternatively, one can also exclusively process sequence variants at some genome regions:
                                        <code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --regions-in chr2,chr4:21212-233454</code> </p>


                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Ignore variants within some genes: <tt>--genes-out</tt> </p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --genes-out TCF4,CNNM2,ANK3</code>
                                        <img src="imgs/leaf.gif" width="10" height="8" />  Alternatively, one can KEEP variants only within some genes:
                                        <code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --genes-in TCF4,CNNM2,ANK3</code> </p>


                                    <h3><a  id="SupperDuplicateRegionFiltering" > Super duplicate region filtering</a> </h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Filter out variants in super duplicate regions: <tt>--superdup-filter</tt> </p>
                                    <code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --superdup-filter</code>
<<<<<<< HEAD
                                    <p> KGGSeq can eliminate sequence variants within putative super-duplicate genomic regions defined in a dataset (genomicSuperDups) from <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/genomicSuperDups.sql" target="_blank">UCSC datasets </a>, which have higher genotyping error rate. (See<a href="http://blog.goldenhelix.com/?p=1153" target="_blank"> more </a> ).

                                    </p>
                                    <p><strong>Note</strong> Another option <tt>--superdup-annot</tt> allows you to just mark (but not to remove) the variants in the super duplicate regions. A column named <b>SuperDupKValue</b> (<a href="http://garfield.library.upenn.edu/classics1990/A1990CZ67100002.pdf" target="_blank">K-value calculated with Jukes-Cantor model</a>) will be added in the resulting dataset</p>
=======
                                    <p> KGGSeq can eliminate sequence variants within putative super-duplicate genomic regions defined in a dataset (genomicSuperDups) from <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/" target="_blank">UCSC datasets </a>, which have higher genotyping error rate. (See<a href="http://blog.goldenhelix.com/?p=1153" target="_blank"> more </a> ).

                                    </p>
                                    <p><strong>Note</strong> Another option <tt>--superdup-annot</tt> allows you to just mark (but not to remove) the variants in the super duplicate regions. A column named <b>SuperDupKValue</b> will be added in the resulting dataset</p>
>>>>>>> origin/master

                                    <h3><a  id="TooManyVarGeneFiltering" >Gene&#39;s Variant number</a> </h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Filter out genes with too many variants: <tt>--gene-var-filter</tt> </p>
                                    <code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 [options to filter out common and neutral variants] --gene-var-filter 4</code>
                                    After a number of hard-filtering by allele frequencies and pathogenic prediction, you may see some genes harboring candidate variants.  Empirically, these genes are usually not interesting and their mutations are often produced by some unknown factors such pseudogene  or platform bias.  As a rule of thumb, it is safe to set a cutoff 4 or more.

                                    <h3><a  id="PhenotypeRelevantFiltering" >Case-control Phenotype</a> </h3>
                                    <br />
                                    <img src="imgs/leaf.gif" width="10" height="8" /> Filter for case or control unique variants: <tt> --filter-model case-unique</tt> or <tt> --filter-model control-unique</tt>
                                    <code>java -jar ./kggseq.jar --buildver hg19  --vcf-file path/to/file1 --ped-file path/to/file2 --filter-model case-unique </code>
                                    Case unique: extract variants with alternative alleles only present among affected individuals.<br />
                                    Control unique: extract variants with alternative alleles only present among unaffected individuals.
                                    <br />	<br />
                                    <p>
                                        <strong>Note</strong>	The <tt>--filter-model</tt>	only works for inputs specified by <tt>--vcf-file</tt> .<br />
                                        <strong>Note</strong> This function can work with  options to flexibly set the sample size and missingness of genotypes, detalied in the <a href="#VariantQC" target="_self">Variant QC</a> part.<br />

                                    </p>
                                    <p>&nbsp;                                </p>

                                </td>
                            </tr>

                            <tr>
                                <td>
                                    <h2><a id="Annotation" >Annotation</a></h2>
                                    <h3> <a name="dbSNPRSID" > dbSNP RSID </a></h3>
                                    <br />
                                    Assign dbSNP RSID to variants:<tt>--rsid</tt><br />
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2   --rsid</code>
                                    KGGSeq can update the variants' <a href="http://www.ncbi.nlm.nih.gov/projects/SNP/" target="_blank">dbSNP rsID </a>(if available)  according to their genomic coordinates. The resource data used are from
                                    <b>SNPChrPosOnRef</b> (<a href="ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b142_GRCh37p13/database/organism_data/b142_SNPChrPosOnRef_105.bcp.gz">hg19</a>/<a href="ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b142_GRCh38/database/organism_data/b142_SNPChrPosOnRef_106.bcp.gz">hg38</a>) [<a href="http://www.ncbi.nlm.nih.gov/SNP/snp_db_table_description.cgi?t=SNPChrPosOnRef">help</a>]. 
                                    <br/>
                                    <h3><a name="splicingAnnote" >Alternative splicing </a></h3>
                                    Potential of altering splicing<tt>--scsnv-annot</tt><br />

                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --scsnv-annot</code>
<<<<<<< HEAD
                                    Annotate a SNV with predicted potential of altering splicing from <a href="https://sites.google.com/site/jpopgen/dbNSFP" target="_blank"> dbscSNV </a>, which include all possible human SNVs within splicing consensus regions &#40;-3 to +8 at the 5&#39;  splice site and -12 to +2 at the 3&#39; splice site&#41;.
=======
                                    Annotate a SNV with predicted potential of altering splicing from <a href="https://sites.google.com/site/jpopgen/dbNSFP" target="_blank"> dbscSNV </a>, which include all possible human SNVs within splicing consensus regions &#40;-3 to +8 at the 5  splice site and -12 to +2 at the 3&#39; splice site&#41;.
>>>>>>> origin/master
                                    <br/>                                    The command will append a field to the output file. 
                                    <h4>
                                        ada_score@dbScSNV: The prediction score by ADA approach which combines multiple existing tools. See more in <a  href="http://nar.oxfordjournals.org/content/42/22/13534" target="_blank"> Jian et al (2014)</a> <br />
                                    </h4>

                                    <h3> <a name="dgvcnvAnnote" >Structure variation</a></h3>
                                    Map a variant against known structure variation <tt>--dgv-cnv-annot</tt><br />

                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --dgv-cnv-annot</code>
                                    Annotate a variant whether it is within a known insertion or deletion registered in the <a href="http://dgv.tcag.ca/dgv/app/home" target="_blank"> Database of Genomic Variants </a>, which will help exclude candidates which are not interesting.
                                    <br/>                                    The command will append a field to the output file. 
                                    <h4>                                       
                                        <b>DGVIDs:</b> the ID in the Database of Genomic Variants.<br />
                                        <b>CNVSampleSize:</b> the size of sample, used to derive the structure. <br />
                                        <b>LossCNV:</b> Number of subjects having loss of copy number variation.<br />
                                        <b>GainCNV:</b> Number of subjects having excessive copy number variation.<br />
                                    </h4>



                                    <h3><a name="SharedProteinProteinInteractions" >Shared protein-protein interactions (PPIs)</a></h3>
                                    <br/> Examine whether variants share the same PPIs with some genes of interest: <tt>--candi-list gene1,gene2,...,geneN --ppi-annot ppiDatabase</tt>
                                    <br/>


                                    <p><code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --candi-list gene1,gene2,...,geneN --ppi-annot string --ppi-depth 1</code></p>
                                    <p><strong>Note</strong> Currently, the only supported PPI database is <a href="http://string-db.org/">STRING</a> (i.e., <tt>string</tt>).</p>
                                    <p><strong>Note</strong> Instead of specifying the gene list using the <tt>--candi-list</tt> option, users can specify a file containing the list using the <tt>--candi-file</tt> option.</p>
<<<<<<< HEAD
                                    In the file containing candidate genes, each row can only have one gene. The format is like the following:
                                    <h4>
                                        GeneSymbol1<br />
                                        GeneSymbol2<br />
                                        GeneSymbol3<br />
                                        GeneSymbol4<br />
                                    </h4>
=======
>>>>>>> origin/master
                                    <p><strong>Note</strong> The maximum distance of a PPI between a specifified candidate  genes and a gene containing the potential causal variants can be adjusted using the <tt>--ppi-depth</tt> option. The defaul is <tt>--ppi-depth 1</tt>.<br/>
                                    </p>

                                    The option will append the following fields to the output file:
                                    <h4>
                                        IsWithinCandidateGene:                    Whether the variant is within a candidate gene<br />
                                        PPI:                                      PPI shared by at least one gene of interest and the gene
                                        containing the variant
                                    </h4>
                                    <h3><a name="SharedGeneSets" >Shared genesets</a></h3>
                                    <br />
                                    Examine whether variants share the same genesets with some genes of interest: <tt>--candi-list gene1,gene2,...,geneN --geneset-annot GenesetDatabase</tt>
                                    <br/>
                                    <p><code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --candi-list gene1,gene2,...,geneN --geneset-annot GeneSetDatabase</code></p>

                                    <p>Currently, users can choose between 5 geneset databases from <a href="http://www.broadinstitute.org/gsea/msigdb/collections.jsp">GSEA</a>. The following are brief descriptions of the 5 datasets COPIED from the <a href="http://www.broadinstitute.org/gsea/msigdb/collections.jsp" target="_blank">page</a> of GSEA.<br/>
                                        <table width="90%" class="list">
                                            <tr>
                                                <th width="19%" scope="col">GeneSetDatabase</th>
                                                <th width="66%" scope="col">Description</th>
                                                <th width="15%" scope="col">Version</th>
                                            </tr>
                                            <tr>
                                                <td>cano</td>
                                                <td>Gene sets from the geneset databases. Usually, these gene sets are canonical representations of a biological process compiled by domain experts. (1452 gene sets)</td>
                                                <td> MSigDB 3.1</td>
                                            </tr>
                                            <tr>
                                                <td> cura </td>
                                                <td>Curated gene sets: Gene sets collected from various sources such as online geneset databases, publications in PubMed, and knowledge of domain experts. The gene set page for each gene set lists its source. (4850 gene sets)</td>
                                                <td> MSigDB 3.1</td>
                                            </tr>
                                            <tr>
                                                <td> onco </td>
                                                <td> Oncogenic signatures: Gene sets represent signatures of cellular genesets which are often dis-regulated in cancer. The majority of signatures were generated directly from microarray data from NCBI GEO or from internal unpublished profiling experiments which involved perturbation of known cancer genes. In addition, a small number of oncogenic signatures were curated from scientific publications. (189 gene sets)</td>
                                                <td> MSigDB 3.1</td>
                                            </tr>
                                            <tr>
                                                <td> cmop </td>
                                                <td>Computational gene sets: Computational gene sets defined by mining large collections of cancer-oriented microarray data. (858 gene sets)</td>
                                                <td> MSigDB 3.1</td>
                                            </tr>
                                            <tr>
                                                <td> onto </td>
                                                <td>GO gene sets Gene sets are named by GO term and contain genes annotated by that term. GSEA users: Gene set enrichment analysis identifies gene sets consisting of co-regulated genes; GO gene sets are based on ontologies and do not necessarily comprise co-regulated genes. (1454 gene sets)</td>
                                                <td> MSigDB 3.1</td>
                                            </tr>
                                        </table>

                                    </p>
                                    <p><strong>Note</strong> The <tt>--candi-list</tt> and <tt>--candi-file</tt> options can be shared with <tt>--ppi-annot</tt>. </p>

                                    The option will append the following fields to the output file:
                                    <h4>
                                        IsWithinCandidateGene:                    Whether the variant is within a candidate gene<br />
                                        SharedGeneSet:                            GeneSet shared by at least one gene of interest and the gene
                                        containing the variant
                                    </h4>


                                    <h3><a name="MousePhenotype">Mouse Phenotype</a></h3>
                                    <br/>
                                    Annotate the genes with known mouse phenotypes as reference: --mouse-pheno.  
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --mouse-pheno </code>
                                    The genes harboring interested mutations will be annotated by known mouse phenotypes of ortholog genes in the file of <a href="ftp://ftp.informatics.jax.org/pub/reports/HMD_HumanPhenotype.rpt" target="_blank" >Mouse/Human Orthology with Phenotype Annotations</a>  from <a href="http://www.informatics.jax.org/" target="_blank" >Mouse Genome Informatics(MGI)</a> and in the file of <a href=" ftp://ftp.ebi.ac.uk/pub/databases/impc/release-4.0/csv/ALL_genotype_phenotype.csv.gz" target="_blank" > phenotype-genotype </a> from  <a href="https://www.mousephenotype.org/" target="_blank" >International Mouse Phenotyping Consortium (IMPC)</a> <br/>
                                    An column will be appended in the main output file:<br/>
                                    <h4>
                                        MousePhenotype (MGI):mouse phenotypes of ortholog genes. Multiple phenotypes are separated by '||'. <br/>
                                        MousePhenotype (IMPC):mouse phenotypes of ortholog genes. Multiple phenotypes are separated by '||'.
                                    </h4>
                                    <br />
                                    <h3><a name="zebrafihserPhenotype">Zebrafish Phenotype</a></h3>
                                    <br/>
                                    Annotate the genes with known zebrafish phenotypes as reference: --zebrafish-pheno.  
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --zebrafish-pheno </code>
                                    The genes harboring interested mutations will be annotated by known zebrafish phenotypes of ortholog genes in the file of <a href="http://zfin.org/" target="_blank" >Phenotypic Zebrafish genes with Human Orthology </a>  from <a href="http://zfin.org/" target="_blank" >The Zebrafish Model Organism Database</a>. <br/>
                                    An column will be appended in the main output file:<br/>
                                    <h4>
                                        ZebrafishPhenotype :Affected structure or process subterm after the knock-out of the human ortholog genes in Zebrafish.
                                    </h4>
                                    <br />
                                    <h3><a name="ddditem">DDD Study</a></h3>
                                    <br/>
                                    Annotate by disease names in Deciphering Developmental Disorders (DDD) study: --ddd-annot.  
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --ddd-annot </code>
<<<<<<< HEAD
                                    The genes harboring interested mutations will be annotated by the documented disease names in <a href="http://www.ebi.ac.uk/gene2phenotype/gene2phenotype-webcode/cgi-bin/handler.cgi?show_downloads=all#" target="_blank" >Gene-based genotype to phenotype</a> of <a href="http://www.ddduk.org/" target="_blank" > Deciphering Developmental Disorders (DDD) study </a>. <br/>
=======
                                    The genes harboring interested mutations will be annotated by the documented disease names in <a href=" https://decipher.sanger.ac.uk/files/ddd/ddg2p.zip" target="_blank" >Gene-based genotype to phenotype</a> of <a href="http://www.ddduk.org/" target="_blank" > Deciphering Developmental Disorders (DDD) study </a>. <br/>
>>>>>>> origin/master
                                    An column will be appended in the main output file:<br/>
                                    <h4>
                                        DDDPhenotype : confirmation status;disease;pubmed_ids
                                    </h4>
                                    <br />
                                    <h3><a name="PubMedTextMining" >PubMed literature</a></h3>
                                    <br/>
                                    Assign PubMed ID to a gene:</a><tt>--phenotype-term searchTerm</tt> and <tt>--pubmed-mining</tt><br />
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --phenotype-term search+Term1,search+Term2 --pubmed-mining</code>
                                    KGGSeq can mine the titles and abstracts of published papers in PubMed via the NCBI <a href="http://www.ncbi.nlm.nih.gov/books/NBK25500/" target="_blank" >E-utilities</a> to find a co-mention with the searched term(s) of interest (i.e., disease) and the genes/cytogeneic regions in which the
                                    variants are located. The option <tt>--pubmed-mining</tt> will append the following two fields to the output file.
                                    <h4>
                                        PubMedIDIdeogram :                        PubMed ID of articles in which the term and the cytogeneic
                                        position of the variant are co-mentioned<br />
                                        PubMedIDGene :                            PubMed ID of articles in which the term and the gene
                                        containing the variant are co-mentioned</h4>
                                    <br />
                                    <strong>IMPORTANT</strong>  Please use + instead of blank characters within a search term.<br/>

                                    <h3><a name="OMIMTextMining" >OMIM</a></h3>

                                    Assign OMIM term to a gene <tt>--omim-annot</tt><br />
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --omim-annot</code>
                                    KGGSeq can extract all available disorder names linked to a gene in which the
                                    variants are located from the OMIM dataset <a href="http://www.omim.org/downloads" target="_blank" >morbidmap</a>. The command will append two extra fields to the output file:
                                    <h4>
                                        DiseaseName(s)MIMid :                        Disorder, &lt;disorder MIM no.&gt; (&lt;phene mapping key&gt;)<br />
                                        Phenotype mapping method &lt;phene mapping key&gt;:<br />
                                        1 - the disorder is placed on the map based on its association with
                                        a gene, but the underlying defect is not known.<br />
                                        2 - the disorder has been placed on the map by linkage; no mutation has
                                        been found.<br />
                                        3 - the molecular basis for the disorder is known; a mutation has been
                                        found in the gene.<br />
                                        4 - a contiguous gene deletion or duplication syndrome, multiple genes
                                        are deleted or duplicated causing the phenotype.<br />

                                        GeneMIMid :                                                    Gene/locus MIM no.
                                    </h4>


                                    <h3><a name="COSMICMining" >COSMIC cancer information</a></h3>
                                    Assign COSMIC cancer annotation to a gene: </a><tt>--cosmic-annot</tt><br />
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --indiv-pair NONTUMOR.1:TUMOR.1,NONTUMOR.2:TUMOR.2 --genotype-filter 8 --cosmic-annot</code>
                                    KGGSeq can retrieve the confirmed somatic mutation information registered for variants and genes in the <a href="http://cancer.sanger.ac.uk/cancergenome/projects/cosmic/" target="_blank" >COSMIC</a> database.
                                    The command will append a field to the output file. e.g.,
                                    <h4>
                                        COSMICCancerInfo : known cancer and occurrence time of the somatic mutation in COSMIC dataset. e.g., {serous_carcinoma=3, astrocytoma_Grade_IV=1}<br />
                                        This somatic mutation occurred 3 times in samples of 'serous_carcinoma' and once in samples of 'astrocytoma_Grade_IV'.
                                    </h4>

                                    <p>&nbsp;</p>
                                    <h2><a id="PredictionVar" >Prediction at variant</a></h2>
                                    KGGSeq provides a series of methods to combine existing functional scores of DNA variants from multiple algorithms (i.e.,  <a href="http://sift.jcvi.org/" target="_blank">SIFT</a>, <a href="http://genetics.bwh.harvard.edu/pph2/" target="_blank">PolyPhen2</a>, <a href="https://www.sanger.ac.uk/resources/software/gwava/" target="_blank">GWAVA</a>  and <a href="http://cadd.gs.washington.edu/ " target="_blank">CADD</a>) for a more accurate pathogenic prediction at coding and non-coding variants.

                                    <h3><a name="PredictingMendelianDiseaseCausingVariants" >Predicting Mendelian disease-causing variants</a></h3>
                                    <p><br/>
                                        Predict Mendelian disease-causing variants: <tt>--db-score dbnsfp --mendel-causing-predict</tt><br />
<<<<<<< HEAD
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --db-score dbnsfp --mendel-causing-predict best --filter-nondisease-variant</code>
                                        Use functional or deleterious scores (listed below) collected in <a href="https://sites.google.com/site/jpopgen/dbNSFP" target="new">dbNSFP v3.0+</a> database to RE-predict whether a nonsynonymous single nucleotide variant (SNV) will potentially be Mendelian disease causal or not by Logistic regression model (See methods in our paper, <a href="http://www.plosgenetics.org/article/info%3Adoi/10.1371/journal.pgen.1003143" target="_blank">Li et al. PLoS Genet. 2013 Jan;9(1):e1003143</a>). By default, KGGSeq try the top 10 subsets of the scores (according to their area under the curves (AUC) of receiver operating characteristic (ROC)) for a combinatorial prediction until a positive prediction occurs. This is a way to relief influence of missing values. Users can try more subsets by a number on the option, say <tt>--mendel-causing-predict best30</tt>. However, a larger number will increase the false positive rate. In conjunction with the prediction model,  the <tt>--filter-nondisease-variant</tt> tag will be used to filter out the  nondisease-causative variants predicted by Logistic regression model (See methods in <a href="http://www.plosgenetics.org/article/info%3Adoi/10.1371/journal.pgen.1003143" target="_blank">Li et al. PLoS Genet. 2013 Jan;9(1):e1003143</a>). .Note that this trying will increase the false positive rate although it decreases the false negative rate. So you may need more information to justify the involved gene(s).</p>
=======
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --db-score dbnsfp --mendel-causing-predict all --filter-nondisease-variant</code>
                                        Use functional or deleterious scores (listed below) collected in <a href="https://sites.google.com/site/jpopgen/dbNSFP" target="new">dbNSFP v3.0+</a> database to RE-predict whether a nonsynonymous single nucleotide variant (SNV) will potentially be Mendelian disease causal or not by Logistic regression model (See methods in our paper, <a href="http://www.plosgenetics.org/article/info%3Adoi/10.1371/journal.pgen.1003143" target="_blank">Li et al. PLoS Genet. 2013 Jan;9(1):e1003143</a>). By default, KGGSeq tries all subsets of the scores (at least two) for a combinatorial prediction and finally present a prediction which can pass the threshold to classify the SNV as a potentially disease causal or no.  In conjunction with the prediction model,  the <tt>--filter-nondisease-variant</tt> tag will be used to filter out the  nondisease-causative variants predicted by Logistic regression model (See methods in <a href="http://www.plosgenetics.org/article/info%3Adoi/10.1371/journal.pgen.1003143" target="_blank">Li et al. PLoS Genet. 2013 Jan;9(1):e1003143</a>). .Note that this trying will increase the false positive rate although it decreases the false negative rate. So you may need more information to justify the involved gene(s).</p>
>>>>>>> origin/master
                                    <img src="imgs/mendelAUC.png" width="800" height="400" /><br />
                                    <b>Figure:</b> Receiver operating characteristic (ROC) and Precision recall(PR) curves and  area under the curves (AUC) of individual scores and combined score by Logistic regression model<br />
                                    Note: this is an updated plot of Figure 1 in <a href="http://www.plosgenetics.org/article/info%3Adoi/10.1371/journal.pgen.1003143" target="_blank" >our paper</a> . Here we combined more individual functional impact scores, which is the only difference.

                                    <br />
                                    <p> On the other hand, one can FIX the prediction using a specified subset or full set of the 5 impact scores by option like <tt>--mendel-causing-predict 4,6,7</tt><br />

                                        The coding for the functional or deleterious scores used in'--mendel-causing-predict'options is listed below:                                    </p>
                                    <table  class="list">
                                        <tr>
                                            <td width="52" valign="top"><b>Coding</b></td>
                                            <td width="114" valign="top"><b>Method</b></td>
                                            <td    valign="top"><b>Description (Mostly copied from <a href="https://sites.google.com/site/jpopgen/dbNSFP" target="_blank">  dbNSFP</a>)</b></td>
                                        </tr>                                      
                                        <tr>
                                            <td width="52" valign="top">1</td>
                                            <td width="114" valign="top"><a href="http://www.ncbi.nlm.nih.gov/pubmed?term=22689647" target="_blank">SIFT</a></td>
                                            <td  valign="top">SIFT uses the 'Sorting Tolerant From Intolerant' (SIFT) algorithm to predict whether a single amino acid substitution affects protein function or not, based on the assumption that important amino acids in a protein sequence should be conserved throughout evolution and substitutions at highly conserved sites are expected to affect protein function.A small scoreindicates a high chance for a substitutionto damage the protein function. </td>
                                        </tr>

                                        <tr>
                                            <td width="52" valign="top">2</td>
                                            <td width="114" valign="top"><a href="http://www.ncbi.nlm.nih.gov/pubmed?term=20354512" target="new">Polyphen2_HDIV</a></td>
                                            <td  valign="top">Polyphen2 score based on HumDiv, i.e. hdiv_prob.
                                                The score ranges from 0 to 1, and the corresponding prediction is "probably damaging"
                                                if it is in [0.957,1]; "possibly damaging" if it is in [0.453,0.956]; "benign"
                                                if it is in [0,0.452]. Score cutoff for binary classification is 0.5, i.e.
                                                the prediction is "neutral" if the score is smaller than 0.5 and "deleterious" if the
                                                score is larger than 0.5. Multiple entries separated by ";"</td>
                                        </tr>
                                        <tr>
                                            <td width="52" valign="top">3</td>
                                            <td width="114" valign="top"><a href="http://www.ncbi.nlm.nih.gov/pubmed?term=20354512" target="_blank">Polyphen2_HVAR</a></td>
                                            <td  valign="top">Polyphen2 predicts the possible impact of an amino acid substitution on the structure and function of a human protein using straightforward physical and comparative considerations by an iterative greedy algorithm. In the present study, we use the original scores generated by the HumVar (instead ofHumDiv) trained model as it is preferred for the diagnosis of Mendelian diseases. The scores range from 0 to 1. A substitution with larger score has a higher possibility to damage the protein function.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">4</td>
                                            <td valign="top"><a href="http://www.ncbi.nlm.nih.gov/pubmed?term=19602639" target="_blank">LRT</a></td>
                                            <td valign="top">LRT employed a likelihood ratio test to assess variant deleteriousness based on a comparative genomics data set of 32 vertebrate species. The identified deleterious mutations could disrupt highly conserved amino acids within protein-coding sequences, which are likely to be unconditionally deleterious.The scores range from 0 to 1. A larger score indicates a larger deleterious effect.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">5</td>
                                            <td valign="top"><a href="http://www.ncbi.nlm.nih.gov/pubmed?term=20676075" target="_blank">MutationTaster</a></td>
                                            <td valign="top">MutationTaster assesses the impact of the disease-causing potential of a sequence variant by a naive Bayes classifier using multiple resources such as evolutionary conservation, splice-site changes, loss of protein features and changes that might affect mRNA level. The scores range from 0 to 1. The larger score suggests a higher probability to cause a human disease.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">6</td>
                                            <td valign="top"><a href="http://mutationassessor.org/" target="new">MutationAssessor</a></td>
                                            <td valign="top">MutationAssessor "functional impact of a variant :
                                                predicted functional (high, medium), predicted non-functional (low, neutral)"
                                                Please refer to Reva et al. Nucl. Acids Res. (2011) 39(17):e118 for details</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">7</td>
                                            <td valign="top"><a href="http://fathmm.biocompute.org.uk/" target="new">FATHMM</a></td>
                                            <td valign="top">FATHMM default score (weighted for human inherited-disease mutations with Disease Ontology); If a score is smaller than -1.5 the corresponding NS is predicted as &quot;D(AMAGING)&quot;; otherwise it is predicted as &quot;T(OLERATED)&quot;. If there's more than one scores associated with the same NS due to isoforms, the smallest score (most damaging) was used. Please refer to Shihab et al Hum. Mut. (2013) 34(1):57-65 for details</td>
                                        </tr>

                                        <tr>
                                            <td valign="top">8</td>
                                            <td valign="top"><a href="http://karchinlab.org/apps/appVest.html" target="new">VEST3</a></td>
                                            <td valign="top">VEST 3.0 score. Score ranges from 0 to 1. The larger the score the more likely
                                                the mutation may cause functional change. In case there are multiple scores for the
                                                same variant, the largest score (most damaging) is presented. Please refer to
                                                Carter et al., (2013) BMC Genomics. 14(3) 1-16 for details.
                                                Please note this score is free for non-commercial use. For more details please refer to
                                                http://wiki.chasmsoftware.org/index.php/SoftwareLicense. Commercial users should contact
                                                the Johns Hopkins Technology Transfer office.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">9</td>
                                            <td valign="top"><a href="http://provean.jcvi.org/index.php" target="new">PROVEAN</a></td>
                                            <td valign="top">PROVEAN score (PROVEANori). Scores range from -14 to 14. The smaller the score
                                                the more likely the SNP has damaging effect..</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">10</td>
                                            <td valign="top"><a href="http://cadd.gs.washington.edu/" target="new">CADD</a></td>
                                            <td valign="top">Combined Annotation Dependent Depletion (CADD) score for funtional prediction of a SNP. Please refer to <a href="http://www.nature.com/ng/journal/v46/n3/full/ng.2892.html" target="new"> Kircher et al.
                                                    (2014) Nature Genetics 46(3):310-5  </a> for details. The larger the score the more likely
                                                the SNP has damaging effect.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">11</td>
                                            <td valign="top"><a href="http://mendel.stanford.edu/SidowLab/downloads/gerp/index.html" target="new">GERP++_NR</a></td>
                                            <td valign="top">Neutral rate</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">12</td>
                                            <td valign="top"><a href="http://mendel.stanford.edu/SidowLab/downloads/gerp/index.html" target="new">GERP++_RS</a></td>
                                            <td valign="top">RS score, the larger the score, the more conserved the site</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">13</td>
                                            <td valign="top"><a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/" target="_blank">phastCons7way_vertebrate</a></td>
                                            <td valign="top">phastCons conservation score based on the multiple alignments of 7 vertebrate genomes (including human). The larger the score, the more conserved the site. Scores range from 0 to 1. </td>
                                        </tr>
                                        <tr>
                                            <td valign="top">14</td>
                                            <td valign="top"><a href="http://bioinformatics.oxfordjournals.org/content/25/12/i54.short" target="new">SiPhy_29way_logOdds</a></td>
                                            <td valign="top">SiPhy score based on 29 mammals genomes. The larger the score,he more conserved the site. Scores range from 0 to 37.9718 in dbNSFP.</td>
                                        </tr>
                                    </table>

                                    <strong>Note</strong> Predictions at variants with missing scores at specified methods will be ignored!<br/>
                                    <strong>Note</strong>  If you use this feature in any published work, please cite the following manuscript:
                                    <b>Li et al. Predicting Mendelian disease-causing non-synonymous single nucleotide variants in exome sequencing studies. <span title="PLoS genetics"><a href="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1003143" target="_blank">PLoS Genet</a></span><a href="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1003143">. 2013 Jan;9(1):e1003143</a></b>


                                    <p>The option will append the following fields to the output file: </p>
                                    <h4>

                                        ...<br />
                                        Polyphen2_HDIV_pred:                                 Polyphen2 prediction based on HumDiv, "D" ("porobably damaging"),
                                        "P" ("possibly damaging") and "B" ("benign"). Multiple entries separated by ";"<br />

                                        Polyphen2_HVAR_pred: Polyphen2 prediction based on HumVar, "D" ("porobably damaging"),
                                        "P" ("possibly damaging") and "B" ("benign"). Multiple entries separated by ";"                                       <br />
                                        LRT_pred:                                 Classification using LRT (D = deleterious, N = neutral,
                                        or U = unknown)<br />
                                        MutationTaster_pred:                      Classification using MutationTaster (A =
                                        disease_causing_automatic, D = disease_causing, N =
                                        polymorphism, or P = polymorphism_automatic)<br />
                                        MutationAssessor_pred: MutationAssessor "functional impact of a variant :
                                        predicted functional (high, medium), predicted non-functional (low, neutral)"<br />
<<<<<<< HEAD
                                        MetaSVM_pred: Prediction of an SVM based ensemble prediction score by Dong et al. [(2015) Human Molecular Genetics 24(8):2125-2137]."T(olerated)" or"D(amaging)". <br />
                                        MetaLR_pred: Prediction of a MetaLR based ensemble prediction score by Dong et al. [(2015) Human Molecular Genetics 24(8):2125-2137]."T(olerated)" or "D(amaging)". <br />
                                        clinvar_clnsig: clinical significance as to the clinvar data set; 2 - Benign, 3 - Likely benign, 4 - Likely pathogenic, 5 - Pathogenic, 6 - drug response, 7 - histocompatibility. A negative score means the the score is for the ref allele
                                        clinvar_trait: the trait/disease the clinvar_clnsig referring to (Please see more description about the attributes at <a href="https://drive.google.com/file/d/0B60wROKy6OqcWllRZjZXM2Vuc2c/view" target="_blank">  https://drive.google.com/file/d/0B60wROKy6OqcWllRZjZXM2Vuc2c/view </a><br />
=======
>>>>>>> origin/master
                                        [...]<br/>
                                        DiseaseCausalProb_ExoVarTrainedModel:     Conditional probability of being Mendelian disease-causing
                                        given the above prediction scores under a logistic regression
                                        model trained by our dataset<a href="http://grass.cgs.hku.hk/limx/kggseq/download/ExoVar.xls" target="_blank"> ExoVar</a>.<br />
                                        IsRareDiseaseCausal_ExoVarTrainedModel:   Classification using the logistic regression model
                                        (Y = disease-causing or N = neutral)<br />
                                        BestCombinedTools:OptimalCutoff:TP:TN : The  subset of original prediction tools (out of the 13 tools) used for the combined prediction by our Logistic Regression model which have the largest posterior probability among all possible combinatorial subsets: the cutoff leads to the maximal Matthews correlation coefficient (MCC): the corresponding true positive and true negative at the maximal MCC.
                                    </h4>
<<<<<<< HEAD
                                    <strong>Note</strong> By default the option <tt>--mendel-causing-predict best </tt> is always switched on  unless other prediction types are specified.
=======
                                    <strong>Note</strong> By default the option <tt>--mendel-causing-predict all </tt> is always switched on  unless other prediction types are specified.
>>>>>>> origin/master
                                    <h3><a name="PredictingCancerCausingVariants" >Predicting driver mutations of cancers</a></h3>
                                    <p><br/>
                                        Predict non-synonyms cancer-driver somatic-mutations of cancers by an integrative approach based on : <tt>--db-score dbnsfp --cancer-driver-predict</tt><br />
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --indiv-pair NONTUMOR.1:TUMOR.1,NONTUMOR.2:TUMOR.2 --genotype-filter 8 --cosmic-annot --db-score dbnsfp --cancer-driver-predict --filter-nondisease-variant</code>
                                        A driver mutation is defined as the mutation that leads to selective advantage to a cancer clone by either increasing the cancer cells' survival or reproduction. Use the above mentioned 14 functional or deleterious scores collected in dbNSFP (dbNSFP v3.0+) database to jointly predict whether a nonsynonymous single nucleotide variant (SNV) will potentially be the driver mutations by a standard random forest model trained under <a href="http://www.sanger.ac.uk/genetics/CGP/cosmic/" target="new">COSMIC</a> database. After the prediction, the <tt>--filter-nondisease-variant</tt> tag can be used to filter out all predicted non-driver missense variants.</p>
                                    <img src="imgs/cancerdriverpredic.png" width="735" height="402" /><br />
                                    <b>Table:</b> The area under the curves (AUC) of receiver operating characteristic (ROC) and Precision recall(PR) curves  and Matthews correlation coefficient (MCC)  of individual scores and combined score by the random forest model and Logistic regression model in a 10-fold cross-validation. As shown in the above table, the combined prediction by the random forest outperformed that by the Logistic regression.<br />

                                    <p>The option will append the following fields to the output file: </p>
                                    <h4>

                                        ...<br />                                        
                                        <b>IsCancerDriver_COSMICTrainedModel</b>:   Classification results by the random forest model
                                        (Y = cancer-driver or N = passenger)<br />
                                        <b>RandomForestScore</b>:Proportion of positive predictions in the random forest model in the benchmark dataset. It ranges from 0 to 1. The larger value, the more confident a positive prediction (Y).
                                        <br />
                                    </h4>

                                    <p>These predicted cancer-driver somatic mutations are evaluated by our statistical approach. </p>
                                    <h4>Kwan JS, Ng OL, Sham PC and Li MX.  Functional prediction of driver mutations in cancer genomes. (In preparation)
                                        <br/>
                                        <b>Benchmark dataset:</b> This dataset is composed of 3,479 cancer-causing somatic nsSNVs that occurred at least twice in the Catalogue of Somatic Mutations in Cancer (COSMIC) database and located in genes from the Cancer Gene Census (CGC), which are treated as positive control variants, together with an equal number of somatic mutations randomly sampled without replacement from a set of neutral mutations (i.e., mutations occurred only once in COSMIC and not  located in CGC genes), which are treated as negative control variants.
                                    </h4>
                                    <br />


                                    <h3><a name="PredictingComplexDiseaseCausingVariants" >Predicting complex disease-causing variants</a></h3>
                                    <p>
<<<<<<< HEAD
                                        Predict pathogenic potential of complex diseases at non-coding variants by different models.
                                    </p>
                                    <p>
                                        <img src="imgs/leaf.gif" width="10" height="8" /> <b>Gene feature specific model</b> <tt>--db-score dbncfp_known [or dbncfp_all] --ldgf-func-predict </tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --db-score dbncfp_known --ldgf-func-predict --buildver hg19</code>
                                    </p>
                                    <p>
                                        1) Annotate non-coding variants by 10 existing functional prediction scores (See below table) in a known variant dataset and 2) jointly predict the functions importance by a new statistical model (linkage disequilibrium expanded and gene feature specific model). The known dataset contains around 110 million variants from multiple high throughput sequencing resources, including <a href="http://www.1000genomes.org/" target="_blank">the 1000 Genomes Project Phase 3</a>, <a href="http://www.uk10k.org/" target="_blank">UK10K Project</a>, <a href="http://evs.gs.washington.edu/EVS/" target="_blank">ESP6500 Project</a> and <a href="http://exac.broadinstitute.org/" target="_blank">ExAC database</a>. The joint prediction also uses over 700 epigenomic markers for open chromatin, transcription factor binding, histone modification, RNA polymerase binding, and DNA methylation. 
                                        <br/>
                                        <br/>
                                        If you have a lot of variants without annotation of existing tools, you can use the whole genome annotation dataset, <tt>--db-score dbncfp_all</tt>. However, you have to manually download the file from <a href="ftp://jjwanglab.org/MX/score/ANN/" target="_blank">ftp://jjwanglab.org/MX/score/ANN/hgXX_all_SNV_dbNCFP.gz </a> by efficient tools (e.g. FileZilla) and put it into the folder of KGGSeqFolder/resources/hgXX/hgXX_all_SNV_dbNCFP.gz.
=======
                                        Predict pathogenicity of complex diseases at non-coding variants by two models, gene feature specific model and cell-type specific model.
                                    </p>
                                    <p>
                                        <img src="imgs/leaf.gif" width="10" height="8" /> <b>Gene feature specific model</b>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --db-score noncoding --complex-pathogenic-predict --buildver hg19</code>
                                    </p>
                                    <p>
                                        Rapid technology developments have dramatically reduced the cost of next-generation sequencing, which has accelerated identification of single nucleotide variants (SNVs) in the human genome. Unlike coding variants which could be easily detected the biological consequences, non-coding variants are very hard to observe the true functionality. In this respect, we proposed the method which is specifically designed for non-coding variants prediction and prioritization. In this model, we incorporate features from 11 existing functional prediction scores (See below table) and over 700 epigenomic markers for open chromatin, transcription factor binding, histone modification, RNA polymerase binding, and DNA methylation, to build gene feature specific model for predicting the pathogenicity of each non-coding variant.
                                        <br/>

>>>>>>> origin/master
                                        <p>  The existing functional prediction scores at non-coding variants are listed below:                                    </p>
                                        <table class="list">
                                            <tbody>
                                                <tr>
                                                    <td width="52" valign="top"><b>Coding</b></td>
                                                    <td width="114" valign="top"><b>Method</b></td>
                                                    <td valign="top"><b>Description</b></td>
                                                </tr>
                                                <tr>
<<<<<<< HEAD
                                                    <td valign="top">1</td>
                                                    <td valign="top"><a href="http://www.nature.com/nmeth/journal/v11/n3/abs/nmeth.2832.html" target="_blank">GWAVA_Region</a></td>
                                                    <td valign="top">GWAVA uses the random forest algorithm to build three classifiers using all available annotations to discriminate between the disease variants and variants from each of the three control sets. This control set first was composed of all 1KG variants in the 1 kb surrounding each of the HGMD variants.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">2</td>
                                                    <td valign="top"><a href="http://www.nature.com/nmeth/journal/v11/n3/abs/nmeth.2832.html" target="_blank">GWAVA_TSS</a></td>
                                                    <td valign="top">GWAVA uses the random forest algorithm to build three classifiers using all available annotations to discriminate between the disease variants and variants from each of the three control sets. This control set first was matched for distance to the nearest TSS genome-wide.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">3</td>
                                                    <td valign="top"><a href="http://www.nature.com/nmeth/journal/v11/n3/abs/nmeth.2832.html" target="_blank">GWAVA_Unmatched</a></td>
                                                    <td valign="top">GWAVA uses the random forest algorithm to build three classifiers using all available annotations to discriminate between the disease variants and variants from each of the three control sets. This control set first was constructed from a random selection of SNVs from across the genome in order to sample overall background.</td>
                                                </tr>
                                                <tr>
                                                    <td width="52" valign="top">4</td>
                                                    <td width="114" valign="top"><a href="http://www.nature.com/ng/journal/v46/n3/full/ng.2892.html" target="_blank">CADD_CScore</a></td>
                                                    <td valign="top"> "Raw" CADD scores come straight from the CADD model, and are interpretable as the extent to which the annotation profile for a given variant suggests that that variant is likely to be "observed" (negative values) vs "simulated" (positive values). These values have no absolute unit of meaning and are incomparable across distinct annotation combinations, training sets, or model parameters. However, raw values do have relative meaning, with higher values indicating that a variant is more likely to be simulated (or "not observed") and therefore more likely to have deleterious effects.</td>
                                                </tr>
                                                <!--
=======
                                                    <td width="52" valign="top">1</td>
                                                    <td width="114" valign="top"><a href="http://www.nature.com/ng/journal/v46/n3/full/ng.2892.html" target="_blank">CADD_CScore</a></td>
                                                    <td valign="top"> "Raw" CADD scores come straight from the CADD model, and are interpretable as the extent to which the annotation profile for a given variant suggests that that variant is likely to be "observed" (negative values) vs "simulated" (positive values). These values have no absolute unit of meaning and are incomparable across distinct annotation combinations, training sets, or model parameters. However, raw values do have relative meaning, with higher values indicating that a variant is more likely to be simulated (or "not observed") and therefore more likely to have deleterious effects.</td>
                                                </tr>
>>>>>>> origin/master
                                                <tr>
                                                    <td width="52" valign="top">2</td>
                                                    <td width="114" valign="top"><a href="http://www.nature.com/ng/journal/v46/n3/full/ng.2892.html" target="_blank">CADD_PHRED</a></td>
                                                    <td valign="top">Since the CScores do have relative meaning, one can take a specific group of variants, define the rank for each variant within that group, and then use that value as a "normalized" and now externally comparable unit of analysis. CADD scored and ranked all ~8.6 billion SNVs of the GRCh37/hg19 reference and then "PHRED-scaled" those values by expressing the rank in order of magnitude terms rather than the precise rank itself.</td>
                                                </tr>
<<<<<<< HEAD
                                                -->
                                                <tr>
                                                    <td width="52" valign="top">5</td>
=======

                                                <tr>
                                                    <td width="52" valign="top">3</td>
>>>>>>> origin/master
                                                    <td width="114" valign="top"><a href="http://bioinformatics.oxfordjournals.org/content/31/5/761" target="_blank">DANN</a></td>
                                                    <td valign="top">DANN uses the same feature set and training data as CADD to train a deep neural network (DNN). DNNs can capture nonlinear relationships among features and are better suited than SVMs for problems with a large number of samples and features.</td>
                                                </tr>
                                                <tr>
<<<<<<< HEAD
                                                    <td valign="top">6</td>
                                                    <td valign="top"><a href="http://bioinformatics.oxfordjournals.org/content/31/10/1536.long" target="_blank">FATHMM-MKL</a></td>
                                                    <td valign="top">FATHMM-MKL uses MKL classifier to predict the functional consequences of both coding and non-coding sequence variants from various genomic annotations and weights the significance of each component annotation source.</td>
                                                </tr>
                                                <tr>
                                                    <td width="52" valign="top">7</td>
=======
                                                    <td width="52" valign="top">4</td>
>>>>>>> origin/master
                                                    <td width="114" valign="top"><a href="http://www.sciencemag.org/content/342/6154/1235587" target="_blank">FunSeq</a></td>
                                                    <td valign="top">FunSeq filters mutations overlapping 1000 Genomes variants and then prioritizes those in regions under strong selection (sensitive and ultrasensitive), breaking TF motifs, and those associated with hubs. It can score the deleterious potential of variants in single or multiple genomes. The scores for each noncoding variant vary from 0 to 6, with 6 corresponding to maximum deleterious effect. When multiple tumor genomes are given as input, FunSeq also identifies recurrent mutations in the same element.</td>
                                                </tr>
                                                <tr>
<<<<<<< HEAD
                                                    <td valign="top">8</td>
=======
                                                    <td valign="top">5</td>
>>>>>>> origin/master
                                                    <td valign="top"><a href="http://genomebiology.com/2014/15/10/480" target="_blank">FunSeq2</a></td>
                                                    <td valign="top"> FunSeq2 is originally to annotate and prioritize somatic alterations integrating various resources from genomic and cancer studies. The framework consists of two components: (1) data context from uniformly processing large-scale datasets; and (2) a high-throughput variant prioritization pipeline. FunSeq2 can also be used to prioritize noncoding genetic variants.</td>
                                                </tr>
                                                <tr>
<<<<<<< HEAD
                                                    <td valign="top">9</td>
                                                    <td valign="top"><a href="http://nar.oxfordjournals.org/content/41/W1/W150.long" target="_blank">GWAS3D</a></td>
                                                    <td valign="top">GWAS3D systematically assesses the genetic variants that could affect regulatory elements, by integrating annotations from cell type-specific chromatin states, epigenetic modifications, sequence motifs and cross-species conservation. It combines the original GWAS signal, risk haplotype, binding affinity significance and conservation information to prioritize the leading variants, and infer the putative causal variant in the LD of leading variant.</td>
                                                </tr>                                                
=======
                                                    <td valign="top">6</td>
                                                    <td valign="top"><a href="http://nar.oxfordjournals.org/content/41/W1/W150.long" target="_blank">GWAS3D</a></td>
                                                    <td valign="top">GWAS3D systematically assesses the genetic variants that could affect regulatory elements, by integrating annotations from cell type-specific chromatin states, epigenetic modifications, sequence motifs and cross-species conservation. It combines the original GWAS signal, risk haplotype, binding affinity significance and conservation information to prioritize the leading variants, and infer the putative causal variant in the LD of leading variant.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">7</td>
                                                    <td valign="top"><a href="http://www.nature.com/nmeth/journal/v11/n3/abs/nmeth.2832.html" target="_blank">GWAVA_Region</a></td>
                                                    <td valign="top">GWAVA uses the random forest algorithm to build three classifiers using all available annotations to discriminate between the disease variants and variants from each of the three control sets. This control set first was composed of all 1KG variants in the 1 kb surrounding each of the HGMD variants.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">8</td>
                                                    <td valign="top"><a href="http://www.nature.com/nmeth/journal/v11/n3/abs/nmeth.2832.html" target="_blank">GWAVA_TSS</a></td>
                                                    <td valign="top">GWAVA uses the random forest algorithm to build three classifiers using all available annotations to discriminate between the disease variants and variants from each of the three control sets. This control set first was matched for distance to the nearest TSS genome-wide.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">9</td>
                                                    <td valign="top"><a href="http://www.nature.com/nmeth/journal/v11/n3/abs/nmeth.2832.html" target="_blank">GWAVA_Unmatched</a></td>
                                                    <td valign="top">GWAVA uses the random forest algorithm to build three classifiers using all available annotations to discriminate between the disease variants and variants from each of the three control sets. This control set first was constructed from a random selection of SNVs from across the genome in order to sample overall background.</td>
                                                </tr>
>>>>>>> origin/master
                                                <tr>
                                                    <td valign="top">10</td>
                                                    <td valign="top"><a href="http://genomemedicine.com/content/6/10/79" target="_blank">SuRFR</a></td>
                                                    <td valign="top">SuRFR integrates functional annotation and prior biological knowledge to prioritise candidate functional variants by regression model. It introduces novel training and validation datasets that i) capture the regional heterogeneity of genomic annotation better than previously applied approaches, and ii) facilitate understanding of which annotations are most important for discriminating different classes of functionally relevant variants from background variants.</td>
                                                </tr>
<<<<<<< HEAD

=======
                                                <tr>
                                                    <td valign="top">11</td>
                                                    <td valign="top"><a href="http://bioinformatics.oxfordjournals.org/content/31/10/1536.long" target="_blank">FATHMM-MKL</a></td>
                                                    <td valign="top">FATHMM-MKL uses MKL classifier to predict the functional consequences of both coding and non-coding sequence variants from various genomic annotations and weights the significance of each component annotation source.</td>
                                                </tr>
>>>>>>> origin/master
                                            </tbody></table>
                                        <strong>Note</strong> Predictions at variants with missing scores at specified methods will use population mean of corresponding method!



                                        <br/>

                                        <br/>
                                        <strong>Note</strong> Add the option tag <tt>--verbose-noncoding</tt> to show all detailed features used in the pathogenic prediction. However, this will consume <b>a lot of memory</b>!<br />
                                    </p>

                                    <img src="imgs/noncoding1.png" width="800" height="600"/><img src="imgs/noncoding2.png" width="800" height="600"/>
                                    <p>
                                        <b>Figure</b>: Receiver operating characteristic (ROC) curves for performances from different gene feature model.
                                    </p>


                                    Besides the above existing functional prediction scores, this option will append the following fields to the output file:
                                    <h4>
                                        <b>IsComplexDiseasePathogenic</b>: whether the variant is predicted to be pathogenic by the model trained with known pathogenic variants and neutral control variants.  <br />	
                                        <b>RandomForestScore:</b> the pathogenic scores of the random forest prediction model.

                                    </h4>
                                    <br />
<<<<<<< HEAD
                                    <b>Reference</b>: Pan et al. An improved functional prediction by linkage disequilibrium expanded and gene feature specific scores at non-coding variants. (Submitted)
                                    <br/><br/><br/> 
                                    <hr></hr>
                                    <img src="imgs/leaf.gif" width="10" height="8" /><b>Predicting noncoding regulatory variants based on composite strategy</b> <br/> 
                                    <br/>
                                    Predict genes' regulatory variants  by above 10 functional scores at non-coding variants: <tt>--db-score dbncfp_known [or dbncfp_all] --regulatory-causing-predict </tt><br/>
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --db-score dbncfp_known --regulatory-causing-predict all </code>
                                    Use 11 existing functional prediction scores (See above table) to RE-predict whether a SNV will potentially be regulatory. By default, KGGSeq uses all scores for a combinatorial prediction. On the other hand, one can FIX the prediction using any specified subset (>2) or full set (all) of the impact scores by option like --regulatory-causing-predict 1,6,8,10.
                                    <br/>
                                    <br/> 
                                    Again, if you have a lot of variants without annotation of existing tools, you can use the whole genome annotation dataset, --db-score dbncfp_all. However, you have to manually download the file from ftp://jjwanglab.org/MX/score/ANN/hgXX_all_SNV_dbNCFP.gz by efficient tools (e.g. <a href="https://filezilla-project.org/"> FileZilla</a>) and put it into the folder of KGGSeqFolder/resources/hgXX/hgXX_all_SNV_dbNCFP.gz.

                                    <br/> <br/> 
                                    Besides the above existing functional prediction scores, this option will append the following fields to the output file:
                                    <h4>                                       
                                        <b>BF</b>:Bayes factor (BF) that compares the two probability models, in which the null hypothesis is that the variant is neutral, and the alternative hypothesis is that the variant is causal. <br />	
                                        <b>Composite_P</b>: the overall probability of the variant being causal, which is calculated by multiplying the probabilities of being causal of multiple scores.
                                    </h4>
                                    <br />
                                    <b>Reference</b>: Li et al. Predicting regulatory variants with composite statistic. <a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=27273672" target="_blank">Bioinformatics. 2016 Sep 15;32(18):2729-36.</a>
                                    <br/><br/><br/> 
                                    <!--
                                    <img src="imgs/leaf.gif" width="10" height="8" /><b>Cell-type specific model</b> <br/> 
                                    <br/>
                                    Predict pathogenic variants by adding cell-type specific regulatory epigenomic markers: <tt>--db-score dbncfp_known --regulatory-pathogenic-predict  --cell </tt><br/>
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --db-score dbncfp_known --regulatory-pathogenic-predict  --cell GM12878 </code>
=======
                                    <b>Reference</b>: Pan et al. A gene-feature specific pathogenic composite prediction at non-coding variants. (Submitted)
                                    <br/><br/><br/>
                                    <hr></hr>
                                    <img src="imgs/leaf.gif" width="10" height="8" /><b>Cell-type specific model</b> <br/> 
                                    <br/>
                                    Predict pathogenic variants by adding cell-type specific regulatory epigenomic markers: <tt>--db-score dbncfp --regulatory-pathogenic-predict  --cell </tt><br/>
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --db-score dbncfp --regulatory-pathogenic-predict  --cell GM12878 </code>
>>>>>>> origin/master
                                    Assigning a cell type, KGGSeq uses a logit model to measure the probability of regulatory causality for given variants in selected condition. It finally combine the above composite model and condition-dependent model into an unified model to estimate the posterior probability of regulatory potential.
                                    <br/> <br/> 
                                    KGGSeq supports <a href="https://www.encodeproject.org/"  target="_blank">16 ENCODE cell types</a> as follows: 

                                    <table class="list">
                                        <tbody><tr>
                                                <td width="72" valign="top"><b>Coding</b></td>
                                                <td width="72" valign="top"><b>Tissue</b></td>
                                                <td valign="top"><b>Description</b></td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">A549</td>
                                                <td width="72" valign="top">Epithelium</td>
                                                <td valign="top">epithelial cell line derived from a lung carcinoma tissue. (PMID: 175022), "This line was initiated in 1972 by D.J. Giard, et al. through explant culture of lung carcinomatous tissue from a 58-year-old caucasian male." - ATCC, newly promoted to tier 2: not in 2011 analysis.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">CD14</td>
                                                <td width="72" valign="top">Monocytes</td>
                                                <td valign="top">Monocytes-CD14+ are CD14-positive cells from human leukapheresis production, from donor RO 01746 (draw 1 ID is RO 01746, draw 2 ID is RO 01826), newly promoted to tier 2: not in 2011 analysis.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">GM12878</td>
                                                <td width="72" valign="top">Blood</td>
                                                <td valign="top">B-lymphocyte, lymphoblastoid, International HapMap Project - CEPH/Utah - European Caucasion, Epstein-Barr Virus.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">H1-hESC</td>
                                                <td width="72" valign="top">Embryonic stem cell</td>
                                                <td valign="top">embryonic stem cells.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">HMEC</td>
                                                <td width="72" valign="top">Breast</td>
                                                <td valign="top">mammary epithelial cells.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">HSMM</td>
                                                <td width="72" valign="top">Muscle</td>
                                                <td valign="top">skeletal muscle myoblasts.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">HSMMT</td>
                                                <td width="72" valign="top">Muscle</td>
                                                <td valign="top">HSMM cell derived skeletal muscle myotubes cell line.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">HUVEC</td>
                                                <td width="72" valign="top">Vessel</td>
                                                <td valign="top">umbilical vein endothelial cells.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">HeLa-S3</td>
                                                <td width="72" valign="top">Cervix</td>
                                                <td valign="top">cervical carcinoma.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">HepG2</td>
                                                <td width="72" valign="top">Liver</td>
                                                <td valign="top">hepatocellular carcinoma.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">IMR90</td>
                                                <td width="72" valign="top">Lung</td>
                                                <td valign="top">fetal lung fibroblasts, newly promoted to tier 2: not in 2011 analysis.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">K562</td>
                                                <td width="72" valign="top">Blood</td>
                                                <td valign="top">leukemia, "The continuous cell line K-562 was established by Lozzio and Lozzio from the pleural effusion of a 53-year-old female with chronic myelogenous leukemia in terminal blast crises." - ATCC.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">NH-A</td>
                                                <td width="72" valign="top">Brain</td>
                                                <td valign="top">astrocytes (also called Astrocy).</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">NHDF</td>
                                                <td width="72" valign="top">Skin</td>
                                                <td valign="top">dermal fibroblasts from temple / breast</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">NHEK</td>
                                                <td width="72" valign="top">Skin</td>
                                                <td valign="top">epidermal keratinocytes.</td>
                                            </tr>
                                            <tr>
                                                <td width="72" valign="top">NHLF</td>
                                                <td width="72" valign="top">Lung</td>
                                                <td valign="top">lung fibroblasts.</td>
                                            </tr>
                                        </tbody></table>
                                    <strong>Note</strong>By default, KGGSeq will use GM12878 for condition-dependent prioritization.
                                    <br/><br/>


                                    KGGSeq also supports <a href="http://www.roadmapepigenomics.org/", target="_blank"> 127 RoadMap human reference epigenomes</a> : 

<<<<<<< HEAD
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --db-score dbncfp_known --regulatory-pathogenic-predict --cell E116 </code>
=======
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --db-score dbncfp --regulatory-pathogenic-predict --cell E116 </code>
>>>>>>> origin/master
                                    <table class="list">
                                        <tbody>
                                            <tr>
                                                <td style="font-weight: bold;">Coding</td>
                                                <td style="font-weight: bold;">Lineage Group</td>
                                                <td style="font-weight: bold;">Epigenome Mnemonic</td>
                                                <td style="font-weight: bold;">Epigenome Name</td>
                                                <td style="font-weight: bold;">Anatomy</td>
                                                <td style="font-weight: bold;">Type</td>
                                            </tr>
                                            <tr>
                                                <td>E001</td>
                                                <td>ESC</td>
                                                <td>ESC.I3</td>
                                                <td>ES-I3 Cell Line</td>
                                                <td>ESC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E002</td>
                                                <td>ESC</td>
                                                <td>ESC.WA7</td>
                                                <td>ES-WA7 Cell Line</td>
                                                <td>ESC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E003</td>
                                                <td>ESC</td>
                                                <td>ESC.H1</td>
                                                <td>H1 Cell Line</td>
                                                <td>ESC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E004</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.H1.BMP4.MESO</td>
                                                <td>H1 BMP4 Derived Mesendoderm Cultured Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E005</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.H1.BMP4.TROP</td>
                                                <td>H1 BMP4 Derived Trophoblast Cultured Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E006</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.H1.MSC</td>
                                                <td>H1 Derived Mesenchymal Stem Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E007</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.H1.NEUR.PROG</td>
                                                <td>H1 Derived Neuronal Progenitor Cultured Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E008</td>
                                                <td>ESC</td>
                                                <td>ESC.H9</td>
                                                <td>H9 Cell Line</td>
                                                <td>ESC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E009</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.H9.NEUR.PROG</td>
                                                <td>H9 Derived Neuronal Progenitor Cultured Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E010</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.H9.NEUR</td>
                                                <td>H9 Derived Neuron Cultured Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E011</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.CD184.ENDO</td>
                                                <td>hESC Derived CD184+ Endoderm Cultured Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E012</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.CD56.ECTO</td>
                                                <td>hESC Derived CD56+ Ectoderm Cultured Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E013</td>
                                                <td>ES-deriv</td>
                                                <td>ESDR.CD56.MESO</td>
                                                <td>hESC Derived CD56+ Mesoderm Cultured Cells</td>
                                                <td>ESC_DERIVED</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E014</td>
                                                <td>ESC</td>
                                                <td>ESC.HUES48</td>
                                                <td>HUES48 Cell Line</td>
                                                <td>ESC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E015</td>
                                                <td>ESC</td>
                                                <td>ESC.HUES6</td>
                                                <td>HUES6 Cell Line</td>
                                                <td>ESC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E016</td>
                                                <td>ESC</td>
                                                <td>ESC.HUES64</td>
                                                <td>HUES64 Cell Line</td>
                                                <td>ESC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E017</td>
                                                <td>IMR90</td>
                                                <td>LNG.IMR90</td>
                                                <td>IMR90 fetal lung fibroblasts Cell Line</td>
                                                <td>LUNG</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E018</td>
                                                <td>iPSC</td>
                                                <td>IPSC.15b</td>
                                                <td>iPS-15b Cell Line</td>
                                                <td>IPSC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E019</td>
                                                <td>iPSC</td>
                                                <td>IPSC.18</td>
                                                <td>iPS-18 Cell Line</td>
                                                <td>IPSC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E020</td>
                                                <td>iPSC</td>
                                                <td>IPSC.20B</td>
                                                <td>iPS-20b Cell Line</td>
                                                <td>IPSC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E021</td>
                                                <td>iPSC</td>
                                                <td>IPSC.DF.6.9</td>
                                                <td>iPS DF 6.9 Cell Line</td>
                                                <td>IPSC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E022</td>
                                                <td>iPSC</td>
                                                <td>IPSC.DF.19.11</td>
                                                <td>iPS DF 19.11 Cell Line</td>
                                                <td>IPSC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E023</td>
                                                <td>Mesench</td>
                                                <td>FAT.MSC.DR.ADIP</td>
                                                <td>Mesenchymal Stem Cell Derived Adipocyte Cultured Cells</td>
                                                <td>FAT</td>
                                                <td>CellLineDerived</td>
                                            </tr>
                                            <tr>
                                                <td>E024</td>
                                                <td>ESC</td>
                                                <td>ESC.4STAR</td>
                                                <td>&nbsp; </span>Cell Line</td>
                                                <td>ESC</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E025</td>
                                                <td>Mesench</td>
                                                <td>FAT.ADIP.DR.MSC</td>
                                                <td>Adipose Derived Mesenchymal Stem Cell Cultured Cells</td>
                                                <td>FAT</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E026</td>
                                                <td>Mesench</td>
                                                <td>STRM.MRW.MSC</td>
                                                <td>Bone Marrow Derived Cultured Mesenchymal Stem Cells</td>
                                                <td>NECTIVE</span></td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E027</td>
                                                <td>Epithelial</td>
                                                <td>BRST.MYO</td>
                                                <td>Breast Myoepithelial Primary Cells</td>
                                                <td>BREAST</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E028</td>
                                                <td>Epithelial</td>
                                                <td>BRST.HMEC.35</td>
                                                <td>Breast variant Human Mammary Epithelial Cells (vHMEC)</td>
                                                <td>BREAST</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E029</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.CD14.PC</td>
                                                <td>Primary monocytes from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E030</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.CD15.PC</td>
                                                <td>Primary neutrophils from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E031</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.CD19.CPC</td>
                                                <td>Primary B cells from cord blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E032</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.CD19.PPC</td>
                                                <td>Primary B cells from peripheral blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E033</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD3.CPC</td>
                                                <td>Primary T cells from cord blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E034</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD3.PPC</td>
                                                <td>Primary T cells from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E035</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.CD34.PC</td>
                                                <td>Primary hematopoietic stem cells</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E036</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.CD34.CC</td>
                                                <td>Primary hematopoietic stem cells short term culture</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E037</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.MPC</td>
                                                <td>Primary T helper memory cells from peripheral blood 2</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E038</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.NPC</td>
                                                <td>Primary T helper naive cells from peripheral blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E039</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.CD25M.CD45RA.NPC</td>
                                                <td>Primary T helper naive cells from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E040</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.CD25M.CD45RO.MPC</td>
                                                <td>Primary T helper memory cells from peripheral blood 1</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E041</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.CD25M.IL17M.PL.TPC</td>
                                                <td>Primary T helper cells PMA-I stimulated</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E042</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.CD25M.IL17P.PL.TPC</td>
                                                <td>Primary T helper 17 cells PMA-I stimulated</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E043</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.CD25M.TPC</td>
                                                <td>Primary T helper cells from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E044</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.CD25.CD127M.TREGPC</td>
                                                <td>Primary T regulatory cells from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E045</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD4.CD25I.CD127.TMEMPC</td>
                                                <td>Primary T cells effector/memory enriched
                                                    from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E046</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.CD56.PC</td>
                                                <td>Primary Natural Killer cells from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E047</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD8.NPC</td>
                                                <td>Primary T killer naive cells from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E048</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.CD8.MPC</td>
                                                <td>Primary T killer memory cells from peripheral blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E049</td>
                                                <td>Mesench</td>
                                                <td>STRM.CHON.MRW.DR.MSC</td>
                                                <td>Mesenchymal Stem Cell Derived Chondrocyte Cultured Cells</td>
                                                <td>NECTIVE</span></td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E050</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.MOB.CD34.PC.F</td>
                                                <td>Primary hematopoietic stem cells G-CSF-mobilized Female</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E051</td>
                                                <td>HSC &amp; B-cell</td>
                                                <td>BLD.MOB.CD34.PC.M</td>
                                                <td>Primary hematopoietic stem cells G-CSF-mobilized Male</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E052</td>
                                                <td>Myosat</td>
                                                <td>MUS.SAT</td>
                                                <td>Muscle Satellite Cultured Cells</td>
                                                <td>MUSCLE</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E053</td>
                                                <td>Neurosph</td>
                                                <td>BRN.CRTX.DR.NRSPHR</td>
                                                <td>Cortex derived primary cultured neurospheres</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E054</td>
                                                <td>Neurosph</td>
                                                <td>BRN.GANGEM.DR.NRSPHR</td>
                                                <td>Ganglion Eminence derived primary cultured neurospheres</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E055</td>
                                                <td>Epithelial</td>
                                                <td>SKIN.PEN.FRSK.FIB.01</td>
                                                <td>Foreskin Fibroblast Primary Cells skin01</td>
                                                <td>SKIN</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E056</td>
                                                <td>Epithelial</td>
                                                <td>SKIN.PEN.FRSK.FIB.02</td>
                                                <td>Foreskin Fibroblast Primary Cells skin02</td>
                                                <td>SKIN</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E057</td>
                                                <td>Epithelial</td>
                                                <td>SKIN.PEN.FRSK.KER.02</td>
                                                <td>Foreskin Keratinocyte Primary Cells skin02</td>
                                                <td>SKIN</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E058</td>
                                                <td>Epithelial</td>
                                                <td>SKIN.PEN.FRSK.KER.03</td>
                                                <td>Foreskin Keratinocyte Primary Cells skin03</td>
                                                <td>SKIN</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E059</td>
                                                <td>Epithelial</td>
                                                <td>SKIN.PEN.FRSK.MEL.01</td>
                                                <td>Foreskin Melanocyte Primary Cells skin01</td>
                                                <td>SKIN</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E061</td>
                                                <td>Epithelial</td>
                                                <td>SKIN.PEN.FRSK.MEL.03</td>
                                                <td>Foreskin Melanocyte Primary Cells skin03</td>
                                                <td>SKIN</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E062</td>
                                                <td>Blood &amp; T-cell</td>
                                                <td>BLD.PER.MONUC.PC</td>
                                                <td>Primary mononuclear cells from&nbsp;peripheral&nbsp;blood</td>
                                                <td>BLOOD</td>
                                                <td>PrimaryCell</td>
                                            </tr>
                                            <tr>
                                                <td>E063</td>
                                                <td>Adipose</td>
                                                <td>FAT.ADIP.NUC</td>
                                                <td>Adipose Nuclei</td>
                                                <td>FAT</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E065</td>
                                                <td>Heart</td>
                                                <td>VAS.AOR</td>
                                                <td>Aorta</td>
                                                <td>VASCULAR</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E066</td>
                                                <td>Other</td>
                                                <td>LIV.ADLT</td>
                                                <td>Liver</td>
                                                <td>LIVER</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E067</td>
                                                <td>Brain</td>
                                                <td>BRN.ANG.GYR</td>
                                                <td>Brain Angular Gyrus</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E068</td>
                                                <td>Brain</td>
                                                <td>BRN.ANT.CAUD</td>
                                                <td>Brain Anterior Caudate</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E069</td>
                                                <td>Brain</td>
                                                <td>BRN.CING.GYR</td>
                                                <td>Brain Cingulate Gyrus</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E070</td>
                                                <td>Brain</td>
                                                <td>BRN.GRM.MTRX</td>
                                                <td>Brain Germinal Matrix</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E071</td>
                                                <td>Brain</td>
                                                <td>BRN.HIPP.MID</td>
                                                <td>Brain Hippocampus Middle</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E072</td>
                                                <td>Brain</td>
                                                <td>BRN.INF.TMP</td>
                                                <td>Brain Inferior Temporal Lobe</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E073</td>
                                                <td>Brain</td>
                                                <td>BRN.DL.PRFRNTL.CRTX</td>
                                                <td>Brain Dorsolateral Prefrontal Cortex</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E074</td>
                                                <td>Brain</td>
                                                <td>BRN.SUB.NIG</td>
                                                <td>Brain Substantia Nigra</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E075</td>
                                                <td>Digestive</td>
                                                <td>GI.CLN.MUC</td>
                                                <td>Colonic Mucosa</td>
                                                <td>GI_COLON</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E076</td>
                                                <td>Sm. Muscle</td>
                                                <td>GI.CLN.SM.MUS</td>
                                                <td>Colon Smooth Muscle</td>
                                                <td>GI_COLON</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E077</td>
                                                <td>Digestive</td>
                                                <td>GI.DUO.MUC</td>
                                                <td>Duodenum Mucosa</td>
                                                <td>GI_DUODENUM</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E078</td>
                                                <td>Sm. Muscle</td>
                                                <td>GI.DUO.SM.MUS</td>
                                                <td>Duodenum Smooth Muscle</td>
                                                <td>GI_DUODENUM</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E079</td>
                                                <td>Digestive</td>
                                                <td>GI.ESO</td>
                                                <td>Esophagus</td>
                                                <td>S</span></td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E080</td>
                                                <td>Other</td>
                                                <td>ADRL.GLND.FET</td>
                                                <td>Fetal Adrenal Gland</td>
                                                <td>ADRENAL</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E081</td>
                                                <td>Brain</td>
                                                <td>BRN.FET.M</td>
                                                <td>Fetal Brain Male</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E082</td>
                                                <td>Brain</td>
                                                <td>BRN.FET.F</td>
                                                <td>Fetal Brain Female</td>
                                                <td>BRAIN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E083</td>
                                                <td>Heart</td>
                                                <td>HRT.FET</td>
                                                <td>Fetal Heart</td>
                                                <td>HEART</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E084</td>
                                                <td>Digestive</td>
                                                <td>GI.L.INT.FET</td>
                                                <td>Fetal Intestine Large</td>
                                                <td>GI_INTESTINE</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E085</td>
                                                <td>Digestive</td>
                                                <td>GI.S.INT.FET</td>
                                                <td>Fetal Intestine Small</td>
                                                <td>GI_INTESTINE</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E086</td>
                                                <td>Other</td>
                                                <td>KID.FET</td>
                                                <td>Fetal Kidney</td>
                                                <td>KIDNEY</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E087</td>
                                                <td>Other</td>
                                                <td>PANC.ISLT</td>
                                                <td>Pancreatic Islets</td>
                                                <td>PANCREAS</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E088</td>
                                                <td>Other</td>
                                                <td>LNG.FET</td>
                                                <td>Fetal Lung</td>
                                                <td>LUNG</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E089</td>
                                                <td>Muscle</td>
                                                <td>MUS.TRNK.FET</td>
                                                <td>Fetal Muscle Trunk</td>
                                                <td>MUSCLE</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E090</td>
                                                <td>Muscle</td>
                                                <td>MUS.LEG.FET</td>
                                                <td>Fetal Muscle Leg</td>
                                                <td>MUSCLE_LEG</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E091</td>
                                                <td>Other</td>
                                                <td>PLCNT.FET</td>
                                                <td>Placenta</td>
                                                <td>PLACENTA</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E092</td>
                                                <td>Digestive</td>
                                                <td>GI.STMC.FET</td>
                                                <td>Fetal Stomach</td>
                                                <td>GI_STOMACH</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E093</td>
                                                <td>Thymus</td>
                                                <td>THYM.FET</td>
                                                <td>Fetal Thymus</td>
                                                <td>THYMUS</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E094</td>
                                                <td>Digestive</td>
                                                <td>GI.STMC.GAST</td>
                                                <td>Gastric</td>
                                                <td>GI_STOMACH</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E095</td>
                                                <td>Heart</td>
                                                <td>HRT.VENT.L</td>
                                                <td>Left Ventricle</td>
                                                <td>HEART</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E096</td>
                                                <td>Other</td>
                                                <td>LNG</td>
                                                <td>Lung</td>
                                                <td>LUNG</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E097</td>
                                                <td>Other</td>
                                                <td>OVRY</td>
                                                <td>Ovary</td>
                                                <td>OVARY</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E098</td>
                                                <td>Other</td>
                                                <td>PANC</td>
                                                <td>Pancreas</td>
                                                <td>PANCREAS</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E099</td>
                                                <td>Other</td>
                                                <td>PLCNT.AMN</td>
                                                <td>Placenta Amnion</td>
                                                <td>PLACENTA</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E100</td>
                                                <td>Muscle</td>
                                                <td>MUS.PSOAS</td>
                                                <td>Psoas Muscle</td>
                                                <td>MUSCLE</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E101</td>
                                                <td>Digestive</td>
                                                <td>GI.RECT.MUC.29</td>
                                                <td>Rectal Mucosa Donor 29</td>
                                                <td>GI_RECTUM</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E102</td>
                                                <td>Digestive</td>
                                                <td>GI.RECT.MUC.31</td>
                                                <td>Rectal Mucosa Donor 31</td>
                                                <td>GI_RECTUM</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E103</td>
                                                <td>Sm. Muscle</td>
                                                <td>GI.RECT.SM.MUS</td>
                                                <td>Rectal Smooth Muscle</td>
                                                <td>GI_RECTUM</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E104</td>
                                                <td>Heart</td>
                                                <td>HRT.ATR.R</td>
                                                <td>Right Atrium</td>
                                                <td>HEART</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E105</td>
                                                <td>Heart</td>
                                                <td>HRT.VNT.R</td>
                                                <td>Right Ventricle</td>
                                                <td>HEART</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E106</td>
                                                <td>Digestive</td>
                                                <td>GI.CLN.SIG</td>
                                                <td>Sigmoid Colon</td>
                                                <td>GI_COLON</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E107</td>
                                                <td>Muscle</td>
                                                <td>MUS.SKLT.M</td>
                                                <td>Skeletal Muscle Male</td>
                                                <td>MUSCLE</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E108</td>
                                                <td>Muscle</td>
                                                <td>MUS.SKLT.F</td>
                                                <td>Skeletal Muscle Female</td>
                                                <td>MUSCLE</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E109</td>
                                                <td>Digestive</td>
                                                <td>GI.S.INT</td>
                                                <td>Small Intestine</td>
                                                <td>GI_INTESTINE</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E110</td>
                                                <td>Digestive</td>
                                                <td>GI.STMC.MUC</td>
                                                <td>Stomach Mucosa</td>
                                                <td>GI_STOMACH</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E111</td>
                                                <td>Sm. Muscle</td>
                                                <td>GI.STMC.MUS</td>
                                                <td>Stomach Smooth Muscle</td>
                                                <td>GI_STOMACH</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E112</td>
                                                <td>Thymus</td>
                                                <td>THYM</td>
                                                <td>Thymus</td>
                                                <td>THYMUS</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E113</td>
                                                <td>Other</td>
                                                <td>SPLN</td>
                                                <td>Spleen</td>
                                                <td>SPLEEN</td>
                                                <td>PrimaryTissue</td>
                                            </tr>
                                            <tr>
                                                <td>E114</td>
                                                <td>ENCODE</td>
                                                <td>LNG.A549.ETOH002.CNCR</td>
                                                <td>A549 EtOH 0.02pct Lung Carcinoma Cell Line</td>
                                                <td>LUNG</td>
                                                <td>CellLine_Cancer</td>
                                            </tr>
                                            <tr>
                                                <td>E115</td>
                                                <td>ENCODE</td>
                                                <td>BLD.DND41.CNCR</td>
                                                <td>Dnd41 TCell Leukemia Cell Line</td>
                                                <td>BLOOD</td>
                                                <td>CellLine_Cancer</td>
                                            </tr>
                                            <tr>
                                                <td>E116</td>
                                                <td>ENCODE</td>
                                                <td>BLD.GM12878</td>
                                                <td>GM12878 Lymphoblastoid Cell Line</td>
                                                <td>BLOOD</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E117</td>
                                                <td>ENCODE</td>
                                                <td>CRVX.HELAS3.CNCR</td>
                                                <td>HeLa-S3 Cervical Carcinoma Cell Line</td>
                                                <td>CERVIX</td>
                                                <td>CellLine_Cancer</td>
                                            </tr>
                                            <tr>
                                                <td>E118</td>
                                                <td>ENCODE</td>
                                                <td>LIV.HEPG2.CNCR</td>
                                                <td>HepG2 Hepatocellular Carcinoma Cell Line</td>
                                                <td>LIVER</td>
                                                <td>CellLine_Cancer</td>
                                            </tr>
                                            <tr>
                                                <td>E119</td>
                                                <td>ENCODE</td>
                                                <td>BRST.HMEC</td>
                                                <td>HMEC Mammary Epithelial Primary Cells</td>
                                                <td>BREAST</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E120</td>
                                                <td>ENCODE</td>
                                                <td>MUS.HSMM</td>
                                                <td>HSMM Skeletal Muscle Myoblasts Cell Line</td>
                                                <td>MUSCLE</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E121</td>
                                                <td>ENCODE</td>
                                                <td>MUS.HSMMT</td>
                                                <td>HSMM cell derived Skeletal Muscle Myotubes Cell Line</td>
                                                <td>MUSCLE</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E122</td>
                                                <td>ENCODE</td>
                                                <td>VAS.HUVEC</td>
                                                <td>HUVEC Umbilical Vein Endothelial Cells Cell Line</td>
                                                <td>VASCULAR</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E123</td>
                                                <td>ENCODE</td>
                                                <td>BLD.K562.CNCR</td>
                                                <td>K562 Leukemia Cell Line</td>
                                                <td>BLOOD</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E124</td>
                                                <td>ENCODE</td>
                                                <td>BLD.CD14.MONO</td>
                                                <td>Monocytes-CD14+ RO01746 Cell Line</td>
                                                <td>BLOOD</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E125</td>
                                                <td>ENCODE</td>
                                                <td>BRN.NHA</td>
                                                <td>NH-A Astrocytes Cell Line</td>
                                                <td>BRAIN</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E126</td>
                                                <td>ENCODE</td>
                                                <td>SKIN.NHDFAD</td>
                                                <td>NHDF-Ad Adult Dermal Fibroblast Primary Cells</td>
                                                <td>SKIN</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E127</td>
                                                <td>ENCODE</td>
                                                <td>SKIN.NHEK</td>
                                                <td>NHEK-Epidermal Keratinocyte Primary Cells</td>
                                                <td>SKIN</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E128</td>
                                                <td>ENCODE</td>
                                                <td>LNG.NHLF</td>
                                                <td>NHLF Lung Fibroblast Primary Cells</td>
                                                <td>LUNG</td>
                                                <td>CellLine</td>
                                            </tr>
                                            <tr>
                                                <td>E129</td>
                                                <td>ENCODE</td>
                                                <td>BONE.OSTEO</td>
                                                <td>Osteoblast Primary Cells</td>
                                                <td>BONE</td>
                                                <td>CellLine</td>
                                            </tr>
                                        </tbody>
                                    </table>


                                    The option will append the following fields to the output file:
                                    <h4>                                        	
                                        <b>BF</b>: Bayes factor  <br />	
                                        <b>Composite_P</b>:Posterior probability of being pathogenic given the above 11 functional predictions.  <br />	
                                        <b>Cell_P</b>: cell type-specific regulatory probability.  <br />	
                                        <b>Combine_P</b> combined posterior probability of pathogenicity.

                                    </h4>
                                    <strong>Note</strong> Epigenomes of some specific marks are missing in some tissues/cells, which may affect the prediction. We will use imputed epigenomes to handle this missing data problem soon! 
                                    <br/>
                                    <br/>
                                    <b>Figure:</b> Using GWAS fine mapped SNPs and 127 recently profiled human epigenomes, we showed that our condition-dependent score can comprehensively depict the phenotypic cell type specificity. <br/>
                                    <img src="./imgs/heatmap.png" width="1300" height="600"/><br/>

                                    <br/>
                                    <br/>
                                    <b>Reference</b>: Li et al. Cell-type specific pathogenic composite prediction at non-coding variants. (In preparation)
                                    <p>&nbsp;</p>
                                </td>
                            </tr>
<<<<<<< HEAD
                                    -->
                                    <tr>
                                        <td><h2><a name="PredictionGene" >Prediction at gene </a></h2> 
                                            <h3><a name="Pathogenicity" >Pathogenicity </a></h3>                                     
                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> Predict genes&rsquo; pathogenicity: <tt>--patho-gene-predict</tt>
                                                <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --excel --db-gene refgene --patho-gene-predict </code>
                                                Gene-based pathogenicity estimation was developed for Autosomal Dominant (AD), Autosomal Recessive (AR), X chromosome linked (XL) and Paediatric diseases (PAE). 
                                                By using random forest approach to combine various important gene characteristics with six existing gene prioritization methods for disease causal genes in <a href="http://research.nhgri.nih.gov/CGD/", target="_blank" >Clinical Genomic Database</a>, the inheritance-specific pathogenicity prioritization (ISPP) could be estimated. 
                                                The higher score indicates the more similar to Clinical Genomic Disease genes in many molecular characteristics. The six existing gene prioritization methods are Haploinsufficiency (HI), Recessive (REC), Network centrality (NET), Genic Intolerance (RVIS), Gene Damage Index (GDI) and Gene Constraint score (CONS). 
                                                <p>  The output is a series of pathogenic prediction scores for each gene. By comparing all ISPP scores of each gene, the possible inheritance mode and corresponding pathogenicity could be estimated. 
                                                    <h4>
                                                        <b>ISPP_AD</b>: Inheritance specific gene pathogenicity for Dominant diseases <br/>
                                                        <b>ISPP_AR</b>: Inheritance specific gene pathogenicity for Recessive diseases <br/>
                                                        <b>ISPP_XL</b>: Inheritance specific gene pathogenicity for X chromosome linked diseases<br/>
                                                        <!--<b>ISPP_PAE</b>: Paediatric disease gene pathogenicity<br/>-->
                                                        <b>HI</b>: Gene level haploinsufficiency estimation by <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1001154",target="_blank">Ni Huang et al</a><br/>
                                                        <b>REC</b>: A model to estimate the probability that a mutation causes a recessive disease by <a href="http://science.sciencemag.org/content/335/6070/823",target="_blank"> Daniel G. MacArthur et al.</a><br/>
                                                        <b>NET</b>: Gene centrality and indispensability in various protein-protein interactions and regulatory networks by <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002886",target="_blank"> Ekta Khurana et al.</a><br/> 
                                                        <b>RVIS</b>: Residual Variation Intolerance Score , a gene score based module intended to help in the interpretation of human sequence data by <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003709",target="_blank"> Slave Petrovski et al.</a><br/>
                                                        <b>GDI</b>: Gene Damage Index (GDI) by <a href="http://www.ncbi.nlm.nih.gov/pubmed/26483451",target="_blank"> Itan Y et al.</a><br/> 
                                                        <b>CONS</b>: De novo mutation effectiveness prediction by <a href="http://www.nature.com/ng/journal/v46/n9/abs/ng.3050.html",target="_blank"> Kaitlin E Samocha et al.</a> <br/>
                                                    </h4>
                                                    <br/>
                                                    <b>Reference</b>: Jacob Hsu et al. Inheritance Modes Specific Pathogenicity Prioritization (ISPP) for Human Protein Coding Genes. Bioinformatics 2016 (Accepted)
                                                </p>
                                        </td>                              	
                                    </tr>

                                    <tr>                          	                          	                           	
                                        <td><h3><a name="PhenotypeMining" >Phenotype Mining </a></h3>
                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> Mine the phenotype of disease: <tt>--phenotype-term searchTerm and --phenolyzer-prediction</tt>
                                                <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --excel --db-gene refgene --db-filter 1kg201305 --phenotype-term searchTerm1,searchTerm2 --phenolyzer-prediction  </code>
                                                KGGseq has integrated the advanced tool <a href="http://phenolyzer.usc.edu/" ,target="_blank">phenolyzer</a> as a plugin to implement the function of phenoytpe mining and predection. The score marked by phenolyer will by used to annotate the related genes in the output file. 
                                                <p><strong> Note</strong>To run Phenolyzer, your computer need install Perl. Please read the <a href="http://learnperl.scratchcomputing.com/install/" ,target="_blank"> installation instructions. </a> </p>
                                            </p>
                                        </td>
                                    </tr>                            

                                    <tr>
                                        <td><h3><a name="Expression" > Expression </a></h3>
                                            <p><strong>Note</strong>This feature will be available soon. Stay tuned for an update!<br />
                                            </p>                                    
                                            <p>&nbsp;</p>
                                        </td>                              	
                                    </tr>                           

                                    <tr>
                                        <td><h2><a name="AssociationEnrichment" >Statistical tests</a></h2>
                                            <h3><a name="AssociationVariant" >Association at variants</a></h3>

                                            <p><img src="imgs/leaf.gif" width="10" height="8" />Conventional statistical association tests at each individual variants: <tt> --var-assoc </tt></p>
                                            <p><code>java -jar ./kggseq.jar --buildver hg19  --var-assoc --p-value-cutoff 0.01 --multiple-testing benfdr --qqplot</code> 
                                                Calculate p-values (by chi-squared test) for case-control association and odds ratio. The <tt>'--p-value-cutoff'</tt> sets p-value cutoffs for filtration; the <tt>'--multiple-testing'</tt> sets the approaches for multiple testing comparisons; and the <tt>'--qqplot'</tt> enables <a href="http://en.wikipedia.org/wiki/Q-Q_plot", target="_blank">Q-Q plot</a> of the association results.</p>
                                            <table  class="list">
                                                <tr>
                                                    <td valign="top"><b>Multiple testing method</b></td>
                                                    <td valign="top"><b>Description</b></td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">no</td>
                                                    <td valign="top">No multiple testing methods. Set a fixed p value cutoff by <tt>--p-value-cutoff</tt>.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">benfdr</td>
                                                    <td valign="top">Benjamini-Hochberg method, which aims to control the FDR (specified by  <tt>--p-value-cutoff</tt>), or "False Discovery Rate," rather than the FWE or "FamilyWise Error rate".</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">bonhol</td>
                                                    <td valign="top">Bonferroni-Holm method, which is less conservative and more powerful than the Standard Bonferroni method given a family-wise error rate specified by <tt>--p-value-cutoff</tt>.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">bonf</td>
                                                    <td valign="top">Standard Bonferroni correction given a family-wise error rate specified by <tt>--p-value-cutoff</tt>.</td>
                                                </tr>
                                            </table>

                                            The output of this function is a list of p-values and odds ratios at a variant.
                                            <h4>
                                                <b> PValueAllelic: </b>The p-value calculated by chi-square test for allelic association between case-control status.<br />
                                                <b> OddAllelic: </b>The odds ratio of alleles in cases and controls.<br />
                                                <b> PValueDom: </b>The p-value calculated by chi-square test for genotype association between case-control status under dominant mode.<br />
                                                <b> OddDom: </b>The odds ratio of genotypes in cases and controls under dominant mode.<br />
                                                <b> PValueDom: </b>The p-value calculated by chi-square test for genotype association between case-control status under recessive mode.<br />
                                                <b> OddDom: </b>The odds ratio of genotypes in cases and controls under recessive mode.<br />
                                                <b> PValueGeno: </b>The p-value calculated by chi-square test for genotype association between case-control status.<br />
                                            </h4>

                                            <p><strong>Note</strong> Currently, this function only work for case-control study without confounding factors as covariables. </p>

                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> Variant-based association test by <a href="http://zhanxw.github.io/rvtests/", target="_blank">Rvtests</a>: <tt> --rvtest-var</tt></p>
                                            <p><code>java -jar ./kggseq.jar --rvtest-var score,wald,exact,... --rvtest-vcf --rvtest-remove-set --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene </code>
                                                KGGSeq will automatically download  the C++ sources codes from  <a href="http://zhanxw.github.io/rvtests/", target="_blank">Github</a> and un-compress them into a folder named "plugin/rvtests-master" under the KGGSeq working folder. Users are required go into the "rvtests-master" to make an executable file  from the C++ source codes. See more in <a  href="http://grass.cgs.hku.hk/limx/kggseq/doc10/demov10.mp4" target="_blank"> a demonstration video</a> .
                                            </p>
                                            <table  class="list">
                                                <tr>
                                                    <td valign="top"><b>Arguments/values</b></td>
                                                    <td valign="top"><b>Annotation</b></td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-var model names</td>
                                                    <td valign="top">Initiate the calling of RVTest for gene based association analysis. The model names are defined in <a href="http://zhanxw.github.io/rvtests/#models", target="_blank"><b>Rvtests models</b></a>. Currently, it supports score, wald, exact, dominantExact, famLRT, famScore, famGrammarGamma and firthc models. The default models are "score,wald".                                                                        
                                                    </td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-vcf</td>
                                                    <td valign="top">Ask KGGSeq to produce a new simplified VCF file, in which only genotypes are included in the FORMAT column, as the input of RVTests. </td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-remove-set</td>
                                                    <td valign="top">Remove the intermediate file defining sets of variants for RVTests group-wise test in your hard disk.</td>
                                                </tr>
                                            </table>
                                            <p>An extra file with the suffix ".rvtest.variant.xlsx" will be created in the targeted output path. </p>


                                            <h3><a name="AssociationGene">Association at gene</a></h3>
                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> <a name="GeneSkat" > Gene-based association test</a> by <a href="http://www.hsph.harvard.edu/skat/", target="_blank">SKAT</a>: <tt> --skat-gene</tt></p>
                                            <p><code>java -jar ./kggseq.jar  --skat-gene --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene [--p-value-cutoff 0.01 --multiple-testing benfdr --qqplot]</code>
                                                The gene-based analysis will be carried out by <a href="http://www.hsph.harvard.edu/skat/",target="_blank">SKAT</a> which was integrated as a plugin in KGGseq. 
                                                The <tt>'--p-value-cutoff'</tt> sets p-value cutoffs for filtration; the <tt>'--multiple-testing'</tt> sets the approaches for multiple testing comparisons; and the <tt>'--qqplot'</tt> enables <a href="http://en.wikipedia.org/wiki/Q-Q_plot", target="_blank">Q-Q plot</a> of the association results.</p>
                                            <table  class="list">
                                                <tr>
                                                    <td valign="top"><b>Multiple testing method</b></td>
                                                    <td valign="top"><b>Description</b></td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">no</td>
                                                    <td valign="top">No multiple testing methods. Set a fixed p value cutoff by <tt>--p-value-cutoff</tt>.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">benfdr</td>
                                                    <td valign="top">Benjamini-Hochberg method, which aims to control the FDR (specified by  <tt>--p-value-cutoff</tt>), or "False Discovery Rate," rather than the FWE or "FamilyWise Error rate".</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">bonhol</td>
                                                    <td valign="top">Bonferroni-Holm method, which is less conservative and more powerful than the Standard Bonferroni method given a family-wise error rate specified by <tt>--p-value-cutoff</tt>.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">bonf</td>
                                                    <td valign="top">Standard Bonferroni correction given a family-wise error rate specified by <tt>--p-value-cutoff</tt>.</td>
                                                </tr>
                                            </table>
                                            <p>An extra file with the suffix ".skat.gene.xlsx" will be created in the targeted output path. The <tt>'--skat-gene'</tt> will provide the p values of SKAT,SKATO and Burden test. If the phenotype is binary, the option <tt>'--skat-gene binary'</tt> can be used, but the speed will decrease. </p>

                                            <h4>                                       
                                                <b>SKAT:</b>The p-value calculated by SKAT. <br />
                                                <b>SKATO:</b>The p-value calculated by SKAT-O. <br />
                                                <b>Burden:</b>The p-value calculated by Burden test.<br />
                                            </h4>

                                            <p><strong>Note</strong>  To use this function, the user should install <a href="https://www.r-project.org/", target="_blank">R</a> language platform first. </p>

                                            <p><strong>Note</strong> When total minor allele count (MAC) is very small in a gene, SKATBinary ( <tt>'--skat-gene binary'</tt> ) has better performance than SKAT( <tt>'--skat-gene'</tt> ). However, because SKATBinary uses a resampling approach to get a p-value, it is slower (sometimes quite slower) than SKAT function. When MAC is large, the performance of the two tests is similar.</p>
                                            <p><strong>Note</strong> The names of phenotype ( <tt>'--phe'</tt> ) and co-variables ( <tt>'--cov'</tt> ) are defined in the  <a href="#PhenotypesAndPedigree">pedigree file</a> .  </p>
                                            <p><strong>Note</strong> If the option <tt>--skat-cutoff N</tt> is added, the genes with variants less than <tt>N</tt> will not be involved into SKAT. The default value of <tt>N</tt> is 0.</p>
                                            <p><strong>Note</strong> If the option <tt>--perm-pheno </tt> is added, phenotypes will be permuted prior to association analysis. </p>

                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> <a name="GeneRvtests" >Gene-based association test</a> by <a href="http://zhanxw.github.io/rvtests/", target="_blank">Rvtests</a>: <tt> --rvtest-gene</tt></p>
                                            <p><code>java -jar ./kggseq.jar --rvtest-gene cmc,cmat,price,skat[nPerm=1000:alpha=0.001:beta1=1:beta2=20],... --rvtest-vcf --rvtest-remove-set --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene </code>
                                                KGGSeq will automatically download  the C++ sources codes from  <a href="http://zhanxw.github.io/rvtests/", target="_blank">Github</a> and un-compress them into a folder named "plugin/rvtests-master" under the KGGSeq working folder. Users are required go into the "rvtests-master" to make an executable file  from the C++ source codes. See more in <a  href="http://grass.cgs.hku.hk/limx/kggseq/doc10/demov10.mp4" target="_blank"> a demonstration video</a> .
                                            </p>
                                            <table  class="list">
                                                <tr>
                                                    <td valign="top"><b>Arguments/values</b></td>
                                                    <td valign="top"><b>Annotation</b></td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-gene model names</td>
                                                    <td valign="top">Initiate the calling of RVTest for gene based association analysis. The model names are defined in <a href="http://zhanxw.github.io/rvtests/#models", target="_blank"><b>Rvtests models</b></a>. Currently, it supports cmc, zeggini, mb, fp, exactCMC, cmcWald, rarecover and cmat models.The default models are "cmc,skato[nPerm=1000:alpha=0.001:beta1=1:beta2=20]".</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-vcf</td>
                                                    <td valign="top">Ask KGGSeq to produce a new VCF file as the input of RVTests. However, this is very time-consuming for large sample. It is advised to use <tt>--o-vcf</tt> to produce a VCF file with good quality once and to use that cleaned VCF file as the new KGGSeq input. In this scenario, the <tt>--rvtest-vcf</tt> is not needed, which may substantially reduce the analysis time.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-remove-set</td>
                                                    <td valign="top">Remove the intermediate file defining sets of variants for RVTests group-wise test in your hard disk.</td>
                                                </tr>
                                            </table>

                                            <p>An extra file with the suffix ".rvtest.gene.xlsx" will be created in the targeted output path. </p>
                                            <!--
                                                                                        <p><img src="imgs/leaf.gif" width="10" height="8" /><a name="GeneDoubleHitTrio" > Association for double-hit genes in case-only sample</a>: <tt>--double-hit-gene-trio-case-pseudo</tt></p>
                                                                                        <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --db-gene refgene --gene-feature-in 0,1,2,3,4,5,6 --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.05 --db-score dbnsfp --mendel-causing-predict all --double-hit-gene-trio-case-pseudo</code></p>
                                                                                        <p>This function is designed for a disorder suspected to be under compound-heterozygous or recessive inheritance mode, in which both copies of a gene on the two homologous chromosomes of a patient are damaged by two different mutations or the same mutation. For recessive mode, it simply checks variants with homozygous genotypes in patients. For the compound-heterozygous mode. Here the trio refers to <b>the two parents and an offspring</b>.  When these alleles causing a disease at one locus, the frequency of the double-hit genes should be different between case and the pseudo-controls. The difference can be evaluated by a statistical test.  </p>
                                                                                        <p>Besides the file storing prioritized sequence variants, this analysis with <tt>--double-hit-gene-trio-pseudo</tt> option will lead to a list of double-hit genes in at least one subject. The counts of double-hit genes of each subject are saved in *.doublehit.gene.trios.flt.count.xls (or *.doublehit.gene.trios.flt.count.txt) and the genotypes of involved sequence variants are saved in *.doublehit.gene.trios.flt.gty.xls (or *.doublehit.gene.trios.flt.gty.txt).  <br/></p>
                                                                                        <p>
                                                                                            The '*doublehit.gene.trios.flt.count*' and the '*doublehit.gene.trios.flt.gty*' files or sheets contain multiple columns:
                                                                                            <h4>
                                                                                                Gene: Gene symbols	<br />
                                                                                                PubMed: The PumMed ID(s) of papers co-mentioning the gene and phenotypes(or diseases) specified by '--phenotype-term'. <br />
                                                                                                ExonLen: The length of total exons in kb. The overlapped regions are ignored.<br />                                           
                                                                                                CountHitCase: Number of cases having double-hit variants in the gene.<br />
                                                                                                CountHitPseudoControl: Number of pseudo-controls having double-hit variants in the gene.<br />
                                                                                                TotalNum: Number of trios having genotypes in the gene.<br />
                                                                                                p-Value:  The p-value of Pearson chi-square test or Fisher's exact test (when the count of a component is &#60;&#61;5 ) <br />
                                            
                                                                                                &#60;Subjects&#62;: Whether a subject carry the double-hit variants,1 for yes and 0 for no in the 'flt.count' file or sheet. In the 'gty.count' file or sheet, the variants positions and genotypes are shown. <br/>
                                            
                                                                                                Example: <br/>
                                                                                                17797224:0/0,0/1,0/1;17827193:0/1,0/0,0/1: The genotypes of <b>father</b>, <b>mother</b>, and <b>child</b> are <b>0/0</b>, <b>0/1</b> and <b>0/1</b> at the variant with coordinate 17797224. At the other variant, the genotypes of <b>father</b>, <b>mother</b>, and <b>child</b> are <b>0/1</b>, <b>0/0</b> and <b>0/1</b>. Apparently, both copies of this gene in the child are hit by the two mutations. 
                                                                                            </h4>
                                                                                        </p>
                                            
                                                                                        <p><img src="imgs/leaf.gif" width="10" height="8" /> Association for double-hit genes in case/control sample: <tt>--double-hit-gene-trio-case-control</tt></p>
                                                                                        <p><code>java -jar ./kggseq.jar --buildver hg19 --vcf-file path/to/file1 --ped-file path/to/file2 --db-gene refgene --gene-feature-in 0,1,2,3,4,5,6 --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.05 --db-score dbnsfp --mendel-causing-predict all --double-hit-gene-trio-case-control</code></p>
                                                                                        <p>This function is designed for a disorder suspected to be under compound-heterozygous or recessive inheritance mode, in which both copies of a gene on the two homologous chromosomes of a patient are damaged by two different mutations or the same mutation. For recessive mode, it simply checks variants with homozygous genotypes in patients. For the compound-heterozygous mode. Here the trio refers to <b>the two parents and an offspring</b>.  When these alleles causing a disease at one locus, the frequency of the double-hit genes should be different between case and controls. The difference can be evaluated by a statistical test.  </p>
                                                                                        <p>Besides the file storing prioritized sequence variants, this analysis with <tt>--double-hit-gene-trio-control</tt> option will lead to a list of double-hit genes in at least one subject. The counts of double-hit genes of each subject are saved in *.doublehit.gene.trios.flt.count.xls (or *.doublehit.gene.trios.flt.count.txt) and the genotypes of involved sequence variants are saved in *.doublehit.gene.trios.flt.gty.xls (or *.doublehit.gene.trios.flt.gty.txt).  <br/></p>
                                                                                        <p>
                                                                                            The '*doublehit.gene.trios.flt.count*' and the '*doublehit.gene.trios.flt.gty*' files or sheets contain multiple columns:
                                                                                            <h4>
                                                                                                Gene: Gene symbols	<br />
                                                                                                PubMed: The PumMed ID(s) of papers co-mentioning the gene and phenotypes(or diseases) specified by '--phenotype-term'. <br />
                                                                                                ExonLen: The length of total exons in kb. The overlapped regions are ignored.<br />                                           
                                                                                                CountHitCase: Number of cases having double-hit variants in the gene.<br />
                                                                                                TotalNumCase: Total number of cases  having genotypes in the gene.<br />
                                                                                                CountHitControl: Number of controls having double-hit variants in the gene.<br />
                                                                                                TotalNumControl: Total number of controls having genotypes in the gene.<br />
                                                                                                p-Value:  The p-value of Pearson chi-square test or Fisher's exact test (when the count of a component is &#60;&#61;5 ) <br />
                                                                                                &#60;Subjects&#62;: Whether a subject carry the double-hit variants,1 for yes and 0 for no in the 'flt.count' file or sheet. In the 'gty.count' file or sheet, the variants positions and genotypes are shown. <br/>
                                            
                                                                                                Example: <br/>
                                                                                                17797224:0/0,0/1,0/1;17827193:0/1,0/0,0/1: The genotypes of <b>father</b>, <b>mother</b>, and <b>child</b> are <b>0/0</b>, <b>0/1</b> and <b>0/1</b> at the variant with coordinate 17797224. At the other variant, the genotypes of <b>father</b>, <b>mother</b>, and <b>child</b> are <b>0/1</b>, <b>0/0</b> and <b>0/1</b>. Apparently, both copies of this gene in the child are hit by the two mutations. 
                                                                                            </h4>
                                                                                        </p>
                                            -->

                                            <h3><a name="MutationRateGene" >Mutation rate test at gene</a></h3>
                                            <br />                                  
                                            Test the certain types mutation rate in a gene by a regression mode <tt>--gene-mutation-rate-test --dependent-gene-feature [variants&apos; gene feature codes] --independent-gene-feature [variants&apos; gene feature codes]</tt> 
                                            <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname [other tags] --gene-mutation-rate-test</code>

                                            The model can be expressed as the following:<br/>
                                            <center>[#minor alleles of variants specified by &apos;--dependent-gene-feature&apos;]=[#minor alleles of variants specified by &apos;--independent-gene-feature&apos;] + [length of coding region] + [constraint score]+[expression]+[DNA replication timing]+[HiC-based chromatin state estimation]</center> <br/><br/>
                                            The gene&apos;s constraint scores are from <a href="http://www.nature.com/ng/journal/v46/n9/abs/ng.3050.html" target="_blank">Samocha et al (2014) Nature Genetics. </a>. The last three factors are from <a href="http://www.broadinstitute.org/cancer/software/genepattern/modules/docs/MutSigCV" target="_blank">MutSigCV</a>.The assumption of the analysis is that the number of minor alleles of specified dependent variants in a gene can be normally explained by other factors of the same gene including length of coding region, the number of minor alleles of specified independent variants, constraint score and other expression related features on the human genome. However, if a gene contributes to a certain disease by mutations at specified dependent variants, this gene tends to have more mutant alleles at the specified dependent variants in a sample of this disease.

                                            <p><strong>Note</strong> By default, the setting of <tt>--dependent-gene-feature</tt> is 0,1,2,3,4,5,6 (i.e., all non-synonymous variants) and the setting of <tt>--independent-gene-feature</tt> is 7 (i.e., the synonymous variants only).
                                                <br/> <strong>Note</strong> The gene feature code can be found <a href="#GeneFeatureFiltering" target="_self">HERE</a>.                                    
                                            </p> 
                                            This approach is designed for test gene &apos;s mutation rate in three scenarios.<br/>
                                            1) In cancer samples for gene&apos;s somatic mutation rate: 
                                            <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --genotype-filter 8 --indiv-pair tumorIndivID1:normalIndivID1,tumorIndivID2:normalIndivID2 --db-gene refgene --gene-mutation-rate-test --dependent-gene-feature 0,1,2,3,4,5,6 --independent-gene-feature 7</code>

                                            2)In a cohort for gene&apos;s rare non-synonymous mutation rate : 
                                            <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.01  --db-gene refgene --gene-mutation-rate-test --dependent-gene-feature 0,1,2,3,4,5,6 --independent-gene-feature 7</code>

                                            <!--
                                            2) In trios samples for gene&apos;s <i>de novo</i> mutation rate: 
                                           <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --genotype-filter 7 --db-gene refgene --gene-mutation-rate-test --dependent-gene-feature 0,1,2,3,4,5,6 --independent-gene-feature 7</code>
        
        
                                           4)In case-control samples for gene&apos;s case-unique rare non-synonymous mutation rate: 
                                           <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.01 --filter-model case-unique --db-gene refgene --gene-mutation-rate-test --dependent-gene-feature 0,1,2,3,4,5,6 --independent-gene-feature 7</code>
                                            -->
                                            In the output file of the analysis (*.gene.mutation.rate.xlsx), there are detailed values of all factors used in the regression and the p-values of mutation rate test.
                                            <h4>
                                                <b>GeneSymbol</b>: Official gene symbols defined by HGNC.<br />
                                                <b>AvgCodingLen</b>: The actual or averaged (of multiple isoforms) length (in kb) of all exons in a gene.<br />
                                                <b>NS</b>: The observed number of non-synonymous and splicing somatic sequence variants.<br />
                                                <b>Syn</b>: The observed number of synonymous somatic sequence variants.<br />
                                                [...] <br/>
                                                <b>RegResidue:</b>  Pearson residuals   (raw residuals scaled by square root of variance function) in the regression after correction for the factors. <br />
                                                <b>RegP:</b>  The probability of be equal or over the RegResidue under a standard normal distribution.<br />


                                            </h4>
                                            Note: The detailed method of the analysis will be described in a paper, which is in preparation.
                                            <br />

                                            <h3><a name="GeneSetAssociation">Association at geneset</a></h3>
                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> <a name="GenesetSkat" >Geneset-based association test </a> by <a href="http://www.hsph.harvard.edu/skat/", target="_blank">SKAT</a>: <tt> --skat-geneset</tt></p>
                                            <p><code>java -jar ./kggseq.jar  --skat-geneset --geneset-file path/to/geneset/file --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene [--p-value-cutoff 0.01 --multiple-testing benfdr --qqplot]</code>
                                                The gene-based analysis will be carried out by <a href="http://www.hsph.harvard.edu/skat/",target="_blank">SKAT</a> which was integrated as a plugin in KGGseq. 
                                                The <tt>'--p-value-cutoff'</tt> sets p-value cutoffs for filtration; the <tt>'--multiple-testing'</tt> sets the approaches for multiple testing comparisons; and the <tt>'--qqplot'</tt> enables <a href="http://en.wikipedia.org/wiki/Q-Q_plot", target="_blank">Q-Q plot</a> of the association results.</p>

                                            <strong>Note</strong>, the pathway gene sets are specified in a text file in th following format:<br/>

                                            <h4>
                                                [Column 1: GeneSet ID]<br />
                                                [Column 2: GeneSet URL]<br />
                                                [Column 3..N: Gene Symbols in the geneset; separated by spaces].<br />

                                                No title row is required!!!<br />
                                                <br />

                                                <br />    e.g.,<br />
                                                ----------------------------------------------<br />
                                                KEGG_GLYCOLYSIS_GLUCONEOGENESIS http://www.broadinstitute.org/KEGG_GLYCOLYSIS_GLUCONEOGENESIS.html LDHC LDHB LDHA<br />
                                                KEGG_CITRATE_CYCLE_TCA_CYCLE http://www.broadinstitute.org/KEGG_CITRATE_CYCLE_TCA_CYCLE.html GJB1 OGDHL OGDH PDHB IDH3G LDHB<br />
                                                ...<br />
                                                ...<br />
                                            </h4>

                                            <p>An extra file with the suffix ".skat.geneset.xlsx" will be created in the targeted output path. The <tt>'--skat-geneset'</tt> will provide the p values of SKAT,SKATO and Burden test. If the phenotype is binary, the option <tt>'--skat-gene binary'</tt> can be used, but the speed will decrease. </p>

                                            <h4> 
                                                <b>SKAT:</b>The p-value calculated by SKAT. <br />
                                                <b>SKATO:</b>The p-value calculated by SKAT-O. <br />
                                                <b>Burden:</b>The p-value calculated by Burden test.<br />
                                            </h4>

                                            <p><strong>Note</strong>  To use this function, the user should install <a href="https://www.r-project.org/", target="_blank">R</a> language platform first. </p>

                                            <p><strong>Note</strong> When total minor allele count (MAC) is very small in a gene, SKATBinary ( <tt>'--skat-gene binary'</tt> ) has better performance than SKAT( <tt>'--skat-gene'</tt> ). However, because SKATBinary uses a resampling approach to get a p-value, it is slower (sometimes quite slower) than SKAT function. When MAC is large, the performance of the two tests is similar.</p>
                                            <p><strong>Note</strong> The names of phenotype ( <tt>'--phe'</tt> ) and co-variables ( <tt>'--cov'</tt> ) are defined in the  <a href="#PhenotypesAndPedigree">pedigree file</a> .  </p>
                                            <p><strong>Note</strong> If the option <tt>--skat-cutoff N</tt> is added, the genes with variants less than <tt>N</tt> will not be involved into SKAT. The default value of <tt>N</tt> is 0.</p>
                                            <p><strong>Note</strong> If the option <tt>--perm-pheno </tt> is added, phenotypes will be permuted prior to association analysis. </p>

                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> <a name="GenesetRvtests" >Geneset-based association test </a> by <a href="http://zhanxw.github.io/rvtests/", target="_blank">Rvtests</a>: <tt> --rvtest-geneset</tt></p>
                                            <p><code>java -jar ./kggseq.jar  --rvtest-geneset cmc,skato[nPerm=1000:alpha=0.001:beta1=1:beta2=20],... --rvtest-vcf --rvtest-remove-set --geneset-file path/to/geneset/file --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene [--p-value-cutoff 0.01 --multiple-testing benfdr --qqplot]</code>
                                                KGGSeq will automatically download  the C++ sources codes from  <a href="http://zhanxw.github.io/rvtests/", target="_blank">Github</a> and un-compress them into a folder named "plugin/rvtests-master" under the KGGSeq working folder. Users are required go into the "rvtests-master" to make an executable file  from the C++ source codes. See more in <a  href="http://grass.cgs.hku.hk/limx/kggseq/doc10/demov10.mp4" target="_blank"> a demonstration video</a> .
                                            </p>
                                            <table  class="list">
                                                <tr>
                                                    <td valign="top"><b>Arguments/values</b></td>
                                                    <td valign="top"><b>Annotation</b></td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-gene model names</td>
                                                    <td valign="top">Initiate the calling of RVTest for gene based association analysis. The model names are defined in <a href="http://zhanxw.github.io/rvtests/#models", target="_blank"><b>Rvtests models</b></a>. Currently, it supports cmc, zeggini, mb, fp, exactCMC, cmcWald, rarecover and cmat models.The default models are "cmc,skato[nPerm=1000:alpha=0.001:beta1=1:beta2=20]".</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-vcf</td>
                                                    <td valign="top">Ask KGGSeq to produce a new VCF file as the input of RVTests. However, this is very time-consuming for large sample. It is advised to use <tt>--o-vcf</tt> to produce a VCF file with good quality once and to use that cleaned VCF file as the new KGGSeq input. In this scenario, the <tt>--rvtest-vcf</tt> is not needed, which may substantially reduce the analysis time.</td>
                                                </tr>
                                                <tr>
                                                    <td valign="top">--rvtest-remove-set</td>
                                                    <td valign="top">Remove the intermediate file defining set of variants for RVTests group-wise test.</td>
                                                </tr>
                                            </table>

                                            <p>An extra file with the suffix ".rvtest.geneset.xlsx" will be created in the targeted output path. The <tt>'--rvtest-geneset'</tt> will provide the  p values of multiple set-based tests for rare variants.  </p>


                                            <h3><a name="GeneSetEnrichment" >Enrichment at geneset</a></h3>                                  
                                            <p>Test whether a  set of genes highly ranked in an ordered list of all genes by Wilcoxon rank sum test:<tt> --geneset-enrichment-test </tt>
                                                <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname [other tags] --geneset-enrichment-test</code>

                                                Basically, this analysis is to examine whether a  set of genes are more highly ranked in an ordered list of all genes than would be expected by chance. Therefore,as the rank of gene is a prerequisite, it should be used together with a gene-based statistical test.<br/>


                                                Currently, this enrichment test can only be used together with the gene-based mutation rate test:
                                                <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.01  --db-gene refgene --gene-mutation-rate-test --geneset-enrichment-test --geneset-db &lt;GenesetDatabase&gt;</code>

                                                <p>There are 5 geneset databases from <a href="http://www.broadinstitute.org/gsea/msigdb/collections.jsp">GSEA</a>. The following are brief descriptions of the 5 datasets COPIED from the <a href="http://www.broadinstitute.org/gsea/msigdb/collections.jsp" target="_blank">page</a> of GSEA.<br/>
                                                    <table width="90%" class="list">
                                                        <tr>
                                                            <th width="19%" scope="col">GeneSetDatabase</th>
                                                            <th width="66%" scope="col">Description</th>
                                                            <th width="15%" scope="col">Version</th>
                                                        </tr>
                                                        <tr>
                                                            <td>cano</td>
                                                            <td>Gene sets from the geneset databases. Usually, these gene sets are canonical representations of a biological process compiled by domain experts. (1452 gene sets)</td>
                                                            <td> MSigDB 3.1</td>
                                                        </tr>
                                                        <tr>
                                                            <td> cura </td>
                                                            <td>Curated gene sets: Gene sets collected from various sources such as online geneset databases, publications in PubMed, and knowledge of domain experts. The gene set page for each gene set lists its source. (4850 gene sets)</td>
                                                            <td> MSigDB 3.1</td>
                                                        </tr>
                                                        <tr>
                                                            <td> onco </td>
                                                            <td> Oncogenic signatures: Gene sets represent signatures of cellular pathways which are often dis-regulated in cancer. The majority of signatures were generated directly from microarray data from NCBI GEO or from internal unpublished profiling experiments which involved perturbation of known cancer genes. In addition, a small number of oncogenic signatures were curated from scientific publications. (189 gene sets)</td>
                                                            <td> MSigDB 3.1</td>
                                                        </tr>
                                                        <tr>
                                                            <td> cmop </td>
                                                            <td>Computational gene sets:  Computational gene sets defined by mining large collections of cancer-oriented microarray data. (858 gene sets)</td>
                                                            <td> MSigDB 3.1</td>
                                                        </tr>
                                                        <tr>
                                                            <td> onto </td>
                                                            <td>GO gene sets Gene sets are named by GO term and contain genes annotated by that term. GSEA users: Gene set enrichment analysis identifies gene sets consisting of co-regulated genes; GO gene sets are based on ontologies and do not necessarily comprise co-regulated genes. (1454 gene sets)</td>
                                                            <td> MSigDB 3.1</td>
                                                        </tr>
                                                    </table>
                                                    <br />
                                                    <strong>Alternatively</strong>, uses can specify the own pathway gene sets in a text file by <tt>--geneset-file path/to/geneset/file</tt>. <br/>


                                                    <strong>Note </strong>, please use <tt>--min-geneset X</tt> to ignore gene sets with less X genes and --max-geneset to ignore gene sets with over X genes.
                                                    <br />

                                                    In the output file of the enrichment analysis (*.enrichment.geneset.xls), there are detailed statistical evaluation results of each geneset:
                                                    <h4>
                                                        <b>GeneSetID</b>:the input geneset name or id<br />
                                                        <b>WilcoxonP</b>:the p-values of the enrichment test<br />
                                                        <b>Gene</b>:the gene of the gene set<br />
                                                        <b>GeneP</b>:the p-value of the gene used for the enrichment test<br />
                                                    </h4>
                                                    <br />
                                                </p>
                                                <p>&nbsp;</p>
                                        </td>
                                    </tr>


                                    <tr>
                                        <td><h2><a name="Plotting" >Plotting </a></h2>
                                            <p>Generate a histogram of minor allele frequency (MAF) according to  allele frequencies in reference dataset <tt>--mafplot-ref</tt> or in sample <tt>--mafplot-sample</tt>. 
                                                <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --db-filter 1kg201204,dbsnp141,ESP6500AA,ESP6500EA  --allele-freq 0,1  --mafplot-ref</code>
                                                <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname  --mafplot-sample</code>
                                                <strong> Note</strong> Currently, --mafplot-sample only supports the inputs specified by --vcf-file.<br />
                                                <br />

                                                Generate a Quantile-Quantile (Q-Q) plot for association p-values <tt>--qqplot</tt>
                                                <code>java -jar kggseq.jar --vcf-file path/to/file --out path/to/prefixname --qqplot --skat-gene [or] --var-assoc</code>                                        
                                                <br />
                                            </p>


                                        </td>
                                    </tr>

                                    <tr>
                                        <td><h2><a name="Miscellaneous" >Miscellaneous </a></h2>
                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> <a name="calLD" > Calculate pairwise the linkage disequilibrium or genotypic correlation </a>  <tt>--calc-ld</tt>. 
                                                <code>java -jar kggseq.jar --vcf-file path/to/file --ped-file path/to/file --out path/to/prefixname --calc-ld --db-filter 1kg201204,dbsnp141,ESP6500AA,ESP6500EA  --allele-freq 0.05,0.95   --regions-in chr4:21212-233454</code>                                       
                                                If the genotypes are phased, KGGSeq will calculate the linkage disequilibrium coefficient r-square. For uphased genotypes, KGGSeq will calculate the Pearson genotypic correlation of variants, in which the genotypes are encoded by the number of alternative alleles.
                                                <br /> 
                                                <strong> Note</strong> Currently,  <tt>--calc-ld</tt> only supports variants with two alleles.<br />
                                                <strong> Note</strong> By default, KGGSeq will only calculate LD between variants within a window of 100,000,000 base-pairs. The window size can be changed by <tt>--ld-win X </tt>, in which X denotes a distance in base-pair. <br />
                                                <strong> Note</strong> By default, only the r-square or correlation square >0.01 will be saved in an output file. This can be changed by <tt>--ld-out X</tt>, in which X denotes r-square or correlation square. .<br />
                                                <p>The results will be outputted in a text file  with the suffix ".ld" or ".ld.gz".  The following is an exmaple of the output format:
                                                    <h4> 
                                                        [chromsome] [coordinate of the first variant] [coordinate of the second variant] [r-square OR correlation]<br />
                                                        22	16051249	16052962	0.8680002633575351<br />
                                                        22	16051249	16053659	0.13732498950210686<br />
                                                        22	16051249	16053862	0.984494959737711
                                                    </h4>
                                                </p> 
                                            </p> 

                                            <p><img src="imgs/leaf.gif" width="10" height="8" /> <a name="ldPrune" >Prune variants in high LD (r or genotypic correlation square >= x)</a> <tt> --ld-prune X</tt>. 
                                                <code>java -jar kggseq.jar --vcf-file path/to/file --ped-file path/to/file --out path/to/prefixname --ld-prune 0.9 --db-filter 1kg201204,dbsnp141,ESP6500AA,ESP6500EA  --allele-freq 0.05,0.95 </code> 
                                                KGGSeq will calculate the pairwise LD between variants within a window of 100,000,000 base-pairs. If a pair of variants have r-square or correlation square over 0.9, the variation with larger coordinate will be pruned.   The window size can be changed by <tt>--ld-win X</tt>, in which X denotes a distance in base-pair. 
                                            </p>


                                        </td>
                                    </tr>

                                    <tr>
                                        <td><hr />                                   
                                            <h1> <a name="faq">FAQ?</a></h1>
                                            <p><b>1) Can a compressed VCF file be processed in parallel?</b></p>
                                            <p>Yes, it can. However, the files has to be compressed in a blocked GNU Zip format or Burrows&#8211;Wheeler format. See more here.</p>
                                            <p><b>2) Does the KGGSeq binary genotype format support variants of over two alleles and/or phased genotypes?</b></p>
                                            <p>Yes, it does support variants of over two alleles and/or phased genotypes. See more here.</p>
                                            <p><b>3) How are the settings '--rare-allele-freq c' and '--allele-freq 0,c' different?</b></p>
                                            <p>  '--rare-allele-freq c' will excluded variants with alternative allele frequency EQUAL to or over c in the reference datasets; and '--allele-freq 0,c' will also excluded variants having no allele frequency in the reference datasets beside the variants excluded by '--rare-allele-freq c'</p> 
                                            <p><b>4) Does KGGSeq provide gene expression information in the output?</b></p>
                                            <p>A project aiming to harness tissue-specific expression information to prioritize candidate genes is underway.</p>
                                            <p><b>5) How shall I prepare the candidate gene list for protein-interaction and geneset annotation? What is the gene  list for?</b></p>
                                            <p>The candidate gene list can include any gene interesting to the user. It  may be verified causal gene for the disease, or potential causal gene. The gene  list is used to fish more hidden interesting genes in order to narrow down the  searching region. If no gene list provided, KGGSeq can still perform routine  &lsquo;filtration + sharing&rsquo; strategy for sequencing data.</p>
                                            <p><b>6)Suppose I have in-house control genomes in VCF format, how can I  transform it into KGGSeq acceptable file?</b></p>
                                            <p>There are two alternative ways to allow you use the in-house VCF data (containing  genotypes of multiple individuals) as filter.</p>
                                            <p>1. Convert the VCF data into a standard KGGSeq local filter format by the  command: <em>--make-filter --vcf-file path/to/vcffile --out  test.var </em>first. And then set the --local-filter <em>test.hg19.var</em></p>
                                            <p>2. Directly set <em>--local-filter-vcf path/to/vcffile. </em>However, this  way is less efficient than the above way as it will take additional time to  parse the VCF data.</p>
                                            <p><b>7) Is there a practical guidance for MAF  selection?</b></p>
                                            <p> MAF selection is an open question, but generally three suggestions can be adopted: for rare dominant Mendelian disease, MAF cut-off at 0 or 0.005 is reasonable; for rare recessive Mendelian disease, MAF cut-off at 0 or 0.03 is reasonable;
                                                for common complex disease of  relatively rare alleles with large effect size, MAF cut-off is dependent on  disease prevalence, penetrance and genetic model, so a higher cut-off like 0.05  might be more reasonable.</p>
                                            <p><b>8) Can I use ANNOVAR and KGGSeq together on my dataset?</b></p>
                                            <p>KGGSeq is quite flexible for interacting with other sequence-oriented  analytical programs/software (including SOAPSNP, ANNOVAR, PLINK/seq, VAAST, etc).  It can accept various input formats, and output different kinds of sequence  data. In case of ANNOVAR, KGGSeq can read ANNOVAR-formatted sequence variants,  and write the final remaining variants in ANNOVAR format.</p>
                                            <p><b>9) Can I run KGGSeq on my Lenovo or Mac laptop? Is it time consuming  to run a complete KGGSeq process?</b></p>
                                            <p>Normally, KGGSeq run well and fast with &gt;=1 GB RAM memory. Hence  current laptop are certainty affordable for running KGGSeq. The whole process  need only &lt;10 minutes, unless PubMed searching is set in the parameter file.  For PubMed searching, it really depends on the speed of the network and also  the loading of the NCBI PubMed database.</p>
                                            <p><b>10) How do I report an error  or bugs to KGGSeq?</b></p>
                                            <p>You are welcomed to join our  google group (<a href="http://groups.google.com/group/kggseq_user?hl=en">http://groups.google.com/group/kggseq_user?hl=en</a>).  This site is used for communication and discussion of KGGSeq usage and  functions. You can also directly write an email to <a href="mailto:limx54@yahoo.com">limx54@yahoo.com</a>, or  kggseq_user@googlegroups.com.</p>
                                        </td>
                                    </tr>
                                    </table>



                                    <p>&nbsp;</p>
                                </td>
                            </tr>
                        </table>
                        </div>

                        <div  id="menu">
                            <a href="#Navi:introduction">References</a><br />
                            <a href="#Navi:installation">Installation</a><br />
                            <a href="#Navi:shorttutorials">Tutorials</a>
                            <a href="#MendelianDiseaseHg19">Examples</a><br />
                            <a href="#Navi:options">Options</a><br />
                            <a name="Navi:faq" href="#faq"> FAQ</a>

                        </div>
                        </body>
                        </html>
=======

                            <tr>
                                <td><h2><a name="PredictionGene" >Prediction at gene </a></h2> 
                                    <h3><a name="Pathogenicity" >Pathogenicity </a></h3>                                     
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Predict genes&rsquo; pathogenicity: <tt>--patho-gene-predict</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --excel --db-gene refgene --patho-gene-predict </code>
                                        Gene-based pathogenicity estimation was developed for different disease types including autosomal dominant (AD) , autosomal recessive (AR), X chromosome linked (XL) and paediatric disease (PAE) by a random forest approach combining various important gene characteristics from disease causal genes in <a href="http://research.nhgri.nih.gov/CGD/", target="_blank" >Clinical Genomic Database</a> . The higher score indicates the more similar characteristics to the reported disease causal genes. Other four public available existing genes&rsquo; functional scores, Haploinsufficiency (HI), Recessive (REC), Genic Intolerance (RVIS) and Gene Constraint (CONS) were also included in the combined prediction. 
                                        <p>  The output of this function is a series of pathogenic prediction scores at involved genes. Gene level prioritization can be estimated by comparing these pathogenicity scores across different candidate genes.
                                            <h4>
                                                <b>AD</b>: Autosomal Dominant disease gene pathogenicity <br/>
                                                <b>AR</b>: Autosomal Recessive disease gene pathogenicity <br/>
                                                <b>XL</b>: X chromosome linked disease gene pathogenicity<br/>
                                                <b>PAE</b>: Paediatric disease gene pathogenicity<br/>
                                                <b>HI</b>: Gene level <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1001154",target="_blank">Haploinsufficiency</a> estimation<br/>
                                                <b>REC</b>: A model to estimate the probability that a mutation causes a recessive disease by <a href=" http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3299548/",target="_blank"> MacArthur et al.</a><br/>
                                                <b>CONS</b>: De novo mutation effectiveness prediction by <a href="http://www.nature.com/ng/journal/v46/n9/abs/ng.3050.html",target="_blank"> Samocha et al.</a> <br/>
                                                <b>RVIS</b>: <a href="http://genic-intolerance.org/",target="_blank"> Residual Variation Intolerance Score</a> , a gene score based module intended to help in the interpretation of human sequence data.
                                            </h4>
                                            <br/>
                                            <b>Reference</b>: Hsu et al. Compiling a Classification System of Human Protein Coding Genes Based on the Mode of Mendelian Disease Inheritance. (Submitted)
                                        </p>
                                </td>                              	
                            </tr>

                            <tr>                          	                          	                           	
                                <td><h3><a name="PhenotypeMining" >Phenotype Mining </a></h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Mine the phenotype of disease: <tt>--phenotype-term searchTerm and --phenolyzer-prediction</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --excel --db-gene refgene --db-filter 1kg201305 --phenotype-term searchTerm1,searchTerm2 --phenolyzer-prediction  </code>
                                        KGGseq has integrated the advanced tool <a href="http://phenolyzer.usc.edu/" ,target="_blank">phenolyzer</a> as a plugin to implement the function of phenoytpe mining and predection. The score marked by phenolyer will by used to annotate the related genes in the output file. 
                                        <p><strong> Note</strong>To run Phenolyzer, your computer need install Perl. Please read the <a href="http://learnperl.scratchcomputing.com/install/" ,target="_blank"> installation instructions. </a> </p>
                                    </p>
                                </td>
                            </tr>                            

                            <tr>
                                <td><h3><a name="Expression" > Expression </a></h3>
                                    <p><strong>Note</strong>This feature will be available soon. Stay tuned for an update!<br />
                                    </p>                                    
                                    <p>&nbsp;</p>
                                </td>                              	
                            </tr>                           

                            <tr>
                                <td><h2><a name="AssociationEnrichment" >Statistical tests</a></h2>
                                    <h3><a name="AssociationVariant" >Association at variants</a></h3>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" />Conventional statistical association tests at each individual variants: <tt> --var-assoc </tt></p>
                                    <p><code>java -jar ./kggseq.jar --buildver hg19  --var-assoc --p-value-cutoff 0.01 --multiple-testing benfdr --qqplot</code> 
                                        Calculate p-values (by chi-squared test) for case-control association and odds ratio. The <tt>'--p-value-cutoff'</tt> sets p-value cutoffs for filtration; the <tt>'--multiple-testing'</tt> sets the approaches for multiple testing comparisons; and the <tt>'--qqplot'</tt> enables <a href="http://en.wikipedia.org/wiki/Q-Q_plot", target="_blank">Q-Q plot</a> of the association results.</p>
                                    <table  class="list">
                                        <tr>
                                            <td valign="top"><b>Multiple testing method</b></td>
                                            <td valign="top"><b>Description</b></td>
                                        </tr>
                                        <tr>
                                            <td valign="top">no</td>
                                            <td valign="top">No multiple testing methods. Set a fixed p value cutoff by <tt>--p-value-cutoff</tt>.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">benfdr</td>
                                            <td valign="top">Benjamini-Hochberg method, which aims to control the FDR (specified by  <tt>--p-value-cutoff</tt>), or "False Discovery Rate," rather than the FWE or "FamilyWise Error rate".</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">bonhol</td>
                                            <td valign="top">Bonferroni-Holm method, which is less conservative and more powerful than the Standard Bonferroni method given a family-wise error rate specified by <tt>--p-value-cutoff</tt>.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">bonf</td>
                                            <td valign="top">Standard Bonferroni correction given a family-wise error rate specified by <tt>--p-value-cutoff</tt>.</td>
                                        </tr>
                                    </table>

                                    The output of this function is a list of p-values and odds ratios at a variant.
                                    <h4>
                                        <b> PValueAllelic: </b>The p-value calculated by chi-square test for allelic association between case-control status.<br />
                                        <b> OddAllelic: </b>The odds ratio of alleles in cases and controls.<br />
                                        <b> PValueDom: </b>The p-value calculated by chi-square test for genotype association between case-control status under dominant mode.<br />
                                        <b> OddDom: </b>The odds ratio of genotypes in cases and controls under dominant mode.<br />
                                        <b> PValueDom: </b>The p-value calculated by chi-square test for genotype association between case-control status under recessive mode.<br />
                                        <b> OddDom: </b>The odds ratio of genotypes in cases and controls under recessive mode.<br />
                                        <b> PValueGeno: </b>The p-value calculated by chi-square test for genotype association between case-control status.<br />
                                    </h4>

                                    <p><strong>Note</strong> Currently, this function only work for case-control study without confounding factors as covariables. </p>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Variant-based association test by <a href="http://zhanxw.github.io/rvtests/", target="_blank">Rvtests</a>: <tt> --rvtest-var</tt></p>
                                    <p><code>java -jar ./kggseq.jar --rvtest-var score,wald,exact,... --rvtest-vcf --rvtest-remove-set --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene </code>
                                        KGGSeq will automatically download  the C++ sources codes from  <a href="http://zhanxw.github.io/rvtests/", target="_blank">Github</a> and un-compress them into a folder named "plugin/rvtests-master" under the KGGSeq working folder. Users are required go into the "rvtests-master" to make an executable file  from the C++ source codes. See more in <a  href="http://grass.cgs.hku.hk/limx/kggseq/doc10/demov10.mp4" target="_blank"> a demonstration video</a> .
                                    </p>
                                    <table  class="list">
                                        <tr>
                                            <td valign="top"><b>Arguments/values</b></td>
                                            <td valign="top"><b>Annotation</b></td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-var model names</td>
                                            <td valign="top">Initiate the calling of RVTest for gene based association analysis. The model names are defined in <a href="http://zhanxw.github.io/rvtests/#models", target="_blank"><b>Rvtests models</b></a>. Currently, it supports score, wald, exact, dominantExact, famLRT, famScore, famGrammarGamma and firthc models. The default models are "score,wald".                                                                        
                                            </td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-vcf</td>
                                            <td valign="top">Ask KGGSeq to produce a new VCF file as the input of RVTests. However, this is very time-consuming for large sample. It is advised to use <tt>--o-vcf</tt> to produce a VCF file with good quality once and to use that cleaned VCF file as the new KGGSeq input. In this scenario, the <tt>--rvtest-vcf</tt> is not needed, which may substantially reduce the analysis time.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-remove-set</td>
                                            <td valign="top">Remove the intermediate file defining sets of variants for RVTests group-wise test in your hard disk.</td>
                                        </tr>
                                    </table>
                                    <p>An extra file with the suffix ".rvtest.variant.xlsx" will be created in the targeted output path. </p>


                                    <h3><a name="AssociationGene">Association at gene</a></h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Gene-based association test by <a href="http://www.hsph.harvard.edu/skat/", target="_blank">SKAT</a>: <tt> --skat-gene</tt></p>
                                    <p><code>java -jar ./kggseq.jar  --skat-gene --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene [--p-value-cutoff 0.01 --multiple-testing benfdr --qqplot]</code>
                                        The gene-based analysis will be carried out by <a href="http://www.hsph.harvard.edu/skat/",target="_blank">SKAT</a> which was integrated as a plugin in KGGseq. 
                                        The <tt>'--p-value-cutoff'</tt> sets p-value cutoffs for filtration; the <tt>'--multiple-testing'</tt> sets the approaches for multiple testing comparisons; and the <tt>'--qqplot'</tt> enables <a href="http://en.wikipedia.org/wiki/Q-Q_plot", target="_blank">Q-Q plot</a> of the association results.</p>
                                    <p>An extra file with the suffix ".skat.gene.xlsx" will be created in the targeted output path. The <tt>'--skat-gene'</tt> will provide the p values of SKAT,SKATO and Burden test. If the phenotype is binary, the option <tt>'--skat-gene binary'</tt> can be used, but the speed will decrease. </p>

                                    <h4>                                       
                                        <b>SKAT:</b>The p-value calculated by SKAT. <br />
                                        <b>SKATO:</b>The p-value calculated by SKAT-O. <br />
                                        <b>Burden:</b>The p-value calculated by Burden test.<br />
                                    </h4>

                                    <p><strong>Note</strong>  To use this function, the user should install <a href="https://www.r-project.org/", target="_blank">R</a> language platform first. </p>

                                    <p><strong>Note</strong> When total minor allele count (MAC) is very small in a gene, SKATBinary ( <tt>'--skat-gene binary'</tt> ) has better performance than SKAT( <tt>'--skat-gene'</tt> ). However, because SKATBinary uses a resampling approach to get a p-value, it is slower (sometimes quite slower) than SKAT function. When MAC is large, the performance of the two tests is similar.</p>
                                    <p><strong>Note</strong> The names of phenotype ( <tt>'--phe'</tt> ) and co-variables ( <tt>'--cov'</tt> ) are defined in the  <a href="#PhenotypesAndPedigree">pedigree file</a> .  </p>
                                    <p><strong>Note</strong> If the option <tt>--skat-cutoff N</tt> is added, the genes with variants less than <tt>N</tt> will not be involved into SKAT. The default value of <tt>N</tt> is 0.</p>
                                    <p><strong>Note</strong> If the option <tt>--perm-pheno </tt> is added, phenotypes will be permuted prior to association analysis. </p>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Gene-based association test by <a href="http://zhanxw.github.io/rvtests/", target="_blank">Rvtests</a>: <tt> --rvtest-gene</tt></p>
                                    <p><code>java -jar ./kggseq.jar --rvtest-gene cmc,cmat,price,skat[nPerm=1000:alpha=0.001:beta1=1:beta2=20],... --rvtest-vcf --rvtest-remove-set --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene </code>
                                        KGGSeq will automatically download  the C++ sources codes from  <a href="http://zhanxw.github.io/rvtests/", target="_blank">Github</a> and un-compress them into a folder named "plugin/rvtests-master" under the KGGSeq working folder. Users are required go into the "rvtests-master" to make an executable file  from the C++ source codes. See more in <a  href="http://grass.cgs.hku.hk/limx/kggseq/doc10/demov10.mp4" target="_blank"> a demonstration video</a> .
                                    </p>
                                    <table  class="list">
                                        <tr>
                                            <td valign="top"><b>Arguments/values</b></td>
                                            <td valign="top"><b>Annotation</b></td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-gene model names</td>
                                            <td valign="top">Initiate the calling of RVTest for gene based association analysis. The model names are defined in <a href="http://zhanxw.github.io/rvtests/#models", target="_blank"><b>Rvtests models</b></a>. Currently, it supports cmc, zeggini, mb, fp, exactCMC, cmcWald, rarecover and cmat models.The default models are "cmc,skato[nPerm=1000:alpha=0.001:beta1=1:beta2=20]".</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-vcf</td>
                                            <td valign="top">Ask KGGSeq to produce a new VCF file as the input of RVTests. However, this is very time-consuming for large sample. It is advised to use <tt>--o-vcf</tt> to produce a VCF file with good quality once and to use that cleaned VCF file as the new KGGSeq input. In this scenario, the <tt>--rvtest-vcf</tt> is not needed, which may substantially reduce the analysis time.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-remove-set</td>
                                            <td valign="top">Remove the intermediate file defining sets of variants for RVTests group-wise test in your hard disk.</td>
                                        </tr>
                                    </table>

                                    <p>An extra file with the suffix ".rvtest.gene.xlsx" will be created in the targeted output path. </p>


                                    <h3><a name="MutationRateGene" >Mutation rate test at gene</a></h3>
                                    <br />                                  
                                    Test the certain types mutation rate in a gene by a regression mode <tt>--gene-mutation-rate-test --dependent-gene-feature [variants&apos; gene feature codes] --independent-gene-feature [variants&apos; gene feature codes]</tt> 
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname [other tags] --gene-mutation-rate-test</code>

                                    The model can be expressed as the following:<br/>
                                    <center>[#minor alleles of variants specified by &apos;--dependent-gene-feature&apos;]=[#minor alleles of variants specified by &apos;--independent-gene-feature&apos;] + [length of coding region] + [constraint score]+[expression]+[DNA replication timing]+[HiC-based chromatin state estimation]</center> <br/><br/>
                                    The gene&apos;s constraint scores are from <a href="http://www.nature.com/ng/journal/v46/n9/abs/ng.3050.html" target="_blank">Samocha et al (2014) Nature Genetics. </a>. The last three factors are from <a href="http://www.broadinstitute.org/cancer/software/genepattern/modules/docs/MutSigCV" target="_blank">MutSigCV</a>.The assumption of the analysis is that the number of minor alleles of specified dependent variants in a gene can be normally explained by other factors of the same gene including length of coding region, the number of minor alleles of specified independent variants, constraint score and other expression related features on the human genome. However, if a gene contributes to a certain disease by mutations at specified dependent variants, this gene tends to have more mutant alleles at the specified dependent variants in a sample of this disease.

                                    <p><strong>Note</strong> By default, the setting of <tt>--dependent-gene-feature</tt> is 0,1,2,3,4,5,6 (i.e., all non-synonymous variants) and the setting of <tt>--independent-gene-feature</tt> is 7 (i.e., the synonymous variants only).
                                        <br/> <strong>Note</strong> The gene feature code can be found <a href="#GeneFeatureFiltering" target="_self">HERE</a>.                                    
                                    </p> 
                                    This approach is designed for test gene &apos;s mutation rate in three scenarios.<br/>
                                    1) In cancer samples for gene&apos;s somatic mutation rate: 
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --genotype-filter 8 --indiv-pair tumorIndivID1:normalIndivID1,tumorIndivID2:normalIndivID2 --db-gene refgene --gene-mutation-rate-test --dependent-gene-feature 0,1,2,3,4,5,6 --independent-gene-feature 7</code>

                                    2)In a cohort for gene&apos;s rare non-synonymous mutation rate : 
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.01  --db-gene refgene --gene-mutation-rate-test --dependent-gene-feature 0,1,2,3,4,5,6 --independent-gene-feature 7</code>

                                    <!--
                                     2) In trios samples for gene&apos;s <i>de novo</i> mutation rate: 
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --genotype-filter 7 --db-gene refgene --gene-mutation-rate-test --dependent-gene-feature 0,1,2,3,4,5,6 --independent-gene-feature 7</code>


                                    4)In case-control samples for gene&apos;s case-unique rare non-synonymous mutation rate: 
                                    <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.01 --filter-model case-unique --db-gene refgene --gene-mutation-rate-test --dependent-gene-feature 0,1,2,3,4,5,6 --independent-gene-feature 7</code>
                                    !-->
                                    In the output file of the analysis (*.gene.mutation.rate.xlsx), there are detailed values of all factors used in the regression and the p-values of mutation rate test.
                                    <h4>
                                        <b>GeneSymbol</b>: Official gene symbols defined by HGNC.<br />
                                        <b>AvgCodingLen</b>: The actual or averaged (of multiple isoforms) length (in kb) of all exons in a gene.<br />
                                        <b>NS</b>: The observed number of non-synonymous and splicing somatic sequence variants.<br />
                                        <b>Syn</b>: The observed number of synonymous somatic sequence variants.<br />
                                        [...] <br/>
                                        <b>RegResidue:</b>  Pearson residuals   (raw residuals scaled by square root of variance function) in the regression after correction for the factors. <br />
                                        <b>RegP:</b>  The probability of be equal or over the RegResidue under a standard normal distribution.<br />


                                    </h4>
                                    Note: The detailed method of the analysis will be described in a paper, which is in preparation.
                                    <br />

                                    <h3><a name="GeneSetAssociation">Association at geneset</a></h3>
                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Geneset-based association test by <a href="http://www.hsph.harvard.edu/skat/", target="_blank">SKAT</a>: <tt> --skat-geneset</tt></p>
                                    <p><code>java -jar ./kggseq.jar  --skat-geneset --geneset-file path/to/geneset/file --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene [--p-value-cutoff 0.01 --multiple-testing benfdr --qqplot]</code>
                                        The gene-based analysis will be carried out by <a href="http://www.hsph.harvard.edu/skat/",target="_blank">SKAT</a> which was integrated as a plugin in KGGseq. 
                                        The <tt>'--p-value-cutoff'</tt> sets p-value cutoffs for filtration; the <tt>'--multiple-testing'</tt> sets the approaches for multiple testing comparisons; and the <tt>'--qqplot'</tt> enables <a href="http://en.wikipedia.org/wiki/Q-Q_plot", target="_blank">Q-Q plot</a> of the association results.</p>

                                    <strong>Note</strong>, the pathway gene sets are specified in a text file in th following format:<br/>

                                    <h4>
                                        [Column 1: GeneSet ID]<br />
                                        [Column 2: GeneSet URL]<br />
                                        [Column 3..N: Gene Symbols in the geneset; separated by spaces].<br />

                                        No title row is required!!!<br />
                                        <br />

                                        <br />    e.g.,<br />
                                        ----------------------------------------------<br />
                                        KEGG_GLYCOLYSIS_GLUCONEOGENESIS http://www.broadinstitute.org/KEGG_GLYCOLYSIS_GLUCONEOGENESIS.html LDHC LDHB LDHA<br />
                                        KEGG_CITRATE_CYCLE_TCA_CYCLE http://www.broadinstitute.org/KEGG_CITRATE_CYCLE_TCA_CYCLE.html GJB1 OGDHL OGDH PDHB IDH3G LDHB<br />
                                        ...<br />
                                        ...<br />
                                    </h4>

                                    <p>An extra file with the suffix ".skat.geneset.xlsx" will be created in the targeted output path. The <tt>'--skat-geneset'</tt> will provide the p values of SKAT,SKATO and Burden test. If the phenotype is binary, the option <tt>'--skat-gene binary'</tt> can be used, but the speed will decrease. </p>

                                    <h4> 
                                        <b>SKAT:</b>The p-value calculated by SKAT. <br />
                                        <b>SKATO:</b>The p-value calculated by SKAT-O. <br />
                                        <b>Burden:</b>The p-value calculated by Burden test.<br />
                                    </h4>

                                    <p><strong>Note</strong>  To use this function, the user should install <a href="https://www.r-project.org/", target="_blank">R</a> language platform first. </p>

                                    <p><strong>Note</strong> When total minor allele count (MAC) is very small in a gene, SKATBinary ( <tt>'--skat-gene binary'</tt> ) has better performance than SKAT( <tt>'--skat-gene'</tt> ). However, because SKATBinary uses a resampling approach to get a p-value, it is slower (sometimes quite slower) than SKAT function. When MAC is large, the performance of the two tests is similar.</p>
                                    <p><strong>Note</strong> The names of phenotype ( <tt>'--phe'</tt> ) and co-variables ( <tt>'--cov'</tt> ) are defined in the  <a href="#PhenotypesAndPedigree">pedigree file</a> .  </p>
                                    <p><strong>Note</strong> If the option <tt>--skat-cutoff N</tt> is added, the genes with variants less than <tt>N</tt> will not be involved into SKAT. The default value of <tt>N</tt> is 0.</p>
                                    <p><strong>Note</strong> If the option <tt>--perm-pheno </tt> is added, phenotypes will be permuted prior to association analysis. </p>

                                    <p><img src="imgs/leaf.gif" width="10" height="8" /> Geneset-based association test by <a href="http://zhanxw.github.io/rvtests/", target="_blank">Rvtests</a>: <tt> --rvtest-geneset</tt></p>
                                    <p><code>java -jar ./kggseq.jar  --rvtest-geneset cmc,skato[nPerm=1000:alpha=0.001:beta1=1:beta2=20],... --rvtest-vcf --rvtest-remove-set --geneset-file path/to/geneset/file --vcf-file path/to/vcf/file --ped-file path/to/pedigree/file --phe phenotypeName --cov covariateName1,covariateName2,...,covariateNameN --out path/to/prefixname --excel --db-gene refgene [--p-value-cutoff 0.01 --multiple-testing benfdr --qqplot]</code>
                                        KGGSeq will automatically download  the C++ sources codes from  <a href="http://zhanxw.github.io/rvtests/", target="_blank">Github</a> and un-compress them into a folder named "plugin/rvtests-master" under the KGGSeq working folder. Users are required go into the "rvtests-master" to make an executable file  from the C++ source codes. See more in <a  href="http://grass.cgs.hku.hk/limx/kggseq/doc10/demov10.mp4" target="_blank"> a demonstration video</a> .
                                    </p>
                                    <table  class="list">
                                        <tr>
                                            <td valign="top"><b>Arguments/values</b></td>
                                            <td valign="top"><b>Annotation</b></td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-gene model names</td>
                                            <td valign="top">Initiate the calling of RVTest for gene based association analysis. The model names are defined in <a href="http://zhanxw.github.io/rvtests/#models", target="_blank"><b>Rvtests models</b></a>. Currently, it supports cmc, zeggini, mb, fp, exactCMC, cmcWald, rarecover and cmat models.The default models are "cmc,skato[nPerm=1000:alpha=0.001:beta1=1:beta2=20]".</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-vcf</td>
                                            <td valign="top">Ask KGGSeq to produce a new VCF file as the input of RVTests. However, this is very time-consuming for large sample. It is advised to use <tt>--o-vcf</tt> to produce a VCF file with good quality once and to use that cleaned VCF file as the new KGGSeq input. In this scenario, the <tt>--rvtest-vcf</tt> is not needed, which may substantially reduce the analysis time.</td>
                                        </tr>
                                        <tr>
                                            <td valign="top">--rvtest-remove-set</td>
                                            <td valign="top">Remove the intermediate file defining set of variants for RVTests group-wise test.</td>
                                        </tr>
                                    </table>

                                    <p>An extra file with the suffix ".rvtest.geneset.xlsx" will be created in the targeted output path. The <tt>'--rvtest-geneset'</tt> will provide the  p values of multiple set-based tests for rare variants.  </p>


                                    <h3><a name="GeneSetEnrichment" >Enrichment at geneset</a></h3>                                  
                                    <p>Test whether a  set of genes highly ranked in an ordered list of all genes by Wilcoxon rank sum test:<tt> --geneset-enrichment-test </tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname [other tags] --geneset-enrichment-test</code>

                                        Basically, this analysis is to examine whether a  set of genes are more highly ranked in an ordered list of all genes than would be expected by chance. Therefore,as the rank of gene is a prerequisite, it should be used together with a gene-based statistical test.<br/>


                                        Currently, this enrichment test can only be used together with the gene-based mutation rate test:
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --db-filter 1kg201204,dbsnp137,ESP6500AA,ESP6500EA  --rare-allele-freq 0.01  --db-gene refgene --gene-mutation-rate-test --geneset-enrichment-test --geneset-db &lt;GenesetDatabase&gt;</code>

                                        <p>There are 5 geneset databases from <a href="http://www.broadinstitute.org/gsea/msigdb/collections.jsp">GSEA</a>. The following are brief descriptions of the 5 datasets COPIED from the <a href="http://www.broadinstitute.org/gsea/msigdb/collections.jsp" target="_blank">page</a> of GSEA.<br/>
                                            <table width="90%" class="list">
                                                <tr>
                                                    <th width="19%" scope="col">GeneSetDatabase</th>
                                                    <th width="66%" scope="col">Description</th>
                                                    <th width="15%" scope="col">Version</th>
                                                </tr>
                                                <tr>
                                                    <td>cano</td>
                                                    <td>Gene sets from the geneset databases. Usually, these gene sets are canonical representations of a biological process compiled by domain experts. (1452 gene sets)</td>
                                                    <td> MSigDB 3.1</td>
                                                </tr>
                                                <tr>
                                                    <td> cura </td>
                                                    <td>Curated gene sets: Gene sets collected from various sources such as online geneset databases, publications in PubMed, and knowledge of domain experts. The gene set page for each gene set lists its source. (4850 gene sets)</td>
                                                    <td> MSigDB 3.1</td>
                                                </tr>
                                                <tr>
                                                    <td> onco </td>
                                                    <td> Oncogenic signatures: Gene sets represent signatures of cellular pathways which are often dis-regulated in cancer. The majority of signatures were generated directly from microarray data from NCBI GEO or from internal unpublished profiling experiments which involved perturbation of known cancer genes. In addition, a small number of oncogenic signatures were curated from scientific publications. (189 gene sets)</td>
                                                    <td> MSigDB 3.1</td>
                                                </tr>
                                                <tr>
                                                    <td> cmop </td>
                                                    <td>Computational gene sets:  Computational gene sets defined by mining large collections of cancer-oriented microarray data. (858 gene sets)</td>
                                                    <td> MSigDB 3.1</td>
                                                </tr>
                                                <tr>
                                                    <td> onto </td>
                                                    <td>GO gene sets Gene sets are named by GO term and contain genes annotated by that term. GSEA users: Gene set enrichment analysis identifies gene sets consisting of co-regulated genes; GO gene sets are based on ontologies and do not necessarily comprise co-regulated genes. (1454 gene sets)</td>
                                                    <td> MSigDB 3.1</td>
                                                </tr>
                                            </table>
                                            <br />
                                            <strong>Alternatively</strong>, uses can specify the own pathway gene sets in a text file by <tt>--geneset-file path/to/geneset/file</tt>. <br/>


                                            <strong>Note </strong>, please use <tt>--min-geneset X</tt> to ignore gene sets with less X genes and --max-geneset to ignore gene sets with over X genes.
                                            <br />

                                            In the output file of the enrichment analysis (*.enrichment.geneset.xls), there are detailed statistical evaluation results of each geneset:
                                            <h4>
                                                <b>GeneSetID</b>:the input geneset name or id<br />
                                                <b>WilcoxonP</b>:the p-values of the enrichment test<br />
                                                <b>Gene</b>:the gene of the gene set<br />
                                                <b>GeneP</b>:the p-value of the gene used for the enrichment test<br />
                                            </h4>
                                            <br />
                                        </p>
                                        <p>&nbsp;</p>
                                </td>
                            </tr>


                            <tr>
                                <td><h2><a name="Plotting" >Plotting </a></h2>
                                    <p>Generate a histogram of minor allele frequency (MAF) according to  allele frequencies in reference dataset <tt>--mafplot-ref</tt> or in sample <tt>--mafplot-sample</tt>. 
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname --db-filter 1kg201204,dbsnp141,ESP6500AA,ESP6500EA  --allele-freq 0,1  --mafplot-ref</code>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file1 --ped-file path/to/file2 --out path/to/prefixname  --mafplot-sample</code>
                                        <strong> Note</strong> Currently, --mafplot-sample only supports the inputs specified by --vcf-file.<br />
                                        <br />

                                        Generate a Quantile-Quantile (Q-Q) plot for association p-values <tt>--qqplot</tt>
                                        <code>java -jar kggseq.jar --vcf-file path/to/file --out path/to/prefixname --qqplot --skat-gene [or] --var-assoc</code>                                        
                                        <br />
                                    </p>


                                </td>
                            </tr>

                            <tr>
                                <td><h2><a name="Miscellaneous" >Miscellaneous </a></h2>
                                    <p>Calculate pair-wise the linkage disequilibrium or genotypic correlation  <tt>--calc-ld</tt>. 
                                        <code>java -jar kggseq.jar --vcf-file path/to/file --ped-file path/to/file --out path/to/prefixname --calc-ld --db-filter 1kg201204,dbsnp141,ESP6500AA,ESP6500EA  --allele-freq 0.05,0.95   --regions-in chr4:21212-233454</code>                                       
                                        If the genotypes are phased, KGGSeq will calculate the linkage disequilibrium coefficient r-square. For uphased genotypes, KGGSeq will calculate the Pearson genotypic correlation of variants, in which the genotypes are encoded by the number of alternative alleles.
                                        <br /> 
                                        <strong> Note</strong> Currently, --calc-ld only supports variants with two alleles.<br />

                                        <p>The results will be outputted in a text file  with the suffix ".ld" or ".ld.gz".  The following is an exmaple of the output format:
                                            <h4> 
                                                [chromsome] [coordinate of the first variant] [coordinate of the second variant] [r-square OR correlation]<br />
                                                22	16051249	16052962	0.8680002633575351<br />
                                                22	16051249	16053659	0.13732498950210686<br />
                                                22	16051249	16053862	0.984494959737711
                                            </h4>
                                        </p>

                                    </p>


                                </td>
                            </tr>

                            <tr>
                                <td><hr />                                   
                                    <h1> <a name="faq">FAQ?</a></h1>
                                    <p><b>1) Can a compressed VCF file be processed in parallel?</b></p>
                                    <p>Yes, it can. However, the files has to be compressed in a blocked GNU Zip format or Burrows&#8211;Wheeler format. See more here.</p>
                                    <p><b>2) Does the KGGSeq binary genotype format support variants of over two alleles and/or phased genotypes?</b></p>
                                    <p>Yes, it does support variants of over two alleles and/or phased genotypes. See more here.</p>
                                    <p><b>3) How are the settings '--rare-allele-freq c' and '--allele-freq 0,c' different?</b></p>
                                    <p>  '--rare-allele-freq c' will excluded variants with alternative allele frequency EQUAL to or over c in the reference datasets; and '--allele-freq 0,c' will also excluded variants having no allele frequency in the reference datasets beside the variants excluded by '--rare-allele-freq c'</p> 
                                    <p><b>4) Does KGGSeq provide gene expression information in the output?</b></p>
                                    <p>A project aiming to harness tissue-specific expression information to prioritize candidate genes is underway.</p>
                                    <p><b>5) How shall I prepare the candidate gene list for protein-interaction and geneset annotation? What is the gene  list for?</b></p>
                                    <p>The candidate gene list can include any gene interesting to the user. It  may be verified causal gene for the disease, or potential causal gene. The gene  list is used to fish more hidden interesting genes in order to narrow down the  searching region. If no gene list provided, KGGSeq can still perform routine  &lsquo;filtration + sharing&rsquo; strategy for sequencing data.</p>
                                    <p><b>6)Suppose I have in-house control genomes in VCF format, how can I  transform it into KGGSeq acceptable file?</b></p>
                                    <p>There are two alternative ways to allow you use the in-house VCF data (containing  genotypes of multiple individuals) as filter.</p>
                                    <p>1. Convert the VCF data into a standard KGGSeq local filter format by the  command: <em>--make-filter --vcf-file path/to/vcffile --out  test.var </em>first. And then set the --local-filter <em>test.hg19.var</em></p>
                                    <p>2. Directly set <em>--local-filter-vcf path/to/vcffile. </em>However, this  way is less efficient than the above way as it will take additional time to  parse the VCF data.</p>
                                    <p><b>7) Is there a practical guidance for MAF  selection?</b></p>
                                    <p> MAF selection is an open question, but generally three suggestions can be adopted: for rare dominant Mendelian disease, MAF cut-off at 0 or 0.005 is reasonable; for rare recessive Mendelian disease, MAF cut-off at 0 or 0.03 is reasonable;
                                        for common complex disease of  relatively rare alleles with large effect size, MAF cut-off is dependent on  disease prevalence, penetrance and genetic model, so a higher cut-off like 0.05  might be more reasonable.</p>
                                    <p><b>8) Can I use ANNOVAR and KGGSeq together on my dataset?</b></p>
                                    <p>KGGSeq is quite flexible for interacting with other sequence-oriented  analytical programs/software (including SOAPSNP, ANNOVAR, PLINK/seq, VAAST, etc).  It can accept various input formats, and output different kinds of sequence  data. In case of ANNOVAR, KGGSeq can read ANNOVAR-formatted sequence variants,  and write the final remaining variants in ANNOVAR format.</p>
                                    <p><b>9) Can I run KGGSeq on my Lenovo or Mac laptop? Is it time consuming  to run a complete KGGSeq process?</b></p>
                                    <p>Normally, KGGSeq run well and fast with &gt;=1 GB RAM memory. Hence  current laptop are certainty affordable for running KGGSeq. The whole process  need only &lt;10 minutes, unless PubMed searching is set in the parameter file.  For PubMed searching, it really depends on the speed of the network and also  the loading of the NCBI PubMed database.</p>
                                    <p><b>10) How do I report an error  or bugs to KGGSeq?</b></p>
                                    <p>You are welcomed to join our  google group (<a href="http://groups.google.com/group/kggseq_user?hl=en">http://groups.google.com/group/kggseq_user?hl=en</a>).  This site is used for communication and discussion of KGGSeq usage and  functions. You can also directly write an email to <a href="mailto:limx54@yahoo.com">limx54@yahoo.com</a>, or  kggseq_user@googlegroups.com.</p>
                                </td>
                            </tr>
                        </table>



                        <p>&nbsp;</p>
                    </td>
                </tr>
            </table>
        </div>

        <div  id="menu">
            <a href="#Navi:introduction">References</a><br />
            <a href="#Navi:installation">Installation</a><br />
            <a href="#Navi:shorttutorials">Tutorials</a>
            <a href="#MendelianDiseaseHg19">Examples</a><br />
            <a href="#Navi:options">Options</a>
            <a name="Navi:faq" href="#faq"> FAQ</a>

        </div>
    </body>
</html>
>>>>>>> origin/master
